---
title: "HipSci_iPSC_PRS"
author: "Joni Coleman [JRIC]"
date: "30/03/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## NOTE

This is not a true RMarkdown - it will not work if you run it in R! It is a convenient format in which to keep the analytical pipeline for this project.

## Background

Data in this project comes from the HipSci consortium (http://www.hipsci.org/).
JRIC obtained permission to use this data to assess the effects of polygenic risk scores for psychiatric disorders on iPSC cell morphological traits (2019-12-12).
iPSCs were drawn from "healthy" control lines (see note below)

Data required:
    - Imaging data
        - This is available open-access and is described e.g. here: http://www.hipsci.org/lines/#/lines/HPSI0513i-euir_1/cellbiol-fn
        - This was downloaded from HipSci on 2019-26-11 via wget from ftp://ftp.hipsci.ebi.ac.uk/vol1/ftp/data/ 
    - Genetic data
        - This is a mixture of open-access and controlled access data (dependent on the original consent into HipSci or the Cambridge Bioresource)
        - Open access data was downloaded from ENA (https://www.ebi.ac.uk/ena/browser/view/PRJEB11750) 2020-01-09 via wget from ftp://ftp.sra.ebi.ac.uk/vol1/
        - Access to controlled-access data was provided to JRIC by HipSci via EGA. 
            - Data was downloaded from EGA (https://ega-archive.org/) via the EGA downloader, commencing 2020-01-06 

Notes:
    - "Healthy" control lines
        - HipSci draws its data primarily from the rare disorders field 
        - It has a number of cell lines from individuals with various rare or uncommon disorders, none of which are psychiatric.
        - As such, the control status of the individuals used in this study is relative to those disorders, not to common disorders including psychiatric 
        - As such, it is possible that some of these individuals may have mental health conditions. 
        - No phenotypic information is available on these individuals, and so we cannot ascertain disorder status.

## Intiate folder structure

```{sh Initiate_Folders}
# iPSC PRS folder exists from previous projects
cd iPSC_PRS

mkdir -p HipSci/Genotypes
mkdir -p HipSci/Images

```

## Get image data

```{sh Get_Images}
## Make sub-folders
mkdir -p HipSci/Images/Cell_Results
mkdir -p HipSci/Images/Original_Files
mkdir -p HipSci/Images/Plate_Results

## Get data
cd iPSC_PRS/HipSci/Images/Original_Files
wget -r ftp://ftp.hipsci.ebi.ac.uk/vol1/ftp/data/cellbiol-fn/

## Clean up - remove hashed for safety!
mv ftp.hipsci.ebi.ac.uk/vol1/ftp/data/cellbiol-fn/* .
#rm -r ftp.hipsci.ebi.ac.uk

## Make softlinks in relevant subfolders
cd ../Cell_Results
ln -s ../Original_Files/release/*/*/*cell_results*.tsv .

cd ../Plate_Results
ln -s ../Original_Files/release/*/*/*plate_results*.tsv .
```

## Get OA genotypes

```{sh Get_OA_Genotypes}

## Make subfolder
cd iPSC_PRS/HipSci/Genotypes
mkdir OpenAccess
cd OpenAccess

## Get data
wget -r ftp://ftp.sra.ebi.ac.uk/vol1/ERZ127/ERZ12747*
wget -r ftp://ftp.sra.ebi.ac.uk/vol1/ERZ127/ERZ12748*
wget -r ftp://ftp.sra.ebi.ac.uk/vol1/ERZ127/ERZ12749*
wget -r ftp://ftp.sra.ebi.ac.uk/vol1/ERZ127/ERZ1275*
wget -r ftp://ftp.sra.ebi.ac.uk/vol1/ERZ127/ERZ1276*
wget -r ftp://ftp.sra.ebi.ac.uk/vol1/ERZ127/ERZ12770*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ368/ERZ368718
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ368/ERZ368719
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ368/ERZ36872*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ368/ERZ36873*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ368/ERZ36874*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ368/ERZ36875*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ368/ERZ36876*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ368/ERZ36877*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ368/ERZ36878*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ368/ERZ36879*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ368/ERZ3688*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ368/ERZ3689*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ369/ERZ36900*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ370/ERZ370130
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ494044
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ494045
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ494046
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ494047
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ494048
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ494049
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ49405*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ49406*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ49407*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ49408*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ49409*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ4941*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ49420*
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ494210
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ494211
wget -r ftp.sra.ebi.ac.uk/vol1/ERZ494/ERZ494212

## Clean up
mv -n ftp.sra.ebi.ac.uk/vol1/*/*/* .
#rm -rf ftp.sra.ebi.ac.uk
#rm -rf E*.md5
# Don't really need to keep the md5s 
```

## Get controlled access genotype data

```{Get CA genotype data}
## Softlink EGA client (previously used to get UK Biobank data) into a more local path
mkdir Tools
cd Tools
ln -s Software/EgaDemoClient/EgaDemoClient.jar EgaDemoClient.jar

## Make subfolder
cd iPSC_PRS/HipSci/
mkdir Files_Available

## Get data
cd Files_Available

for dataset in \
EGAD00001000893 \
EGAD00001000897 \
EGAD00001001437 \
EGAD00001001438 \
EGAD00001001932 \
EGAD00001001933 \
EGAD00001003514 \
EGAD00001003529 \
EGAD00010000564 \
EGAD00010000568 \
EGAD00010000768 \
EGAD00010000771 \
EGAD00010000773 \
EGAD00010000775 \
EGAD00010001139 \
EGAD00010001143 \
EGAD00010001147 \
EGAD00010001328 \
EGAD00010001330
do 
   java -Xms256m -Xmx512m -jar ../../../Tools/EgaDemoClient.jar -p jonathan.coleman@kcl.ac.uk PASSWORD -lfd $dataset > ${dataset}_Files.txt
done

#Password is redacted
#Files include data other than genotypes (expression, methylation, sequence). Don't require this data at the moment (and would need new DTA to use it), so did not download

## Identify datasets with genotype data
LANG=C fgrep genotypes.vcf.gz.cip *Files.txt | LANG=C fgrep HumanCoreExome-12_v1 > Datasets_With_Genotypes.txt

for file in EGAD00010000773 EGAD00010001147 EGAD00010001328
do 
    LANG=C fgrep -wv idat ${file}_Files.txt | LANG=C fgrep -wv gtc.cip > ${file}_Download.sh
done

## Manually edit files in emacs to following structure per line:
#     java -Xms256m -Xmx512m -jar ../../../../../Tools/EgaDemoClient.jar -p jonathan.coleman@kcl.ac.uk PASSWORD -rf EGAF00001056572 -re opensesame -label request_EGAD00010000773
# Password is redacted, filenames vary per line

## Make further subfolders and clean up
cd iPSC_PRS/HipSci/Genotypes
mkdir ControlledAccess
mv ../../Files_Available/*_Download.sh .
sh ./*_Download.sh

## Requests sometimes fail, so rerun commands until all downloads are complete - successful downloads are automatically skipped by the downloader
cd iPSC_PRS/HipSci/Genotypes/ControlledAccess
java -Xms256m -Xmx512m -jar ../../../../../Tools/EgaDemoClient.jar -p jonathan.coleman@kcl.ac.uk PASSWORD -dr request_EGAD00010000773 -nt 7
java -Xms256m -Xmx512m -jar ../../../../../Tools/EgaDemoClient.jar -p jonathan.coleman@kcl.ac.uk PASSWORD -dr request_EGAD00010001147 -nt 7
java -Xms256m -Xmx512m -jar ../../../../../Tools/EgaDemoClient.jar -p jonathan.coleman@kcl.ac.uk PASSWORD -dr request_EGAD00010001328 -nt 7

for file in *.cip; do java -Xms256m -Xmx512m -jar ../../../../../Tools/EgaDemoClient.jar -p jonathan.coleman@kcl.ac.uk PASSWORD -dc $file -dck opensesame; done

```

Merge imputed data to single file - NOT RUN (see below)

```{sh Merge_Imputed_NOT_RUN}
cd iPSC_PRS/HipSci/Genotypes

## Make list of imputed vcf files
ls iPSC_PRS/HipSci/Genotypes/*/*imputed_phased*genotypes.vcf.gz > iPSC_PRS/HipSci/Genotypes/Imputed_Phased_VCFs_List.txt

## Merge VCFs
module add bioinformatics/bcftools/1.9
bcftools merge -l Imputed_Phased_VCFs_List.txt -o HPS_Imputed_Phased.vcf.gz -Oz --threads 7
# This fails, and is actually an unnecessary use of space for this project…
# There is some benefit to having extra data for QC, but the ~100 samples with imaging data are acceptable
```

Limit genetic data to imaging results

```{sh Get_Imaging_Data_Genetics}
## List unique cell_results data
cd iPSC_PRS/HipSci/Images/Cell_Results
ls | sed 's/\.cellbiol.*//g' | sort | uniq > ../Unique_Cell_Result_Lines.txt
ls | sed 's/_.*//g' | sort | uniq > ../Unique_Cell_Result_Individuals.txt
# There are 111 unique cell lines from 76 individuals


## Get all genetic data for these 76 individuals (will favour same cell line data, but will use other cell line data if same cell line is not available)
cd ../Genotypes/
cp ../Images/Unique_Cell_Result_Individuals.txt Find_Unique_Cell_Result_Individuals.sh

# Manually edit Find_Unique_Cell_Result_Individuals.sh to the format below:
#    find ./ -name *HPSI0213i-koun* > Unique_Cell_Result_Individuals_Genetic_Data.txt
#    find ./ -name *HPSI0313i-airc* >> Unique_Cell_Result_Individuals_Genetic_Data.txt
#    etc…

sh ./Find_Unique_Cell_Result_Individuals.sh
cp ../Images/Unique_Cell_Result_Lines.txt Find_Unique_Cell_Result_Lines.sh

## Check lines to see how many lines have an accompanying genotype file

# Manually edit Find_Unique_Cell_Result_Lines.sh to the format below:
#    find ./ -name *HPSI0213i-koun_2.* > Unique_Cell_Result_Lines_Genetic_Data.txt
#    find ./ -name *HPSI0313i-airc_2.* >> Unique_Cell_Result_Lines_Genetic_Data.txt

sh ./Find_Unique_Cell_Result_Lines.sh

LANG=C fgrep -v .tbi Unique_Cell_Result_Individuals_Genetic_Data.txt | LANG=C fgrep imputed_phased | wc -l
# 197
LANG=C fgrep -v .tbi Unique_Cell_Result_Lines_Genetic_Data.txt | LANG=C fgrep imputed_phased | wc -l
# 94 - SUPERSEDED BELOW 
# Not every cell line has imputed, phased data - SUPERSEDED BELOW

# Check if every individual has at least one set of imputed phased data
# This was done manually, stored here: (Imaging_Data_Individuals_Genetic_Data)[https://docs.google.com/spreadsheets/d/1Uirl54GWYPrAhIR4_LSMvQo2usay2qTqHTw36q0xbU0/edit#gid=0]
# Only 63/76 have imputed_phased data  - SUPERSEDED BELOW
#     6 of the missing 13 have data, but not imputed, phased data. - SUPERSEDED BELOW
#         Download issue - this data is available, retrieved 2019-01-14
#         Now have imputed, phased data for all lines with cell imaging data
#         Also downloaded all imputed, phased data for all line from all individuals with cell imaging data
#     7 are not controls, so these are excluded

# Rerun get genetic data from above

## Check lines to see how many lines have an accompanying genotype file

cd iPSC_PRS/HipSci/Genotypes/
sh ./Find_Unique_Cell_Result_Individuals.sh
sh ./Find_Unique_Cell_Result_Lines.sh
LANG=C fgrep -v .tbi Unique_Cell_Result_Individuals_Genetic_Data.txt | LANG=C fgrep imputed_phased | wc -l
# 215 -SUPERSEDED BELOW
# This not all of the lines…?
#     The fibroblast lines do not have the same prefix - still want to retain these at this point, so altered Find_Unique_Cell_Result_Individuals.sh to the following
#         find ./ -name *koun* > Unique_Cell_Result_Individuals_Genetic_Data.txt
#         find ./ -name *airc* >> Unique_Cell_Result_Individuals_Genetic_Data.txt
#         etc…

sh ./Find_Unique_Cell_Result_Individuals.sh
LANG=C fgrep -v .tbi Unique_Cell_Result_Individuals_Genetic_Data.txt | LANG=C fgrep imputed_phased | wc -l
# 280
# This is all of the lines

LANG=C fgrep -v .tbi Unique_Cell_Result_Lines_Genetic_Data.txt | LANG=C fgrep imputed_phased | wc -l
# 103
# This is all of the lines

## Make new files for merge
LANG=C fgrep -v .tbi Unique_Cell_Result_Individuals_Genetic_Data.txt | LANG=C fgrep imputed_phased > Unique_Cell_Result_Individuals_VCFs_Only.txt
LANG=C fgrep -v .tbi Unique_Cell_Result_Lines_Genetic_Data.txt | LANG=C fgrep imputed_phased > Unique_Cell_Result_Lines_VCFs_Only.txt
sh ./Make_Targeted_VCF.sh


cd iPSC_PRS/HipSci/Genotypes
module add bioinformatics/bcftools/1.9
bcftools merge -l Unique_Cell_Result_Individuals_VCFs_Only.txt -o Unique_Cell_Result_Individuals_Imputed_Phased.vcf.gz -Oz --threads 8

# Error in download of HPSI1213i-nusw_1 - redownloaded.

## Remove bad HPSI1213i-nusw_1 data and replace with good
cd iPSC_PRS/HipSci/Genotypes
bcftools view -s ^HPSI1213i-nusw_1 Unique_Cell_Result_Individuals_Imputed_Phased.vcf.gz -Oz --threads 1 -o Unique_Cell_Result_Individuals_Imputed_Phased.vcf.gz_TEMP
bcftools merge  -o Unique_Cell_Result_Individuals_Imputed_Phased_20200115.vcf.gz  -Oz --threads 1 Unique_Cell_Result_Individuals_Imputed_Phased.vcf.gz_TEMP ./OpenAccess/HPSI1213i-nusw_1.wec.gtarray.HumanCoreExome-12_v1_0.imputed_phased.20150604.genotypes.vcf.gz
#rm Unique_Cell_Result_Individuals_Imputed_Phased.vcf.gz Unique_Cell_Result_Individuals_Imputed_Phased.vcf.gz_TEMP

## Convert to PLINK binary

cd iPSC_PRS/HipSci/Genotypes

Software/plink \
--vcf Unique_Cell_Result_Individuals_Imputed_Phased_20200115.vcf.gz \
--make-bed \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115
# 280 individuals (cell lines)
# 42,036,610 variants

## Remove data from individuals without imaging data 
cd iPSC_PRS/HipSci/Genotypes/Controlled_Access/

# Don't delete scripts for getting data 
mv EGAD0001000* ..

# Add rm manually - could do this on command line, but dangerous code!
ls | LANG=C fgrep -vf <(sed 's/.*-//g' ../../Images/Unique_Cell_Result_Individuals.txt) > Remove_File.sh

sh ./Remove_File.sh
#rm Remove_File.sh*
cd ../Open_Access
ls | LANG=C fgrep -vf <(sed 's/.*-//g' ../../Images/Unique_Cell_Result_Individuals.txt) > Remove_File.sh
sh ./Remove_File.sh
#rm Remove_File.sh*
```

## Assess data quality and perform basic QC

```{sh Basic_QC}
## Basic PLINK Checks (MAF, Call rate, HWE)

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115 \
--missing \
--freq \
--hardy \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_Checks

## Remove MAF < 0.001 (which is just the monomorphic variants) and delete the old file (don't need to keep so many 0/0s!)

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115 \
--maf 0.001 \
--make-bed \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001

# 280 individuals (cell lines)
# 10,931,966 variants

#rm Unique_Cell_Result_Individuals_Imputed_Phased_20200115.bed \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115.bim \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115.fam \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115.log \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115.nosex

## Store checks for future reference
mkdir Intermediate_Files
mv *Checks* Intermediate_Files

## Assess data quality - same checks as above

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001 \
--missing \
--freq \
--hardy \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_Checks

## Check whether missing rate differs between individuals in R (script below)

module add general/R/3.4.1
R

```

```{r Missing_Rate_Check}
library(data.table)

Miss_Indiv<-fread("Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_Checks.imiss", data.table=F)

summary(Miss_Indiv$F_MISS)

# Min = 1.56e-6
# 1Q = 4.92e-6
# Median = 6.40e-6
# Mean = 6.32e-6
# 3Q = 7.78e-6
# Max = 1.12e-5

# Differences in call rate are trivial
```

```{sh Further QC}

# Confirm expected duplicates by pruning for LD and running IBS analysis

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001 \
--indep-pairwise 1500 150 0.2 \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_LD1

# 714650 variants in approximate LE

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001 \
--extract Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_LD1.prune.in \
--genome \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_LD1

# Max pi_hat is 0.91, which is a little odd, but this was fairly quick-and-dirty confirmatory QC to limit down to the dataset for analysis. 
# Limit data to just cell lines with imaging data, taking the best call rate for each individual where multiple cell lines are used
# Use data from Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_Checks.imiss to add to Imaging_Data_Individuals_Genetic_Data and confirm each relationship from Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_LD1.genome

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001 \
--keep Unique_Cell_Result_Lines_To_Keep.txt \
--make-bed \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups

## Assess data quality
mv *MAF001.* Intermediate_Files/
mv *MAF001_Checks* Intermediate_Files/
mv *MAF001_LD1* Intermediate_Files/

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups \
--missing \
--freq \
--hardy \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_Checks
```

### Analysis 1 - Genotype QC

```{sh Filter_To_High_Quality}

## Remove MAF < 0.05
## Small N means low MAF variants may be unreliably called

/scratch/groups/ukbiobank/Edinburgh_Data/Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups \
--maf 0.05 \
--make-bed \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5 

# 69 people
# 4,120,300 variants removed due to minor allele threshold(s)
# 6,811,666 variants and 69 people pass filters and QC.

mv *NoDups.* Intermediate_Files/
mv *NoDups_Checks* Intermediate_Files/

## Assess data quality - same checks as above

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5 \
--missing \
--freq \
--hardy \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_Checks

sort -k5,5gr Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_Checks.lmiss | head

## There are variants with poor call rate - remove these

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5 \
--geno 0.01 \
--make-bed \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_GENO1

#137 variants removed due to missing genotype data (--geno).

mv *_MAF5 Intermediate_Files/

## Remove variants violating HWE

R Assess_HWE_Cutoff.R

# See Assess_HWE_Cutoff.R below - use 0.00001

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_GENO1 \
--hwe 0.00001 \
--make-bed \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_GENO1_HWE5

# 20,512 variants removed due to Hardy-Weinberg exact test.
# 6,791,017 variants and 69 people pass filters and QC.

mv *Checks* Intermediate_Files/
mv *_GENO1.* Intermediate_Files/

```

```{r Assess_HWE_Cutoff.R}
## Load packages
library(data.table)

## Load data
HWE<-fread("Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_Checks.hwe", data.table=F)

## Quick check
head(HWE)
#   CHR         SNP    TEST A1   A2    GENO O(HET) E(HET)         P
# 1   1 rs140337953 ALL(NP)  G    T  0/8/61 0.1159 0.1092 1.000e+00
# 2   1 rs116400033 ALL(NP)  A    T 0/10/59 0.1449 0.1344 1.000e+00
# 3   1   rs2462492 ALL(NP)  T    C 0/10/59 0.1449 0.1344 1.000e+00
# 4   1  rs76735897 ALL(NP)  G    A 0/44/25 0.6377 0.4344 4.402e-05
# 5   1  rs77573425 ALL(NP)  C    G 0/46/23 0.6667 0.4444 1.635e-05
# 6   1 rs201888535 ALL(NP)  C CCTA 0/38/31 0.5507 0.3991 1.531e-03

## Plot data - 10^-5 looks like a sensible, conservative cut-off
pdf("Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_Checks_HWE_Plot.pdf")
    hist(log10(HWE$P))
    hist(log10(HWE$P), ylim=c(0,100000))
    hist(log10(HWE$P), ylim=c(0,10000))
    hist(log10(HWE$P), ylim=c(0,1000))
dev.off()
```

```{sh More_QC}
# Confirm expected duplicates by pruning for LD and running IBS analysis

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_GENO1_HWE5 \
--indep-pairwise 1500 150 0.2 \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_GENO1_HWE5_LD1

# 172841 variants in approximate LE

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_GENO1_HWE5 \
--extract Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_GENO1_HWE5_LD1.prune.in \
--genome \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_GENO1_HWE5_LD2

# Duplicates are present - examining the HipSci website, this is because some apparently distinct lines are actually from the same person...
# Exclude lower call rates
# Donor fpdl is 4 lines:
#    HPSI0813i-fpdl_1 
#    HPSI0913i-ffdl_1 [EXCLUDE --> Duplicate_Lines.txt]
#    HPSI1213i-foqj_2 [EXCLUDE --> Duplicate_Lines.txt]
#    HPSI0913i-ffdz_1 {EXCLUDE --> Duplicate_Lines.txt]
# Donor fpdr is 3 lines:
#    HPSI0813i-ffdr_3 [EXCLUDE --> Duplicate_Lines.txt]
#    HPSI0813i-fpdr_1
#    HPSI1113i-fumb_1 [EXCLUDE --> Duplicate_Lines.txt]
# Donor fpdk is 3 lines:
#    HPSI0813i-fpdk_2
#    HPSI0813i-ffdk_1 [NOT IN STUDY]
#    HPSI1113i-zuta_1 [EXCLUDE --> Duplicate_Lines.txt]
# Donor ffdj is 3 lines
#    HPSI0813i-ffdj_1 [EXCLUDE --> Duplicate_Lines.txt]
#    HPSI0813i-fpdj_1
#    HPSI1113i-nibo_1 [EXCLUDE --> Duplicate_Lines.txt]
# Donor ffdm is 4 lines
#    HPSI0813i-ffdm_1 [EXCLUDE --> Duplicate_Lines.txt]
#    HPSI0813i-fpdm_1
#    HPSI1114i-qibk_1 [NOT IN STUDY]
#    HPSI1213i-toog_1 [NOT IN STUDY]

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDups_MAF5_GENO1_HWE5 \
--maf 0.05 \
--geno 0.01 \
--hwe 0.00001 \
--remove Duplicate_Lines.txt \
--make-bed \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5

# --remove: 60 people remaining.
# 0 variants removed due to missing genotype data (--geno).
# --hwe: 466 variants removed due to Hardy-Weinberg exact test.
# 116485 variants removed due to minor allele threshold(s)

# 6,674,066 variants and 60 people pass filters and QC.

mv *NoDups_* Intermediate_Files/

# Confirm expected duplicates by pruning for LD and running IBS analysis

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5 \
--indep-pairwise 1500 150 0.2 \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_LD1

# 178141 variants in approximate LE

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5 \
--extract Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_LD1.prune.in \
--genome \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_LD2

## Sex check

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5 \
--extract Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_LD1.prune.in \
--check-sex \
--chr 23 \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_SexCheck

## Everyone appears to be female - cross-check with HipSci website...
## This is not correct - examination of the imputed files suggests that the data is wrong - there are hardly any rare heterozygote calls, even on female participants

## Convert and merge genotyped data for open access

cd OpenAccess

for file in *20141111.genotypes.vcf.gz
do 
    filename=$(echo $file | sed 's/\..*//g')
    Software/plink \
    --vcf $file \
    --make-bed \
    --out $filename
done

# Make MergeList.txt, a file with the PLINK binary for each OA individual on each line
ls *.fam > MergeList.txt
# Then remove .fam manually in emacs

# Merge
Software/plink \
--merge-list MergeList.txt \
--make-bed \
--out MergedOA_Data

## Limit to genotyped data and sex check

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5 \
--check-sex \
--extract OpenAccess/MergedOA_Data.bim \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_SexCheck_SNPs_From_Genotypes

## Confirm check sex was correct

## Make samples.txt (to pass to imputation) and samples_FID_IID.txt (to confirm sex checks)

awk '{if($5 ==1) print $2, "M"; else if($5 == 2) print $2, "F"}' /scratch/users/k1204688/iPSC_PRS/HipSci/Genotypes/OpenAccess//hipsci_20151028.ped | \
LANG=C fgrep -wf <(awk '{print $1"_"$2}' \
/scratch/users/k1204688/iPSC_PRS/HipSci/Genotypes/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5.fam) \
> samples.txt

tr  '_' ' ' < samples.txt > samples_FID_IID.txt

# Run ConfirmCheckSex in R below
# Sex check is correct

```

```{r ConfirmCheckSex}

## Load packages
library(data.table)
library(dplyr)

## Load data
SexCheck<-fread("Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_SexCheck_SNPs_From_Genotypes.sexcheck", data.table=F)
Genders<-fread("samples_FID_IID.txt", data.table=F, head=F)
names(Genders)<-c("FID","IID","Gender")

## Check
SexCheck_Genders<-full_join(SexCheck, Genders)
table(SexCheck_Genders$SNPSEX, SexCheck_Genders$Gender, useNA="a")

#        F  M <NA>
#  1     0 26    0
#  2    34  0    0
#  <NA>  0  0    0

# All good

```

Cell lines consistency was established on imputed data now deemed untrustworthy. Check that consistency is still observed limiting to just genotyped variants

```{sh Confirm cell line consistency}

## Use file of all individuals limited to MAF < 0.001 (just the monomorphic variants)

cd iPSC_PRS/HipSci/Genotypes

/scratch/groups/ukbiobank/Edinburgh_Data/Software/plink \
--bfile Intermediate_Files/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001 \
--missing \
--extract OpenAccess/MergedOA_Data.bim \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_SNPs_From_Genotypes_Checks

## Check whether missing rate differs between individuals in R (script below

module add apps/R/3.6.0

R

```

```{r Missing_Rate_Check}
library(data.table)

Miss_Indiv<-fread("Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_SNPs_From_Genotypes_Checks.imiss", data.table=F)

summary(Miss_Indiv$F_MISS)

# Min = 4.164e-06 
# 1Q = 1.249e-05 
# Median = 2.082e-05 
# Mean = 2.046e-05 
# 3Q = 2.498e-05 
# Max = 5.413e-05

## Compare imputed
# Min = 1.56e-6
# 1Q = 4.92e-6
# Median = 6.40e-6
# Mean = 6.32e-6
# 3Q = 7.78e-6
# Max = 1.12e-5

# Differences in call rate are trivial
```

Need to confirm cell lines are ~genetically identical when considering genotyped data only 

```{sh Further checks}

# Confirm expected duplicates by pruning for LD and running IBS analysis

cd iPSC_PRS/HipSci/Genotypes

Software/plink \
--bfile Intermediate_Files/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001 \
--extract OpenAccess/MergedOA_Data.bim \
--indep-pairwise 1500 150 0.2 \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_SNPs_From_Genotypes_LD1

# 714650 variants in approximate LE

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001 \
--extract Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_SNPs_From_Genotypes_LD1.prune.in \
--genome \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_SNPs_From_Genotypes_LD2

# Consistent

```

```{sh Reimpute}

## Limit to genotyped data to reimpute to HRC

Software/plink \
--bfile Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5 \
--extract OpenAccess/MergedOA_Data.bim \
--make-bed \
--freq \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only

## Prep for reimputation
## The following steps were carried out on JRIC's MacBook, because Rosalind had some issues with W.Rayner's checking script

## Transfer .bim and .frq file to local workstation

rsync [REDACTED]/iPSC_PRS/HipSci/Genotypes/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only.bim .
rsync [REDACTED]/iPSC_PRS/HipSci/Genotypes/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only.frq .

## Download W.Rayner's latest checking tool and the HRC sites

wget https://www.well.ox.ac.uk/\~wrayner/tools/HRC-1000G-check-bim-v4.2.13.zip
unzip HRC-1000G-check-bim-v4.2.13.zip
wget ftp://ngs.sanger.ac.uk/production/hrc/HRC.r1-1/HRC.r1-1.GRCh37.wgs.mac5.sites.tab.gz

## Minor modification to perl script to point to correct location of gunzip on local system

emacs HRC-1000G-check-bim.pl

## Run script

perl HRC-1000G-check-bim.pl -b Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only.bim -r HRC.r1-1.GRCh37.wgs.mac5.sites.tab.gz -h -f Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only.frq

## Gather files and transfer to Rosalind

mkdir HRC_Files
mv *HRC.txt HRC_Files
mv Run-plink.sh HRC_Files
rsync -r HRC_Files [REDACTED]/iPSC_PRS/HipSci/Genotypes/PreImputation

## On Rosalind:

cd iPSC_PRS/HipSci/Genotypes/PreImputation

## Modify Run-plink so the chrX is coded as chr23
emacs Run-plink.sh 

## Run
sh Run-plink.sh

## Merge VCFs
Tools/bcftools/bin/bcftools \
concat \
-Oz \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr1.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr2.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr3.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr4.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr5.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr6.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr7.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr8.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr9.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr10.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr11.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr12.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr13.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr14.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr15.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr16.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr17.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr18.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr19.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr20.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr21.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr22.vcf \
Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-chr23.vcf \
> Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-WG.vcf.gz

## Check combined vcf

Tools/bcftools/bin/bcftools index Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-WG.vcf.gz

Tools/bcftools/bin/bcftools +fixref Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_NoDupsSecondPass_MAF5_GENO1_HWE5_Genotyped_SNPs_Only-updated-WG.vcf.gz -- -f human_g1k_v37.fasta

#Both checks suggest alignment is good

## Transfer to SangerImputationServer (EAGLE2 + imputation) via workstation.
## Retrieve from SangerImputationServer via workstation

rsync -r HipSciReimputationTake2.vcfs [REDACTED]/iPSC_PRS/HipSci/Genotypes

cd iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs

## Extract INFO and MAF from vcfs

for i in {22..1} X
do 
    Tools/bcftools/bin/bcftools \
    query -f '%CHROM  %POS  %REF  %ALT{0} %INFO/INFO %INFO/AC\n' $i.vcf.gz > $i.INFO.AC.txt
done

ls *.INFO.AC.txt >> Make_INFO_AC.sh

emacs Make_INFO_AC.sh

sh Make_INFO_AC.sh
```

```{r Plot_Info}
library(data.table)
WG_Info<-fread("WG.INFO.AC.txt", data.table=F)
pdf("INFO.pdf")
hist(WG_Info$V5, 100)
hist(WG_Info$V5, 100, ylim=c(0,1000000))
hist(WG_Info$V5, 100, ylim=c(0,500000))
hist(WG_Info$V5, 100, ylim=c(0,100000))
dev.off()
```

Can drop most of this - remove INFO below 0.9, AC >= 6 OR AC <=114 (MAF 5%)

```{sh Extract_High_Quality_Data}
## Filter to INFO >= 0.9, MAF >= 0.05 (actually MAC >= 6)
awk '$5 >= 0.9 && $6 >= 6 && $6 <= 114 {print}' WG.INFO.AC.txt > WG.INFO.0.9.AC.0.05.txt
awk '{print $1, $2, $2, "SNP"}' WG.INFO.0.9.AC.0.05.txt > WG.INFO.0.9.AC.0.05_For_Extract.txt

Software/plink \
--vcf WG.vcf.gz \
--make-bed \
--extract range WG.INFO.0.9.AC.0.05_For_Extract.txt \
--out WG_INFO9_MAF5

# --extract range: 35480138 variants excluded.
# --extract range: 4925367 variants remaining.

##Basic QC checks

Software/plink \
--bfile WG_INFO9_MAF5 \
--missing \
--freq \
--hardy \
--out WG_INFO9_MAF5_Checks

##This showed that there were still variants with MAF < 0.05 - I assume because there are some multiallelics. Remove these

Software/plink \
--bfile WG_INFO9_MAF5 \
--maf 0.05 \
--make-bed \
--out WG_INFO9_MAF5_PLINKMAF

# 25596 variants removed due to minor allele threshold(s)
# 4899771 variants and 60 people pass filters and QC.

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF \
--missing \
--freq \
--hardy \
--out WG_INFO9_MAF5_PLINKMAF_Checks

## No missingness (hard-called) 
## Don't care too much about HWE at the moment because the original data was not in violation

## Do sex-check

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF \
--indep-pairwise 1500 150 0.2 \
--out WG_INFO9_MAF5_PLINKMAF_LD1

# 279763 variants

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF \
--check-sex \
--extract WG_INFO9_MAF5_PLINKMAF_LD1.prune.in \
--out WG_INFO9_MAF5_PLINKMAF_SexCheck

```

```{r ConfirmCheckSex}

## Load packages
library(data.table)
library(dplyr)

## Load data
SexCheck<-fread("WG_INFO9_MAF5_PLINKMAF_SexCheck.sexcheck", data.table=F)
Genders<-fread("../samples_FID_IID.txt", data.table=F, head=F)
names(Genders)<-c("FID","IID","Gender")

## Check
SexCheck_Genders<-full_join(SexCheck, Genders)
table(SexCheck_Genders$SNPSEX, SexCheck_Genders$Gender, useNA="a")

#        F  M <NA>
#  1     0 26    0
#  2    34  0    0
#  <NA>  0  0    0

# All good

```

```{sh Further QC}

## Add sex info to binary and rerun checks

mv WG_INFO9_MAF5_PLINKMAF.log WG_INFO9_MAF5_PLINKMAF_First_Run.log

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF \
--update-sex ../samples_FID_IID.txt \
--make-bed \
--out WG_INFO9_MAF5_PLINKMAF

## Remove old PLINKMAF files (hashed for saftey!)
#rm WG_INFO9_MAF5_PLINKMAF*~

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF \
--missing \
--freq \
--hardy \
--out WG_INFO9_MAF5_PLINKMAF_Checks

## Check HWE

sort -k9,9g WG_INFO9_MAF5_PLINKMAF_Checks.hwe | head
# Lowest p = 3.494e-08 - acceptable

## Many variants have a "." as their name - update

awk '{if($2 == ".") print $1, $1"_"$4, $3, $4, $5, $6; else print $0}' WG_INFO9_MAF5_PLINKMAF.bim > TEMP
tr ' ' '\t' < TEMP > WG_INFO9_MAF5_PLINKMAF.bim
#rm TEMP

### Confirm European ancestry by projecting on 1KG data

## Call 1KG Phase 3 data from previous creation - NB THIS WAS JUST EUR DATA, WHICH I DIDN'T INITIALLY REALISE...

ln -s Resources/1KG/1KG_Phase3/1KG_Phase3_Populations.txt 1KG_Phase3_Populations.txt
ln -s Resources/1KG/1KG_Phase3/1KG_Phase3.WG.CLEANED.EUR_MAF001.fam 1KG_Phase3.WG.CLEANED.EUR_MAF001.fam
ln -s Resources/1KG/1KG_Phase3/1KG_Phase3.WG.CLEANED.EUR_MAF001.bim 1KG_Phase3.WG.CLEANED.EUR_MAF001.bim
ln -s Resources/1KG/1KG_Phase3/1KG_Phase3.WG.CLEANED.EUR_MAF001.bed 1KG_Phase3.WG.CLEANED.EUR_MAF001.bed

## Get union variant IDs

LANG=C fgrep -wf <(awk '{print $2}' WG_INFO9_MAF5_PLINKMAF.bim) 1KG_Phase3.WG.CLEANED.EUR_MAF001.bim | awk '{print $2}' > Union_HipSci_1KG3_Variants.txt 

wc -l WG_INFO9_MAF5_PLINKMAF.bim 1KG_Phase3.WG.CLEANED.EUR_MAF001.bim Union_HipSci_1KG3_Variants.txt
#  4899771 WG_INFO9_MAF5_PLINKMAF.bim
# 23247316 1KG_Phase3.WG.CLEANED.EUR_MAF001.bim
#  4659650 Union_HipSci_1KG3_Variants.txt

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF \
--extract Union_HipSci_1KG3_Variants.txt \
--make-bed \
--out WG_INFO9_MAF5_PLINKMAF_1KG_SNPs

Software/plink \
--bfile 1KG_Phase3.WG.CLEANED.EUR_MAF001 \
--extract Union_HipSci_1KG3_Variants.txt \
--make-bed \
--out 1KG_Phase3.WG.CLEANED.EUR_MAF001_HipSci_SNPs

## Merge files

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF_1KG_SNPs \
--bmerge 1KG_Phase3.WG.CLEANED.EUR_MAF001_HipSci_SNPs \
--out 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs

# Some variants are inconsistently coded between the datasets - drop these

cp 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs.missnp Inconsistent_HipSci_1KG3_Variants.txt
mv 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs.log 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_First_Run.log

# Add multiposition and multi-chromosome SNPs to removals
emacs Inconsistent_HipSci_1KG3_Variants.txt

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF_1KG_SNPs \
--exclude Inconsistent_HipSci_1KG3_Variants.txt \
--make-bed \
--out WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_Cleaned

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_Cleaned \
--bmerge 1KG_Phase3.WG.CLEANED.EUR_MAF001_HipSci_SNPs \
--out 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs

## Code phenotypes

cat <(awk '{print $1, $0}' 1KG_Phase3_Populations.txt) <(awk '{print $1, $2, "HipSci"}' WG_INFO9_MAF5_PLINKMAF_1KG_SNPs.fam) > 1KG_HipSci_Pop_Pheno.txt

sed -i -e 's/ACB/3/g' \
-e 's/ASW/4/g' \
-e 's/BEB/5/g' \
-e 's/CDX/6/g' \
-e 's/CEU/7/g' \
-e 's/CHB/8/g' \
-e 's/CHS/10/g' \
-e 's/CLM/11/g' \
-e 's/ESN/12/g' \
-e 's/FIN/13/g' \
-e 's/GBR/14/g' \
-e 's/GIH/15/g' \
-e 's/GWD/16/g' \
-e 's/HipSci/17/g' \
-e 's/IBS/18/g' \
-e 's/ITU/19/g' \
-e 's/JPT/20/g' \
-e 's/KHV/21/g' \
-e 's/LWK/22/g' \
-e 's/MSL/23/g' \
-e 's/MXL/24/g' \
-e 's/PEL/25/g' \
-e 's/PJL/26/g' \
-e 's/PUR/27/g' \
-e 's/STU/28/g' \
-e 's/TSI/29/g' \
-e 's/YRI/30/g' 1KG_HipSci_Pop_Pheno.txt 

## LD prune, limiting to well-typed SNPs

Software/plink \
--bfile 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs \
--maf 0.05 \
--geno 0.01 \
--indep-pairwise 1500 150 0.1 \
--autosome \
--out 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_1

# 38811 variants

Software/plink \
--bfile 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs \
--extract 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_1.prune.in \
--make-bed \
--out 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned \
--pheno 1KG_HipSci_Pop_Pheno.txt

### Project on PCs from 1KG

## Convert to EIGENSTRAT format

Software/EIG-7.2.1/bin/convertf -p <(printf "genotypename: 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.bed
snpname: 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.bim
indivname: 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.fam
outputformat: EIGENSTRAT
genotypeoutname: 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.eigenstratgeno
snpoutname: 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.snp
indivoutname: 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.ind")

## Make poplist

emacs 1KG.poplist
# This contains all the numeric pop IDs above except 17 (no HipSci)

## Run smartpca
Software/EIG-7.2.1/bin/smartpca -p <(printf "genotypename: 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.eigenstratgeno
snpname: 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.snp
indivname: 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.ind
evecoutname: 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.pca
evaloutname: 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.eval
numoutlieriter: 0 
numoutevec: 4 
numoutlierevec: 4 
outliersigmathresh: 6 
poplistname: 1KG.poplist")

## Manipulate for R

cat <(echo "FID IID PC1 PC2 PC3 PC4 Pop") <(\
awk 'NR > 1 {print $1, $2, $3, $4, $5, $6"CHANGE"}' 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.pca | \
sed \
-e 's/10CHANGE/CHS/g' \
-e 's/11CHANGE/CLM/g' \
-e 's/12CHANGE/ESN/g' \
-e 's/13CHANGE/FIN/g' \
-e 's/14CHANGE/GBR/g' \
-e 's/15CHANGE/GIH/g' \
-e 's/16CHANGE/GWD/g' \
-e 's/17CHANGE/HipSci/g' \
-e 's/18CHANGE/IBS/g' \
-e 's/19CHANGE/ITU/g' \
-e 's/20CHANGE/JPT/g' \
-e 's/21CHANGE/KHV/g' \
-e 's/22CHANGE/LWK/g' \
-e 's/23CHANGE/MSL/g' \
-e 's/24CHANGE/MXL/g' \
-e 's/25CHANGE/PEL/g' \
-e 's/26CHANGE/PJL/g' \
-e 's/27CHANGE/PUR/g' \
-e 's/28CHANGE/STU/g' \
-e 's/29CHANGE/TSI/g' \
-e 's/30CHANGE/YRI/g' \
-e 's/3CHANGE/ACB/g' \
-e 's/4CHANGE/ASW/g' \
-e 's/5CHANGE/BEB/g' \
-e 's/6CHANGE/CDX/g' \
-e 's/7CHANGE/CEU/g' \
-e 's/8CHANGE/CHB/g') | \
tr ':' ' ' > 1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.pca_For_R.txt

R PlotPCs.R

```

```{r PlotPCs.R}
### Load packages
library(data.table)
library(ggplot2)

### Load data
PC<-fread("1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.pca_For_R.txt", data.table=F)

### Write pdf
pdf("EUR_PCA_Plot_HipSci.pdf")
   qplot(PC$PC1, PC$PC2, colour=PC$Pop)
   qplot(PC$PC1, PC$PC3, colour=PC$Pop)
   qplot(PC$PC1, PC$PC4, colour=PC$Pop)
   qplot(PC$PC2, PC$PC3, colour=PC$Pop)
   qplot(PC$PC2, PC$PC4, colour=PC$Pop)
   qplot(PC$PC3, PC$PC4, colour=PC$Pop)
dev.off()
```

### Ancestry plot against FULL 1KG data

```{sh PlotAncestry}
## Call 1KG Phase 1 data from previous creation

ln -s Resources/1KG/1kg_phase1_all.fam 1kg_phase1_all.fam
ln -s Resources/1KG/1kg_phase1_all.bim 1kg_phase1_all.bim
ln -s Resources/1KG/1kg_phase1_all.bed 1kg_phase1_all.bed

## Get union variant IDs

LANG=C fgrep -wf <(awk '{print $2}' WG_INFO9_MAF5_PLINKMAF.bim) 1kg_phase1_all.bim | awk '{print $2}' > Union_HipSci_1KG1_Variants.txt

wc -l WG_INFO9_MAF5_PLINKMAF.bim 1kg_phase1_all.bim Union_HipSci_1KG1_Variants.txt
#   4899771 WG_INFO9_MAF5_PLINKMAF.bim
#  39728178 1kg_phase1_all.bim
#   4638789 Union_HipSci_1KG1_Variants.txt

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF \
--extract Union_HipSci_1KG1_Variants.txt \
--make-bed \
--out WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs

Software/plink \
--bfile 1kg_phase1_all \
--extract Union_HipSci_1KG1_Variants.txt \
--make-bed \
--out 1kg_phase1_all_HipSci_SNPs

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs \
--bmerge 1kg_phase1_all_HipSci_SNPs \
--out 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs

# Some variants are inconsistently coded between the datasets - drop these

cp 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs.missnp Inconsistent_HipSci_1KG_Phase1_Variants.txt
mv 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs.log 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_FirstRun.log

# Add multiposition and multi-chromosome SNPs to removals
emacs Inconsistent_HipSci_1KG_Phase1_Variants.txt

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs \
--exclude Inconsistent_HipSci_1KG_Phase1_Variants.txt \
--make-bed \
--out WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_Cleaned

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_Cleaned \
--bmerge 1kg_phase1_all_HipSci_SNPs \
--out 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs

## Make phenos from above - Phase1 has a different FID IID set up

awk '{if($3 != 17) print 0, $2, $3; else print $0}' 1KG_HipSci_Pop_Pheno.txt > 1KG_Phase1_HipSci_Pop_Pheno.txt

## LD prune, limiting to well-typed SNPs

Software/plink \
--bfile 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs \
--maf 0.05 \
--geno 0.01 \
--indep-pairwise 1500 150 0.1 \
--autosome \
--out 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD1

# 45728 variants

Software/plink \
--bfile 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs \
--extract 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD1.prune.in \
--make-bed \
--out 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned \
--pheno 1KG_Phase1_HipSci_Pop_Pheno.txt

### Project on PCs from 1KG

## Convert to EIGENSTRAT format

Software/EIG-7.2.1/bin/convertf -p <(printf "genotypename: 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.bed
snpname: 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.bim
indivname: 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.fam
outputformat: EIGENSTRAT
genotypeoutname: 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.eigenstratgeno
snpoutname: 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.snp
indivoutname: 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.ind")

## Make poplist

emacs 1KG.poplist
# This contains all the numeric pop IDs above except 17 (no HipSci)

## Run smartpca
Software/EIG-7.2.1/bin/smartpca -p <(printf "genotypename: 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.eigenstratgeno
snpname: 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.snp
indivname: 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.ind
evecoutname: 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.pca
evaloutname: 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.eval
numoutlieriter: 0
numoutevec: 4
numoutlierevec: 4
outliersigmathresh: 6
poplistname: 1KG.poplist")

## Manipulate for R

cat <(echo "FID IID PC1 PC2 PC3 PC4 Pop") <(\
awk 'NR > 1 {print $1, $2, $3, $4, $5, $6"CHANGE"}' 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.pca | \
sed \
-e 's/10CHANGE/CHS/g' \
-e 's/11CHANGE/CLM/g' \
-e 's/12CHANGE/ESN/g' \
-e 's/13CHANGE/FIN/g' \
-e 's/14CHANGE/GBR/g' \
-e 's/15CHANGE/GIH/g' \
-e 's/16CHANGE/GWD/g' \
-e 's/17CHANGE/HipSci/g' \
-e 's/18CHANGE/IBS/g' \
-e 's/19CHANGE/ITU/g' \
-e 's/20CHANGE/JPT/g' \
-e 's/21CHANGE/KHV/g' \
-e 's/22CHANGE/LWK/g' \
-e 's/23CHANGE/MSL/g' \
-e 's/24CHANGE/MXL/g' \
-e 's/25CHANGE/PEL/g' \
-e 's/26CHANGE/PJL/g' \
-e 's/27CHANGE/PUR/g' \
-e 's/28CHANGE/STU/g' \
-e 's/29CHANGE/TSI/g' \
-e 's/30CHANGE/YRI/g' \
-e 's/3CHANGE/ACB/g' \
-e 's/4CHANGE/ASW/g' \
-e 's/5CHANGE/BEB/g' \
-e 's/6CHANGE/CDX/g' \
-e 's/7CHANGE/CEU/g' \
-e 's/8CHANGE/CHB/g') | \
tr ':' ' ' > 1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.pca_For_R.txt

R PlotPCsWorldwide.R

```

```{r PlotPCsWorldwide.R}
### Load packages
library(data.table)
library(ggplot2)

### Load data
PC<-fread("1KG_Phase1_WG_INFO9_MAF5_PLINKMAF_1KG_Phase1_SNPs_LD_Pruned.pca_For_R.txt", data.table=F)

### Define colour palette
Colour_Palette<-c("#0DAB76","#EEC170","#99D5C9","#6C969D","#E6ED34","#F2A65A","#F58549","#FF0000","#772F1A","#392B58","#139A43","#BDC22B","#DDE500","#96512D","#053B06")

### Write pdf
pdf("World_PCA_Plot_HipSci.pdf")
   qplot(PC$PC1, PC$PC2, colour=PC$Pop) + scale_colour_manual(values=Colour_Palette) + theme_bw()
   qplot(PC$PC1, PC$PC3, colour=PC$Pop) + scale_colour_manual(values=Colour_Palette) + theme_bw()
   qplot(PC$PC1, PC$PC4, colour=PC$Pop) + scale_colour_manual(values=Colour_Palette) + theme_bw()
   qplot(PC$PC2, PC$PC3, colour=PC$Pop) + scale_colour_manual(values=Colour_Palette) + theme_bw()
   qplot(PC$PC2, PC$PC4, colour=PC$Pop) + scale_colour_manual(values=Colour_Palette) + theme_bw()
   qplot(PC$PC3, PC$PC4, colour=PC$Pop) + scale_colour_manual(values=Colour_Palette) + theme_bw()
dev.off()
```

Data is European, clustering best with Northern Europeans

Can now begin analysis

### Generate PCs for analysis

Two alternatives for PC source - can use PCs from data itself (probably preferable) or can use from (EUR subset of) 1KG. 

Generate internal PCs, and compare to PCs from (EUR subset of) 1KG

```{sh Generate PCs}

### Begin with pruned PC data and exclude high-LD and non-autosomal regions from the pruned file

cp Public/gwas_scripts/highLDregions4bim_b37.awk iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/

cd iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/

### Extract LD filtered

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF \
--extract WG_INFO9_MAF5_PLINKMAF_LD1.prune.in \
--make-bed \
--out WG_INFO9_MAF5_PLINKMAF_LD2

### Extract high-LD regions and limit to autosomes
awk -f highLDregions4bim_b37.awk WG_INFO9_MAF5_PLINKMAF_LD2.bim > highLDexcludes
awk '($1 < 1) || ($1 > 22) {print $2}' WG_INFO9_MAF5_PLINKMAF_LD2.bim > autosomeexcludes
cat highLDexcludes autosomeexcludes > highLD_and_autosomal_excludes  

Software/plink \
--bfile WG_INFO9_MAF5_PLINKMAF_LD2 \
--exclude highLD_and_autosomal_excludes \
--make-bed \
--out WG_INFO9_MAF5_PLINKMAF_LD3

### Generate PCs

## Convert to EIGENSTRAT format

Software/EIG-7.2.1/bin/convertf -p <(printf "genotypename: WG_INFO9_MAF5_PLINKMAF_LD3.bed
snpname: WG_INFO9_MAF5_PLINKMAF_LD3.bim
indivname: WG_INFO9_MAF5_PLINKMAF_LD3.fam
outputformat: EIGENSTRAT
genotypeoutname: WG_INFO9_MAF5_PLINKMAF_LD3.eigenstratgeno
snpoutname: WG_INFO9_MAF5_PLINKMAF_LD3.snp
indivoutname: WG_INFO9_MAF5_PLINKMAF_LD3.ind")

## Run smartpca
Software/EIG-7.2.1/bin/smartpca -p <(printf "genotypename: WG_INFO9_MAF5_PLINKMAF_LD3.eigenstratgeno
snpname: WG_INFO9_MAF5_PLINKMAF_LD3.snp
indivname: WG_INFO9_MAF5_PLINKMAF_LD3.ind
evecoutname: WG_INFO9_MAF5_PLINKMAF_LD3.pca
evaloutname: WG_INFO9_MAF5_PLINKMAF_LD3.eval
numoutlieriter: 0
numoutevec: 20
numoutlierevec: 20
outliersigmathresh: 6")

## Manipulate for R

cat <(echo "FID IID PC1 PC2 PC3 PC4 PC5 PC6 PC7 PC8 PC9 PC10 PC11 PC12 PC13 PC14 PC15 PC16 PC17 PC18 PC19 PC20 Pop") <(awk 'NR > 1 {print}' WG_INFO9_MAF5_PLINKMAF_LD3.pca) | \
tr ':' ' ' > WG_INFO9_MAF5_PLINKMAF_LD3.pca_For_R.txt

## Compare with (EUR subset of) 1KG PCs (first three only) in R

R Compare_PCs_In_R.R

```

```{R Compare_PCs_In_R.R}
### Load packages
library(data.table)
library(ggplot2)
library(dplyr)

### Load data
PC<-fread("WG_INFO9_MAF5_PLINKMAF_LD3.pca_For_R.txt", data.table=F)

### Plot pdf
pdf("WG_INFO9_MAF5_PLINKMAF_LD3.pca.pdf")
   qplot(PC$PC1, PC$PC2) + theme_bw()
   qplot(PC$PC1, PC$PC3) + theme_bw()
   qplot(PC$PC1, PC$PC4) + theme_bw()
   qplot(PC$PC2, PC$PC3) + theme_bw()
   qplot(PC$PC2, PC$PC4) + theme_bw()
   qplot(PC$PC3, PC$PC4) + theme_bw()
dev.off()

### Merge with EUR PCs
EUR_PCs<-fread("1KG_WG_INFO9_MAF5_PLINKMAF_1KG_SNPs_LD_Pruned.pca_For_R.txt", data.table=F)
EUR_PCs<-EUR_PCs[,c(1,3:5)]
names(EUR_PCs)[2:4]<-c("PC_EUR1", "PC_EUR2", "PC_EUR3")

dim(PC)
#60 23

PC_EUR_PCs<-left_join(PC, EUR_PCs)

dim(PC_EUR_PCs)
#60 26

cor(PC_EUR_PCs[,c(3:5,24:26)])

##                   PC1           PC2           PC3    PC_EUR1     PC_EUR2     PC_EUR3
## PC1      1.000000e+00 -8.349932e-06 -1.931868e-05 -0.0650535 -0.07009625  -0.0562770
## PC2     -8.349932e-06  1.000000e+00 -6.730005e-06 -0.2547322  0.01900306   0.1911546
## PC3     -1.931868e-05 -6.730005e-06  1.000000e+00 -0.2623223  0.03552236  -0.2636748
## PC_EUR1 -6.505350e-02 -2.547322e-01 -2.623223e-01  1.0000000 -0.14049658  -0.1443549
## PC_EUR2 -7.009625e-02  1.900306e-02  3.552236e-02 -0.1404966  1.00000000   0.2096328
## PC_EUR3 -5.627700e-02  1.911546e-01 -2.636748e-01 -0.1443549  0.20963281   1.0000000

### Plot PCs against each other
pdf("WG_INFO9_MAF5_PLINKMAF_LD3_vs_EUR_PCs.pca.pdf")
   qplot(PC_EUR_PCs$PC1, PC_EUR_PCs$PC_EUR1) + theme_bw()
   qplot(PC_EUR_PCs$PC1, PC_EUR_PCs$PC_EUR2) + theme_bw()
   qplot(PC_EUR_PCs$PC1, PC_EUR_PCs$PC_EUR3) + theme_bw()
   qplot(PC_EUR_PCs$PC2, PC_EUR_PCs$PC_EUR1) + theme_bw()
   qplot(PC_EUR_PCs$PC2, PC_EUR_PCs$PC_EUR2) + theme_bw()
   qplot(PC_EUR_PCs$PC2, PC_EUR_PCs$PC_EUR3) + theme_bw()
   qplot(PC_EUR_PCs$PC3, PC_EUR_PCs$PC_EUR1) + theme_bw()
   qplot(PC_EUR_PCs$PC3, PC_EUR_PCs$PC_EUR2) + theme_bw()
   qplot(PC_EUR_PCs$PC3, PC_EUR_PCs$PC_EUR3) + theme_bw()
dev.off()

### Check plotting of EUR PCs for HipSci data
pdf("WG_INFO9_MAF5_PLINKMAF_LD3_EUR_PCs.pca.pdf")
   qplot(PC_EUR_PCs$PC_EUR1, PC_EUR_PCs$PC_EUR2) + theme_bw()
   qplot(PC_EUR_PCs$PC_EUR1, PC_EUR_PCs$PC_EUR3) + theme_bw()
   qplot(PC_EUR_PCs$PC_EUR2, PC_EUR_PCs$PC_EUR3) + theme_bw()
dev.off()

## A couple of individuals are somewhat outlying, but they are still European, they just cluster closer to the Southern Europeans compared to the Northern-like others

## The PCs from EUR are quite distinct from those from the data - use those from the data going forwards.

```


Run PRS for height, BMI, schizophrenia, and general psychiatric risk.
For each, run the best within-trait threshold [0.05 for SCZ-CLOZUK, 0.001 for BMI and Height], and the maximal threshold (i.e. pT = 1), which capture all signal + noise
For CDX, there is no best threshold reported in the literature - however, the 10K set chosen for the w/o 23andME data goes down to 0.00165, so a threshold of 0.001 would be comparable.

```{sh ObtainSumstats}

mkdir Sumstats
cd Sumstats/

### Height (Yengo et al, 2018)
wget https://portals.broadinstitute.org/collaboration/giant/images/6/63/Meta-analysis_Wood_et_al%2BUKBiobank_2018.txt.gz

### BMI (Yengo et al, 2018)
wget https://portals.broadinstitute.org/collaboration/giant/images/c/c8/Meta-analysis_Locke_et_al%2BUKBiobank_2018_UPDATED.txt.gz 


#### SUPERSEDED - SEE BELOW
### Schizophrenia (PGC)
### No ability to scrape from PGC, so direct download and upload to remote --> ckqny.scz2snpres.gz
##
##
##
## awk '{if(NR == 1) print "CHR SNP A1 A2 BP INFO OR SE P NGT"; else if(NR > 1) print $0}' ckqny.scz2snpres | \
## sed 's/chr//g' | \
## tr ' ' '\t' > SCZ2_Cleaned_Sumstats
##
####

### Schizophrenia (CLOZUK + PGC)
### Larger than PGC, so may more appropriate as the base GWAS

wget https://walters.psycm.cf.ac.uk/clozuk_pgc2.meta.sumstats.txt.gz --no-check-certificate

gunzip clozuk_pgc2.meta.sumstats.txt.gz

cat \
<(awk '{print $3, $1, $5, $6, $4, $7, $8, $9}' clozuk_pgc2.meta.sumstats.txt | head -1) \
<(awk '{print $3, $1, $5, $6, $4, $7, $8, $9}' clozuk_pgc2.meta.sumstats.txt | fgrep rs | sed 's/\:[^ tab]\+//g') \
> clozuk_pgc2.clumped.sumstats

### Cross-disorder (PGC2)

## Involves data from 23andMe, so include both the limited 10K set and the full set without 23andME

## Limited set
## No ability to scrape from PGC, so direct download and upload to remote --> pgc_cdg2_meta_10k_oct2019_v2.txt.daner.txt.gz

## Full set w/o 23andME
## No ability to scrape from PGC, so direct download and upload to remote --> pgc_cdg2_meta_no23andMe_oct2019_v2.txt.daner.txt.gz


### Unzip all
gunzip *

cd ..
 
mkdir Analysis
cd Analysis/
mkdir PRS_Output

```

```{sh PRSice}
##Height
if [ $1 = 1 ]
then
Software/PRSice_v2.3.1.e/PRSice_linux \
-b iPSC_PRS/HipSci/Sumstats/Meta-analysis_Wood_et_al+UKBiobank_2018.txt \
-t iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF \
--type bed \
--no-regress \
--bar-levels 0.001,1 \
--beta \
-o iPSC_PRS/HipSci/Analysis/PRS_Output/Height \
-n 10 \
--fastscore \
--clump-kb 250 \
--clump-p 1 \
--clump-r2 0.1 \
--A1 Tested_Allele \
--A2 Other_Allele \
--ld iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/1KG_Phase3.WG.CLEANED.EUR_MAF001
fi

##BMI
if [ $1 = 2 ]
then
Software/PRSice_v2.3.1.e/PRSice_linux \
-b iPSC_PRS/HipSci/Sumstats/Meta-analysis_Locke_et_al+UKBiobank_2018_UPDATED.txt \
-t iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF \
--type bed \
--no-regress \
--bar-levels 0.001,1 \
--beta \
-o iPSC_PRS/HipSci/Analysis/PRS_Output/BMI \
-n 10 \
--fastscore \
--clump-kb 250 \
--clump-p 1 \
--clump-r2 0.1 \
--A1 Tested_Allele \
--A2 Other_Allele \
--ld iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/1KG_Phase3.WG.CLEANED.EUR_MAF001
fi

#### SUPERSEDED
## SCZ
##  Initial run of SCZ produced a set of variants that are duplicated in the base file - these were removed with extract below
## if [ $1 = 3 ]
## then
## Software/PRSice_v2.3.1.e/PRSice_linux \
## -b iPSC_PRS/HipSci/Sumstats/SCZ2_Cleaned_Sumstats \
## -t iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF \
## --type bed \
## --no-regress \
## --bar-levels 0.1,1 \
## --or \
## -o iPSC_PRS/HipSci/Analysis/PRS_Output/SCZ \
## -n 10 \
## --fastscore \
## --clump-kb 250 \
## --clump-p 1 \
## --clump-r2 0.1 \
## --ld iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/1KG_Phase3.WG.CLEANED.EUR_MAF001 \
## --extract iPSC_PRS/HipSci/Analysis/PRS_Output/SCZ.valid
## fi
####


## SCZ_CLOZUK
## Initial run of SCZ_CLOZUK produced a set of variants that are duplicated in the base file - these were removed with extract below
if [ $1 = 4 ]
then
Software/PRSice_v2.3.1.e/PRSice_linux \
-b iPSC_PRS/HipSci/Sumstats/clozuk_pgc2.clumped.sumstats \
-t iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF \
--type bed \
--no-regress \
--bar-levels 0.05,1 \
--or \
-o iPSC_PRS/HipSci/Analysis/PRS_Output/SCZ_CLOZUK \
-n 10 \
--fastscore \
--clump-kb 250 \
--clump-p 1 \
--clump-r2 0.1 \
--ld iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/1KG_Phase3.WG.CLEANED.EUR_MAF001 \
--extract iPSC_PRS/HipSci/Analysis/PRS_Output/SCZ_CLOZUK.valid
fi

## Cross Disorder Limited
## Initial run of CXD_10K produced a set of variants that are duplicated in the base file - these were removed with extract below
if [ $1 = 5 ]
then
Software/PRSice_v2.3.1.e/PRSice_linux \
-b iPSC_PRS/HipSci/Sumstats/pgc_cdg2_meta_10k_oct2019_v2.txt.daner.txt \
-t iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF \
--type bed \
--no-regress \
--bar-levels 1 \
--beta \
-o iPSC_PRS/HipSci/Analysis/PRS_Output/CXD_10K \
-n 10 \
--fastscore \
--clump-kb 250 \
--clump-p 1 \
--clump-r2 0.1 \
--ld iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/1KG_Phase3.WG.CLEANED.EUR_MAF001 \
--chr CHROM \
--bp POS \
--a1 ALT \
--a2 REF \
--pvalue PVAL \
--snp ID \
--extract iPSC_PRS/HipSci/Analysis/PRS_Output/CXD_10K.valid
fi

## Cross Disorder Full
## Initial run of CDX_FULL produced a set of variants that are duplicated in the base file - these were removed with extract below
if [ $1 = 6 ]
then
Software/PRSice_v2.3.1.e/PRSice_linux \
-b iPSC_PRS/HipSci/Sumstats/pgc_cdg2_meta_no23andMe_oct2019_v2.txt.daner.txt \
-t iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF \
--type bed \
--no-regress \
--bar-levels 0.001,1 \
--beta \
-o iPSC_PRS/HipSci/Analysis/PRS_Output/CDX_FULL \
-n 10 \
--fastscore \
--clump-kb 250 \
--clump-p 1 \
--clump-r2 0.1 \
--ld iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/1KG_Phase3.WG.CLEANED.EUR_MAF001 \
--chr CHROM \
--bp POS \
--a1 ALT \
--a2 REF \
--pvalue PVAL \
--snp ID \
--extract iPSC_PRS/HipSci/Analysis/PRS_Output/CDX_FULL.valid
fi

```

Running PRSice generates a .prsice file - in this case (without regression) these are the outputs:

```{sh DiplayPRSOverview}

cd iPSC_PRS/HipSci/Analysis/PRS_Output

for file in *.prsice
do
    echo $file
    cat $file | column -t
done

### BMI.prsice
### Pheno  Set   Threshold  Num_SNP
### -      Base  0.001      8567
### -      Base  1          72919

### CDX_FULL.prsice
### Pheno    Set    Threshold    Num_SNP
### -    Base    0.001    3298
### -    Base    1    87849

### CXD_10K.prsice
### Pheno  Set   Threshold  Num_SNP
### -      Base  1          3368

### Height.prsice
### Pheno  Set   Threshold  Num_SNP
### -      Base  0.001      14305
### -      Base  1          72976

### SCZ_CLOZUK.prsice
### Pheno  Set   Threshold  Num_SNP
### -      Base  0.05       23420
### -      Base  1          88633

### SCZ.prsice
### Pheno  Set   Threshold  Num_SNP
### -      Base  0.1        29833
### -      Base  1          81244

```

### Explore and define phenotypes

There are 6 basic phenotypes: area, roundness and width-length ratio, for cells and for nuclei.
There are also a number of potential covariates, of which the most obvious are fibronection conc, well position, clump size and EDU values.
We will also need to include PCs, cell line, donor, and should consider sex, cell type of origin, feeder/feeder-free, reprogramming [all on browswer]
Missing variables potentially of interest as age (NA for most participants) and passage [not available cell-wise for most] 

```{sh BasicSetUp}
cd iPSC_PRS/HipSci/Images/Cell_Results
R ExplorePhenotypes.R
```

```{r ExplorePhenotypes.R}

## Set options
options(stringsAsFactors=F)

## Get packages
library(purrr)
library(tibble)
library(dplyr)
library(tidyr)
library(ggpubr)

## Get data
# Read in all cell data files
temp = list.files(pattern="*.tsv")
# Create list of all cell data files
myfiles = lapply(temp, read.delim)

## The columns in the cell level data are:
# name                   The official name for each line as used within HipSci.  This is especially needed to link to other data from HipSci.
# experiment_name        We assigned a number to each experiment as an internal record
# p_col                  number corresponding to position of cell in well column (A-H)  [Error in README - actually go up to 12]
# p_row                  number corresponding to position of cell in well row (1-8) [I think these correspond to A-H]
# field                  number corresponding to position of cell in fields (1-9)
# i_cell                 unique identifier to cell segmented as an object
# i_cell_unselected      unique identifier to population
# i_clump_singles        unique identifier to population
# i_nuc unique           identifier to population
# i_nuc2 unique          identifier to population
# x_centroid             position
# x_min                  position
# x_max                  position
# y_centroid             position
# y_min                  position
# y_max                  position
# cell_area              morphological value for cell area
# edu_median             intensity value for edU
# oct4_median            N/A
# inten_nuc_dapi_median  intensity value for DAPI
# roundness              morphological value for cell roundness
# ratio_w2l              morphological value for ratio width to length
# nucleus_area           morphological value for nucleus area
# nucleus_roundness      morphological value for nucleus roundness
# nucleus_ratio_w2l      morphological value for nucleus ratio w/l
# clump_size             context feature, number of cells in clump
#
# NOTE - this doesn't include fibronection concentration - this can be inferred from the plate-level data
# Merge can be made on experiment/col/row
#
##

## Initial exploration: create single tibble comprising lists of names and six proposed phenos
dataframe_data<- tibble(
name = map(myfiles, "name"),
cell_area = map(myfiles, "cell_area"),
cell_roundness = map(myfiles, "roundness"),
cell_ratio_w2l = map(myfiles, "ratio_w2l"),
nucleus_area = map(myfiles, "nucleus_area"),
nucleus_roundness = map(myfiles, "nucleus_roundness"),
nucleus_ratio_w2l = map(myfiles, "nucleus_ratio_w2l"))

## Convert lists to a data frame of all entries
Final_Tibble<-dataframe_data %>%
select(name, cell_area, cell_roundness, cell_ratio_w2l, nucleus_area, nucleus_roundness, nucleus_ratio_w2l) %>%
unnest()

### Estimate correlation between phenotypes - this doesn't take into account the inter-relatedness between measures

cor(Final_Tibble[,2:7])

#                     cell_area cell_roundness cell_ratio_w2l nucleus_area  nucleus_roundness nucleus_ratio_w2l
# cell_area          1.00000000     -0.4558940    -0.19619456   0.55526730          0.1172359        0.07901481
# cell_roundness    -0.45589396      1.0000000     0.78267094  -0.14969148          0.1621505        0.20112190
# cell_ratio_w2l    -0.19619456      0.7826709     1.00000000  -0.07262074          0.2127562        0.31239051
# nucleus_area       0.55526730     -0.1496915    -0.07262074   1.00000000          0.2208595        0.14272177
# nucleus_roundness  0.11723590      0.1621505     0.21275618   0.22085952          1.0000000        0.78834594
# nucleus_ratio_w2l  0.07901481      0.2011219     0.31239051   0.14272177          0.7883459        1.00000000
#

### Can also estimate average correlation between measures from cells within the same cell line

Cor_By_Name<-data.frame(name=unique(Final_Tibble$name),
cell_area_cell_roundness=NA,
cell_area_cell_ratio_w2l = NA,
cell_area_nucleus_area = NA,
cell_area_nucleus_roundness = NA,
cell_area_nucleus_ratio_w2l = NA,
cell_roundness_cell_ratio_w2l = NA,
cell_roundness_nucleus_area = NA,
cell_roundness_nucleus_roundness = NA,
cell_roundness_nucleus_ratio_w2l = NA,
cell_ratio_w2l_nucleus_area = NA,
cell_ratio_w2l_nucleus_roundness = NA,
cell_ratio_w2l_nucleus_ratio_w2l = NA,
nucleus_area_nucleus_roundness = NA,
nucleus_area_nucleus_ratio_w2l = NA,
nucleus_roundness_nucleus_ratio_w2l = NA)

Cor_By_Name$cell_area_cell_roundness<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_area, x$cell_roundness))[,2]
Cor_By_Name$cell_area_cell_ratio_w2l<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_area, x$cell_ratio_w2l))[,2]
Cor_By_Name$cell_area_nucleus_area<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_area, x$nucleus_area))[,2]
Cor_By_Name$cell_area_nucleus_roundness<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_area, x$nucleus_roundness))[,2]
Cor_By_Name$cell_area_nucleus_ratio_w2l<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_area, x$nucleus_ratio_w2l))[,2]
Cor_By_Name$cell_roundness_<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_area, x$nucleus_ratio_w2l))[,2]
Cor_By_Name$cell_roundness_cell_ratio_w2l<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_roundness, x$cell_ratio_w2l))[,2]
Cor_By_Name$cell_roundness_nucleus_area<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_roundness, x$nucleus_area))[,2]
Cor_By_Name$cell_roundness_nucleus_roundness<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_roundness, x$nucleus_roundness))[,2]
Cor_By_Name$cell_roundness_nucleus_ratio_w2l<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_roundness, x$nucleus_ratio_w2l))[,2]
Cor_By_Name$cell_ratio_w2l_nucleus_<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_roundness, x$nucleus_ratio_w2l))[,2]
Cor_By_Name$cell_ratio_w2l_nucleus_area<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_ratio_w2l, x$nucleus_area))[,2]
Cor_By_Name$cell_ratio_w2l_nucleus_roundness<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_ratio_w2l, x$nucleus_roundness))[,2]
Cor_By_Name$cell_ratio_w2l_nucleus_ratio_w2l<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$cell_ratio_w2l, x$nucleus_ratio_w2l))[,2]
Cor_By_Name$nucleus_area_nucleus_roundness<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$nucleus_area, x$nucleus_roundness))[,2]
Cor_By_Name$nucleus_area_nucleus_ratio_w2l<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$nucleus_area, x$nucleus_ratio_w2l))[,2]
Cor_By_Name$nucleus_roundness_nucleus_ratio_w2l<-plyr::ddply(Final_Tibble, "name", function(x) cor(x$nucleus_roundness, x$nucleus_ratio_w2l))[,2]
 
apply(Cor_By_Name[,2:18], 2, mean)

#           cell_area_cell_roundness            cell_area_cell_ratio_w2l
#                        -0.45534951                         -0.19390880
#             cell_area_nucleus_area         cell_area_nucleus_roundness
#                         0.58755767                          0.11069182
#        cell_area_nucleus_ratio_w2l       cell_roundness_cell_ratio_w2l
#                         0.07719072                          0.79667119
#        cell_roundness_nucleus_area    cell_roundness_nucleus_roundness
#                        -0.19675862                          0.15604669
#   cell_roundness_nucleus_ratio_w2l         cell_ratio_w2l_nucleus_area
#                         0.19704179                         -0.09026553
#   cell_ratio_w2l_nucleus_roundness    cell_ratio_w2l_nucleus_ratio_w2l
#                         0.21373518                          0.31287373
#     nucleus_area_nucleus_roundness      nucleus_area_nucleus_ratio_w2l
#                         0.18782142                          0.10376160
# nucleus_roundness_nucleus_ratio_w2l   cell_roundness_nucleus_ratio_w2l
#                         0.78584249                          0.07719072
#    cell_ratio_w2l_nucleus_ratio_w2l
#                         0.19704179

apply(Cor_By_Name[,2:18], 2, sd)

#            cell_area_cell_roundness            cell_area_cell_ratio_w2l
#                          0.07682608                          0.05683529
#              cell_area_nucleus_area         cell_area_nucleus_roundness
#                          0.09925987                          0.09308097
#         cell_area_nucleus_ratio_w2l       cell_roundness_cell_ratio_w2l
#                          0.07553163                          0.03796229
#         cell_roundness_nucleus_area    cell_roundness_nucleus_roundness
#                          0.09629467                          0.06351184
#    cell_roundness_nucleus_ratio_w2l         cell_ratio_w2l_nucleus_area
#                          0.05789934                          0.07645087
#    cell_ratio_w2l_nucleus_roundness    cell_ratio_w2l_nucleus_ratio_w2l
#                          0.04806860                          0.05444108
#      nucleus_area_nucleus_roundness      nucleus_area_nucleus_ratio_w2l
#                          0.08762630                          0.08002406
# nucleus_roundness_nucleus_ratio_w2l    cell_roundness_nucleus_ratio_w2l
#                          0.01666989                          0.07553163
#    cell_ratio_w2l_nucleus_ratio_w2l
#                          0.05789934


#                                                                               Without appreciating cell lines
#                     cell_area cell_roundness cell_ratio_w2l nucleus_area  nucleus_roundness nucleus_ratio_w2l
# cell_area          1.00000000     -0.4558940    -0.19619456   0.55526730          0.1172359        0.07901481
# cell_roundness    -0.45534951      1.0000000     0.78267094  -0.14969148          0.1621505        0.20112190
# cell_ratio_w2l    -0.19390880	     0.79667119    1.00000000  -0.07262074          0.2127562        0.31239051
# nucleus_area       0.58755767	    -0.19675862   -0.09026553   1.00000000          0.2208595        0.14272177
# nucleus_roundness  0.11069182	     0.15604669	   0.21373518 	0.18782142	    1.0000000        0.78834594
# nucleus_ratio_w2l  0.07719072	     0.19704179	   0.31287373	0.10376160	    0.78584249	     1.00000000
#                    With appreciating cell lines


```

On the basis of this, the relationship between roundness and w-l is sufficiently high that it doesn't seem productive to include roundness as a phenotype
--> 4 phenotypes, area and w:l in the cell and nucleus.
Need to consider if nucleus is a sensible hypothesis - can see more of an argument for cell than for nucleus.
--> 2 phenotypes, area and w:l in the cell only?

## 2020-07-01 
Phenotype decision - although highly-correlated, each cellular phenotype has interesting features. Particularly height has associations with genes involved in cell shape.
Include all three - I would be uncomfortable making a big deal of a finding between 0.025 and 0.016 anyway!  

Random covariates that must be included to avoid pseudoreplication:
  1. Cell line, nested within donor

Covariates in order of theoretical / known importance:
  1. Fibronectin concentration (has a clear impact on cell shape according to Vigilante et al, and by design)
  2. Genomic PCs (how many? Run with 4 and 10 for comparison)
  3. Cells in clump (consider coding - dichotmous 1 vs 2+, or as a continuous value of clump size?)
  4. Cell type 
  5. Feeder 
  6. Reprogramming method 

Covariates not to be included in first pass (but should be considered in sensitivity analyses if possible)
  1. Well position - this was randomised with respect to fibrinogen as part of the design, so shouldn't have a substantial effect (but worth checking)
  2. Sex - no reason to expect size differences
  3. EDU values
  4. Age - not practical, too much missing data
  5. Passage - missing data
  6. CNV status - this is recorded relative to parent cell - this is less useful than CNVs themselves

### Estimate power landscape

A key concern is whether these analyses are going to be adequately powered to detect significant associations.
Examining this requires making certain assumptions.
Here, the aim is to establish something resembling a general case, as well as a specfic case

```{r EstimatePRSPower.R}

### Install packages

#devtools::install_github("DudbridgeLab/avengeme")

### Load packages
library(avengeme)

### Principal command of interest is polygenescore:
#?polygenescore

###  Usage
# polygenescore(nsnp, n, vg1 = 0, cov12 = vg1, pi0 = 0, pupper = c(0, 1),
#   nested = TRUE, weighted = TRUE, binary = c(FALSE, FALSE),
#   prevalence = c(0.1, 0.1), sampling = prevalence, lambdaS = NA,
#  shrinkage = FALSE, logrisk = FALSE, alpha = 0.05, r2gx = 0,
#  corgx = 0, r2xy = 0, adjustedEffects = FALSE, riskthresh = 0.1)

###  Arguments

## Parameters of interest

# nsnp         Number of independent markers in the polygenic score.

# n    	       Vector with two elements, giving the total sizes of the training and target samples. 
#      	       In case/control studies, n is the sum of the number of cases and number of controls. 
#      	       If only one element of n is given, the training and target samples are assumed to be the same size. No default - a value must be given

# vg1  	       Proportion of variance explained by genetic effects in the training sample.

# cov12        Covariance between genetic effect sizes in the two samples. If the effects are fully correlated then cov12<=sqrt(vg1). 
# 	       If the effects are identical then cov12=vg1 (default).

# pi0          Proportion of markers with no effect on the training trait.

# pupper       Vector of p-value thresholds for selecting markers from training sample. 
# 	       First element is the lower bound of the first interval, second element is the upper bound of the first interval, 
#	       third element is the upper bound of the second interval, etc.

# nested       TRUE if the p-value intervals are nested, that is they have the same lower bound, which is the first element of pupper. 
#  	       If false, lower bound of the second interval is the upper bound of the first and so on.

# weighted     TRUE if estimated effect sizes are used as weights in forming the polygenic score. If false, an unweighted score is used, which is the sum of risk alleles carried.

# binary       TRUE if the training trait is binary. 
#  	       By default, the target trait is binary if the training trait is
#	       Otherwise binary should be a vector with two elements for the training and target samples respectively.

# prevalence   For a binary trait, prevalence in the training sample. 
#  	       By default, prevalence is the same in the target sample. 
#	       Otherwise, prevalence should be a vector with two elements for the training and target samples respectively.

# sampling     For a binary trait, case/control sampling fraction in the training sample. 
#  	       By default, sampling equals the prevalence, as in a cohort study. 
#	       If the sampling fraction is different in the target sample, sampling should be a vector with two elements for the training and target samples respectively.

# shrinkage    TRUE if effect sizes are to be shrunk to BLUPs.

# logrisk      TRUE if binary trait arises from log-risk model rather than liability threshold.

# alpha        Significance level for testing association of the polygenic score in the target sample.

## Not included

# lambdaS      Sibling relative recurrence risk in training sample, can be specified instead of vg1.

# r2gx         Proportion of variance in environmental risk score explained by genetic effects in training sample.

# corgx        Genetic correlation between environmental risk score and training trait.

# r2xy         Proportion of variance in training trait explained by environmental risk score.

# adjustedEffects    TRUE if polygenic and envrionmental scores are combined as a weighted sum. If FALSE, the scores are combined as an unweighted sum even if they are correlated.

# riskthresh    Absolute risk threshold for calculating net reclassification index.

### Determine parameters

# nsnp - 88633 for SCZ_CLOZUK, 72919 for BMI, 72976 for Height, 87849 for PSYCH
# n - c(105318,60) for SCZ_CLOZUK, c(795640,60) for BMI, c(709706,60) for Height, c(438997, 60) for PSYCH 
# vg1 - 0.225 for SCZ_CLOZUK, 0.138 for BMI, 0.347 for Height, 0.25 for PSYCH
# cov12 - to be varied - cov12=cor12*sqrt(vg1vg2)
# pi0 - 0.95 for all
# pupper - c(0.05,1) for SCZ_CLOZUK, c(0.001,1) for BMI, Height and PSYCH
# nested - TRUE for all
# weighted - TRUE for all
# binary - c(TRUE,FALSE) for SCZ_CLOZUK and PSYCH, FALSE for BMI and Height
# prevalence - c(0.007, NA) for SCZ_CLOZUK, c(0.3, NA) for PSYCH, not used for BMI and Height
# sampling - c(0.386, NA) for SCZ_CLOZUK, c(0.368, NA) for PSYCH, not used for BMI and Height
# shrinkage - FALSE for all,
# logrisk - FALSE for SCZ_CLOZUK and PSYCH, NA (not used) for BMI and Height
# alpha - compare 0.05 (single instance) and 0.05/8 (instance used in this analysis)

## Principal paramater to vary is cov12
## cov12 can be varied as a function of vg1 (fixed), vg2 (unknown) and cor12 (unknown)

## Estimate assuming vg2 of 0.05-1 in steps of 0.05, and cor12 of 0.05-1 in steps of 0.05.
## Also consider different alpha levels (0.05, 0.00625 as above)

## Initiate data frames

SCZ_POWER<-data.frame(vg2 = rep(seq(0.05,1,0.05), 40),
		      cor12 = rep(seq(0.05,1,0.05), each = 40),
		      alpha = rep(c(rep(0.05, 20),rep(0.00625, 20)),20),
		      power_best = rep(NA, 800),
		      power_1 = rep(NA, 800))

PSYCH_POWER <- SCZ_POWER
BMI_POWER <- SCZ_POWER
HEIGHT_POWER <- SCZ_POWER

## Fill data frames

for(alphathresh in c(0.05,0.00625)){
    for(cor12est in seq(0.05,1,0.05)){
        for(vg2est in seq(0.05,1,0.05)){
            SCZ_POWER[SCZ_POWER$vg2 == vg2est & SCZ_POWER$cor12 == cor12est & SCZ_POWER$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 88633,
	    		      		     	      	       		  	   			    		             n = c(105318,60),
			  		    										       	     vg1 = 0.225,
			 		    										       	     cov12 = cor12est*sqrt(vg2est*0.225),
			  		    										       	     pi0 = 0.95,
															       	     pupper = c(0,0.05,1),
			  		    										       	     nested = TRUE,
															       	     weighted = TRUE,
			  												       	     binary = c(TRUE, FALSE),
			  												       	     prevalence = c(0.007, NA),
			  												       	     sampling = c(0.386, NA),
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            SCZ_POWER[SCZ_POWER$vg2 == vg2est & SCZ_POWER$cor12 == cor12est & SCZ_POWER$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 88633,
                                                                                                                           	  n = c(105318,60),
                                                                                                                            	  vg1 = 0.225,
                                                                                                                            	  cov12 = cor12est*sqrt(vg2est*0.225),
                                                                                                                            	  pi0 = 0.95,
                                                                                                                            	  pupper = c(0,0.05,1),
                                                                                                                            	  nested = TRUE,
                                                                                                                            	  weighted = TRUE,
                                                                                                                            	  binary = c(TRUE, FALSE),
                                                                                                                            	  prevalence = c(0.007, NA),
                                                                                                                            	  sampling = c(0.386, NA),
                                                                                                                            	  shrinkage = FALSE,
                                                                                                                            	  logrisk = FALSE,
                                                                                                                            	  alpha = alphathresh)$power[2]
            PSYCH_POWER[PSYCH_POWER$vg2 == vg2est & PSYCH_POWER$cor12 == cor12est & PSYCH_POWER$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 87849,
	    		      		     	      	       		  	   			    		             n = c(438997, 60),
			  		    										       	     vg1 = 0.25,
																     cov12 = cor12est*sqrt(vg2est*0.25),
			  		    										       	     pi0 = 0.95,
															       	     pupper = c(0,0.001,1),
			  		    										       	     nested = TRUE,
															       	     weighted = TRUE,
			  												       	     binary = c(TRUE, FALSE),
			  												       	     prevalence = c(0.3, NA),
			  												       	     sampling = c(0.368, NA),
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            PSYCH_POWER[PSYCH_POWER$vg2 == vg2est & PSYCH_POWER$cor12 == cor12est & PSYCH_POWER$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 87849,
                                                                                                                           	  n = c(438997,60),
                                                                                                                            	  vg1 = 0.25,
                                                                                                                            	  cov12 = cor12est*sqrt(vg2est*0.25),
                                                                                                                            	  pi0 = 0.95,
                                                                                                                            	  pupper = c(0,0.001,1),
                                                                                                                            	  nested = TRUE,
                                                                                                                            	  weighted = TRUE,
                                                                                                                            	  binary = c(TRUE, FALSE),
                                                                                                                            	  prevalence = c(0.3, NA),
                                                                                                                            	  sampling = c(0.368, NA),
                                                                                                                            	  shrinkage = FALSE,
                                                                                                                            	  logrisk = FALSE,
                                                                                                                            	  alpha = alphathresh)$power[2]
            BMI_POWER[BMI_POWER$vg2 == vg2est & BMI_POWER$cor12 == cor12est & BMI_POWER$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 72919,
	    		      		     	      	       		  	   			    		       	     n = c(795640,60),
			  		    										       	     vg1 = 0.138,
			 		    										       	     cov12 = cor12est*sqrt(vg2est*0.138),
			  		    										       	     pi0 = 0.95,
			  		    										       	     pupper = c(0,0.001,1),
			  		    										       	     nested = TRUE,
			  												       	     weighted = TRUE,
			  												       	     binary = FALSE,
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            BMI_POWER[BMI_POWER$vg2 == vg2est & BMI_POWER$cor12 == cor12est & BMI_POWER$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 72919,
	    		      		     	      	       		  	   			    		    	  n = c(795640,60),
			  		    										    	  vg1 = 0.138,
			 		    										    	  cov12 = cor12est*sqrt(vg2est*0.138),
			  		    										    	  pi0 = 0.95,
			  		    										    	  pupper = c(0,0.001,1),
			  		    										    	  nested = TRUE,
			  												    	  weighted = TRUE,
			  												    	  binary = FALSE,
			  												    	  shrinkage = FALSE,
			  												    	  logrisk = FALSE,
			  												    	  alpha = alphathresh)$power[2]
            HEIGHT_POWER[HEIGHT_POWER$vg2 == vg2est & HEIGHT_POWER$cor12 == cor12est & HEIGHT_POWER$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 72976,
	    		      		     	      	       		  	   			    		  	           n = c(709706,60),
			  		    										  	     	   vg1 = 0.347,
			 		    										  	     	   cov12 = cor12est*sqrt(vg2est*0.347),
			  		    										  	     	   pi0 = 0.95,
			  		    										  	     	   pupper = c(0,0.001,1),
			  		    										  	     	   nested = TRUE,
			  												  	     	   weighted = TRUE,
			  												  	     	   binary = FALSE,
			  												  	     	   shrinkage = FALSE,
			  												  	     	   logrisk = FALSE,
			  												  	     	   alpha = alphathresh)$power[1]
            HEIGHT_POWER[HEIGHT_POWER$vg2 == vg2est & HEIGHT_POWER$cor12 == cor12est & HEIGHT_POWER$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 72976,
	    		      		     	      	       		  	   			    		  	  	n = c(709706,60),
			  		    										   	  	vg1 = 0.347,
			 		    										  	  	cov12 = cor12est*sqrt(vg2est*0.347),
			  		    										  	  	pi0 = 0.95,
			  		    										  	  	pupper = c(0,0.001,1),
			  		    										  	  	nested = TRUE,
			  												  	  	weighted = TRUE,
			  												  	  	binary = FALSE,
			  												  	  	shrinkage = FALSE,
			  												  	  	logrisk = FALSE,
			  												  	  	alpha = alphathresh)$power[2]
		}
	}
}


## Print output

write.table(BMI_POWER, "iPSC_PRS/HipSci/Analysis/BMI_Power_Table.txt", col.names=T, row.names=F, quote=F)
write.table(HEIGHT_POWER, "iPSC_PRS/HipSci/Analysis/Height_Power_Table.txt", col.names=T, row.names=F, quote=F)
write.table(SCZ_POWER, "iPSC_PRS/HipSci/Analysis/SCZ_CLOZUK_Power_Table.txt", col.names=T, row.names=F, quote=F)
write.table(PSYCH_POWER, "iPSC_PRS/HipSci/Analysis/PSYCH_Power_Table.txt", col.names=T, row.names=F, quote=F)

## What if we had 200 donors?

SCZ_POWER_200<-data.frame(vg2 = rep(seq(0.05,1,0.05), 40),
                      cor12 = rep(seq(0.05,1,0.05), each = 40),
                      alpha = rep(c(rep(0.05, 20),rep(0.00625, 20)),20),
                      power_best = rep(NA, 800),
                      power_1 = rep(NA, 800))

PSYCH_POWER_200 <- SCZ_POWER_200
BMI_POWER_200 <- SCZ_POWER_200
HEIGHT_POWER_200 <- SCZ_POWER_200

## Fill dataframes

for(alphathresh in c(0.05,0.00625)){
    for(cor12est in seq(0.05,1,0.05)){
        for(vg2est in seq(0.05,1,0.05)){
            SCZ_POWER_200[SCZ_POWER_200$vg2 == vg2est & SCZ_POWER_200$cor12 == cor12est & SCZ_POWER_200$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 88633,
	    		      		     	      	       		  	   			    		             n = c(105318,200),
			  		    										       	     vg1 = 0.225,
			 		    										       	     cov12 = cor12est*sqrt(vg2est*0.225),
			  		    										       	     pi0 = 0.95,
															       	     pupper = c(0,0.05,1),
			  		    										       	     nested = TRUE,
															       	     weighted = TRUE,
			  												       	     binary = c(TRUE, FALSE),
			  												       	     prevalence = c(0.007, NA),
			  												       	     sampling = c(0.386, NA),
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            SCZ_POWER_200[SCZ_POWER_200$vg2 == vg2est & SCZ_POWER_200$cor12 == cor12est & SCZ_POWER_200$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 88633,
                                                                                                                           	  n = c(105318,200),
                                                                                                                            	  vg1 = 0.225,
                                                                                                                            	  cov12 = cor12est*sqrt(vg2est*0.225),
                                                                                                                            	  pi0 = 0.95,
                                                                                                                            	  pupper = c(0,0.05,1),
                                                                                                                            	  nested = TRUE,
                                                                                                                            	  weighted = TRUE,
                                                                                                                            	  binary = c(TRUE, FALSE),
                                                                                                                            	  prevalence = c(0.007, NA),
                                                                                                                            	  sampling = c(0.386, NA),
                                                                                                                            	  shrinkage = FALSE,
                                                                                                                            	  logrisk = FALSE,
                                                                                                                            	  alpha = alphathresh)$power[2]
            PSYCH_POWER_200[PSYCH_POWER_200$vg2 == vg2est & PSYCH_POWER_200$cor12 == cor12est & PSYCH_POWER_200$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 87849,
	    		      		     	      	       		  	   			    		             n = c(438997, 200),
			  		    										       	     vg1 = 0.25,
																     cov12 = cor12est*sqrt(vg2est*0.25),
			  		    										       	     pi0 = 0.95,
															       	     pupper = c(0,0.001,1),
			  		    										       	     nested = TRUE,
															       	     weighted = TRUE,
			  												       	     binary = c(TRUE, FALSE),
			  												       	     prevalence = c(0.3, NA),
			  												       	     sampling = c(0.368, NA),
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            PSYCH_POWER_200[PSYCH_POWER_200$vg2 == vg2est & PSYCH_POWER_200$cor12 == cor12est & PSYCH_POWER_200$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 87849,
                                                                                                                           	  n = c(438997,200),
                                                                                                                            	  vg1 = 0.25,
                                                                                                                            	  cov12 = cor12est*sqrt(vg2est*0.25),
                                                                                                                            	  pi0 = 0.95,
                                                                                                                            	  pupper = c(0,0.001,1),
                                                                                                                            	  nested = TRUE,
                                                                                                                            	  weighted = TRUE,
                                                                                                                            	  binary = c(TRUE, FALSE),
                                                                                                                            	  prevalence = c(0.3, NA),
                                                                                                                            	  sampling = c(0.368, NA),
                                                                                                                            	  shrinkage = FALSE,
                                                                                                                            	  logrisk = FALSE,
                                                                                                                            	  alpha = alphathresh)$power[2]
            BMI_POWER_200[BMI_POWER_200$vg2 == vg2est & BMI_POWER_200$cor12 == cor12est & BMI_POWER_200$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 72919,
	    		      		     	      	       		  	   			    		       	     n = c(795640,200),
			  		    										       	     vg1 = 0.138,
			 		    										       	     cov12 = cor12est*sqrt(vg2est*0.138),
			  		    										       	     pi0 = 0.95,
			  		    										       	     pupper = c(0,0.001,1),
			  		    										       	     nested = TRUE,
			  												       	     weighted = TRUE,
			  												       	     binary = FALSE,
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            BMI_POWER_200[BMI_POWER_200$vg2 == vg2est & BMI_POWER_200$cor12 == cor12est & BMI_POWER_200$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 72919,
	    		      		     	      	       		  	   			    		    	  n = c(795640,200),
			  		    										    	  vg1 = 0.138,
			 		    										    	  cov12 = cor12est*sqrt(vg2est*0.138),
			  		    										    	  pi0 = 0.95,
			  		    										    	  pupper = c(0,0.001,1),
			  		    										    	  nested = TRUE,
			  												    	  weighted = TRUE,
			  												    	  binary = FALSE,
			  												    	  shrinkage = FALSE,
			  												    	  logrisk = FALSE,
			  												    	  alpha = alphathresh)$power[2]
            HEIGHT_POWER_200[HEIGHT_POWER_200$vg2 == vg2est & HEIGHT_POWER_200$cor12 == cor12est & HEIGHT_POWER_200$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 72976,
	    		      		     	      	       		  	   			    		  	           n = c(709706,200),
			  		    										  	     	   vg1 = 0.347,
			 		    										  	     	   cov12 = cor12est*sqrt(vg2est*0.347),
			  		    										  	     	   pi0 = 0.95,
			  		    										  	     	   pupper = c(0,0.001,1),
			  		    										  	     	   nested = TRUE,
			  												  	     	   weighted = TRUE,
			  												  	     	   binary = FALSE,
			  												  	     	   shrinkage = FALSE,
			  												  	     	   logrisk = FALSE,
			  												  	     	   alpha = alphathresh)$power[1]
            HEIGHT_POWER_200[HEIGHT_POWER_200$vg2 == vg2est & HEIGHT_POWER_200$cor12 == cor12est & HEIGHT_POWER_200$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 72976,
	    		      		     	      	       		  	   			    		  	  	n = c(709706,200),
			  		    										   	  	vg1 = 0.347,
			 		    										  	  	cov12 = cor12est*sqrt(vg2est*0.347),
			  		    										  	  	pi0 = 0.95,
			  		    										  	  	pupper = c(0,0.001,1),
			  		    										  	  	nested = TRUE,
			  												  	  	weighted = TRUE,
			  												  	  	binary = FALSE,
			  												  	  	shrinkage = FALSE,
			  												  	  	logrisk = FALSE,
			  												  	  	alpha = alphathresh)$power[2]
		}
	}
}


## Print output

write.table(BMI_POWER_200, "iPSC_PRS/HipSci/Analysis/BMI_Power_Table_200.txt", col.names=T, row.names=F, quote=F)
write.table(HEIGHT_POWER_200, "iPSC_PRS/HipSci/Analysis/Height_Power_Table_200.txt", col.names=T, row.names=F, quote=F)
write.table(SCZ_POWER_200, "iPSC_PRS/HipSci/Analysis/SCZ_CLOZUK_Power_Table_200.txt", col.names=T, row.names=F, quote=F)
write.table(PSYCH_POWER_200, "iPSC_PRS/HipSci/Analysis/PSYCH_Power_Table_200.txt", col.names=T, row.names=F, quote=F)

## What if we had 1000 donors?

SCZ_POWER_1000<-data.frame(vg2 = rep(seq(0.05,1,0.05), 40),
                      cor12 = rep(seq(0.05,1,0.05), each = 40),
                      alpha = rep(c(rep(0.05, 20),rep(0.00625, 20)),20),
                      power_best = rep(NA, 800),
                      power_1 = rep(NA, 800))

PSYCH_POWER_1000 <- SCZ_POWER_1000
BMI_POWER_1000 <- SCZ_POWER_1000
HEIGHT_POWER_1000 <- SCZ_POWER_1000


## Fill data frames

for(alphathresh in c(0.05,0.00625)){
    for(cor12est in seq(0.05,1,0.05)){
        for(vg2est in seq(0.05,1,0.05)){
            SCZ_POWER_1000[SCZ_POWER_1000$vg2 == vg2est & SCZ_POWER_1000$cor12 == cor12est & SCZ_POWER_1000$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 88633,
	    		      		     	      	       		  	   			    		             n = c(105318,1000),
			  		    										       	     vg1 = 0.225,
			 		    										       	     cov12 = cor12est*sqrt(vg2est*0.225),
			  		    										       	     pi0 = 0.95,
															       	     pupper = c(0,0.05,1),
			  		    										       	     nested = TRUE,
															       	     weighted = TRUE,
			  												       	     binary = c(TRUE, FALSE),
			  												       	     prevalence = c(0.007, NA),
			  												       	     sampling = c(0.386, NA),
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            SCZ_POWER_1000[SCZ_POWER_1000$vg2 == vg2est & SCZ_POWER_1000$cor12 == cor12est & SCZ_POWER_1000$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 88633,
                                                                                                                           	  n = c(105318,1000),
                                                                                                                            	  vg1 = 0.225,
                                                                                                                            	  cov12 = cor12est*sqrt(vg2est*0.225),
                                                                                                                            	  pi0 = 0.95,
                                                                                                                            	  pupper = c(0,0.05,1),
                                                                                                                            	  nested = TRUE,
                                                                                                                            	  weighted = TRUE,
                                                                                                                            	  binary = c(TRUE, FALSE),
                                                                                                                            	  prevalence = c(0.007, NA),
                                                                                                                            	  sampling = c(0.386, NA),
                                                                                                                            	  shrinkage = FALSE,
                                                                                                                            	  logrisk = FALSE,
                                                                                                                            	  alpha = alphathresh)$power[2]
            PSYCH_POWER_1000[PSYCH_POWER_1000$vg2 == vg2est & PSYCH_POWER_1000$cor12 == cor12est & PSYCH_POWER_1000$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 87849,
	    		      		     	      	       		  	   			    		             n = c(438997, 1000),
			  		    										       	     vg1 = 0.25,
																     cov12 = cor12est*sqrt(vg2est*0.25),
			  		    										       	     pi0 = 0.95,
															       	     pupper = c(0,0.001,1),
			  		    										       	     nested = TRUE,
															       	     weighted = TRUE,
			  												       	     binary = c(TRUE, FALSE),
			  												       	     prevalence = c(0.3, NA),
			  												       	     sampling = c(0.368, NA),
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            PSYCH_POWER_1000[PSYCH_POWER_1000$vg2 == vg2est & PSYCH_POWER_1000$cor12 == cor12est & PSYCH_POWER_1000$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 87849,
                                                                                                                           	  n = c(438997,1000),
                                                                                                                            	  vg1 = 0.25,
                                                                                                                            	  cov12 = cor12est*sqrt(vg2est*0.25),
                                                                                                                            	  pi0 = 0.95,
                                                                                                                            	  pupper = c(0,0.001,1),
                                                                                                                            	  nested = TRUE,
                                                                                                                            	  weighted = TRUE,
                                                                                                                            	  binary = c(TRUE, FALSE),
                                                                                                                            	  prevalence = c(0.3, NA),
                                                                                                                            	  sampling = c(0.368, NA),
                                                                                                                            	  shrinkage = FALSE,
                                                                                                                            	  logrisk = FALSE,
                                                                                                                            	  alpha = alphathresh)$power[2]
            BMI_POWER_1000[BMI_POWER_1000$vg2 == vg2est & BMI_POWER_1000$cor12 == cor12est & BMI_POWER_1000$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 72919,
	    		      		     	      	       		  	   			    		       	     n = c(795640,1000),
			  		    										       	     vg1 = 0.138,
			 		    										       	     cov12 = cor12est*sqrt(vg2est*0.138),
			  		    										       	     pi0 = 0.95,
			  		    										       	     pupper = c(0,0.001,1),
			  		    										       	     nested = TRUE,
			  												       	     weighted = TRUE,
			  												       	     binary = FALSE,
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            BMI_POWER_1000[BMI_POWER_1000$vg2 == vg2est & BMI_POWER_1000$cor12 == cor12est & BMI_POWER_1000$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 72919,
	    		      		     	      	       		  	   			    		    	  n = c(795640,1000),
			  		    										    	  vg1 = 0.138,
			 		    										    	  cov12 = cor12est*sqrt(vg2est*0.138),
			  		    										    	  pi0 = 0.95,
			  		    										    	  pupper = c(0,0.001,1),
			  		    										    	  nested = TRUE,
			  												    	  weighted = TRUE,
			  												    	  binary = FALSE,
			  												    	  shrinkage = FALSE,
			  												    	  logrisk = FALSE,
			  												    	  alpha = alphathresh)$power[2]
            HEIGHT_POWER_1000[HEIGHT_POWER_1000$vg2 == vg2est & HEIGHT_POWER_1000$cor12 == cor12est & HEIGHT_POWER_1000$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 72976,
	    		      		     	      	       		  	   			    		  	           n = c(709706,1000),
			  		    										  	     	   vg1 = 0.347,
			 		    										  	     	   cov12 = cor12est*sqrt(vg2est*0.347),
			  		    										  	     	   pi0 = 0.95,
			  		    										  	     	   pupper = c(0,0.001,1),
			  		    										  	     	   nested = TRUE,
			  												  	     	   weighted = TRUE,
			  												  	     	   binary = FALSE,
			  												  	     	   shrinkage = FALSE,
			  												  	     	   logrisk = FALSE,
			  												  	     	   alpha = alphathresh)$power[1]
            HEIGHT_POWER_1000[HEIGHT_POWER_1000$vg2 == vg2est & HEIGHT_POWER_1000$cor12 == cor12est & HEIGHT_POWER_1000$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 72976,
	    		      		     	      	       		  	   			    		  	  	n = c(709706,1000),
			  		    										   	  	vg1 = 0.347,
			 		    										  	  	cov12 = cor12est*sqrt(vg2est*0.347),
			  		    										  	  	pi0 = 0.95,
			  		    										  	  	pupper = c(0,0.001,1),
			  		    										  	  	nested = TRUE,
			  												  	  	weighted = TRUE,
			  												  	  	binary = FALSE,
			  												  	  	shrinkage = FALSE,
			  												  	  	logrisk = FALSE,
			  												  	  	alpha = alphathresh)$power[2]
		}
	}
}


## Print output

write.table(BMI_POWER_1000, "iPSC_PRS/HipSci/Analysis/BMI_Power_Table_1000.txt", col.names=T, row.names=F, quote=F)
write.table(HEIGHT_POWER_1000, "iPSC_PRS/HipSci/Analysis/Height_Power_Table_1000.txt", col.names=T, row.names=F, quote=F)
write.table(SCZ_POWER_1000, "iPSC_PRS/HipSci/Analysis/SCZ_CLOZUK_Power_Table_1000.txt", col.names=T, row.names=F, quote=F)
write.table(PSYCH_POWER_1000, "iPSC_PRS/HipSci/Analysis/PSYCH_Power_Table_1000.txt", col.names=T, row.names=F, quote=F)

## Donors are non-independent - estimates suggest 850 < NEff < 2435 in this instance. Repeat calculations for these values

SCZ_POWER_850<-data.frame(vg2 = rep(seq(0.05,1,0.05), 40),
                      cor12 = rep(seq(0.05,1,0.05), each = 40),
                      alpha = rep(c(rep(0.05, 20),rep(0.00625, 20)),20),
                      power_best = rep(NA, 800),
                      power_1 = rep(NA, 800))

PSYCH_POWER_850 <- SCZ_POWER_850
BMI_POWER_850 <- SCZ_POWER_850
HEIGHT_POWER_850 <- SCZ_POWER_850


## Fill data frames

for(alphathresh in c(0.05,0.00625)){
    for(cor12est in seq(0.05,1,0.05)){
        for(vg2est in seq(0.05,1,0.05)){
            SCZ_POWER_850[SCZ_POWER_850$vg2 == vg2est & SCZ_POWER_850$cor12 == cor12est & SCZ_POWER_850$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 88633,
	    		      		     	      	       		  	   			    		             n = c(105318,850),
			  		    										       	     vg1 = 0.225,
			 		    										       	     cov12 = cor12est*sqrt(vg2est*0.225),
			  		    										       	     pi0 = 0.95,
															       	     pupper = c(0,0.05,1),
			  		    										       	     nested = TRUE,
															       	     weighted = TRUE,
			  												       	     binary = c(TRUE, FALSE),
			  												       	     prevalence = c(0.007, NA),
			  												       	     sampling = c(0.386, NA),
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            SCZ_POWER_850[SCZ_POWER_850$vg2 == vg2est & SCZ_POWER_850$cor12 == cor12est & SCZ_POWER_850$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 88633,
                                                                                                                           	  n = c(105318,850),
                                                                                                                            	  vg1 = 0.225,
                                                                                                                            	  cov12 = cor12est*sqrt(vg2est*0.225),
                                                                                                                            	  pi0 = 0.95,
                                                                                                                            	  pupper = c(0,0.05,1),
                                                                                                                            	  nested = TRUE,
                                                                                                                            	  weighted = TRUE,
                                                                                                                            	  binary = c(TRUE, FALSE),
                                                                                                                            	  prevalence = c(0.007, NA),
                                                                                                                            	  sampling = c(0.386, NA),
                                                                                                                            	  shrinkage = FALSE,
                                                                                                                            	  logrisk = FALSE,
                                                                                                                            	  alpha = alphathresh)$power[2]
            PSYCH_POWER_850[PSYCH_POWER_850$vg2 == vg2est & PSYCH_POWER_850$cor12 == cor12est & PSYCH_POWER_850$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 87849,
	    		      		     	      	       		  	   			    		             n = c(438997, 850),
			  		    										       	     vg1 = 0.25,
																     cov12 = cor12est*sqrt(vg2est*0.25),
			  		    										       	     pi0 = 0.95,
															       	     pupper = c(0,0.001,1),
			  		    										       	     nested = TRUE,
															       	     weighted = TRUE,
			  												       	     binary = c(TRUE, FALSE),
			  												       	     prevalence = c(0.3, NA),
			  												       	     sampling = c(0.368, NA),
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            PSYCH_POWER_850[PSYCH_POWER_850$vg2 == vg2est & PSYCH_POWER_850$cor12 == cor12est & PSYCH_POWER_850$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 87849,
                                                                                                                           	  n = c(438997,850),
                                                                                                                            	  vg1 = 0.25,
                                                                                                                            	  cov12 = cor12est*sqrt(vg2est*0.25),
                                                                                                                            	  pi0 = 0.95,
                                                                                                                            	  pupper = c(0,0.001,1),
                                                                                                                            	  nested = TRUE,
                                                                                                                            	  weighted = TRUE,
                                                                                                                            	  binary = c(TRUE, FALSE),
                                                                                                                            	  prevalence = c(0.3, NA),
                                                                                                                            	  sampling = c(0.368, NA),
                                                                                                                            	  shrinkage = FALSE,
                                                                                                                            	  logrisk = FALSE,
                                                                                                                            	  alpha = alphathresh)$power[2]
            BMI_POWER_850[BMI_POWER_850$vg2 == vg2est & BMI_POWER_850$cor12 == cor12est & BMI_POWER_850$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 72919,
	    		      		     	      	       		  	   			    		       	     n = c(795640,850),
			  		    										       	     vg1 = 0.138,
			 		    										       	     cov12 = cor12est*sqrt(vg2est*0.138),
			  		    										       	     pi0 = 0.95,
			  		    										       	     pupper = c(0,0.001,1),
			  		    										       	     nested = TRUE,
			  												       	     weighted = TRUE,
			  												       	     binary = FALSE,
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            BMI_POWER_850[BMI_POWER_850$vg2 == vg2est & BMI_POWER_850$cor12 == cor12est & BMI_POWER_850$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 72919,
	    		      		     	      	       		  	   			    		    	  n = c(795640,850),
			  		    										    	  vg1 = 0.138,
			 		    										    	  cov12 = cor12est*sqrt(vg2est*0.138),
			  		    										    	  pi0 = 0.95,
			  		    										    	  pupper = c(0,0.001,1),
			  		    										    	  nested = TRUE,
			  												    	  weighted = TRUE,
			  												    	  binary = FALSE,
			  												    	  shrinkage = FALSE,
			  												    	  logrisk = FALSE,
			  												    	  alpha = alphathresh)$power[2]
            HEIGHT_POWER_850[HEIGHT_POWER_850$vg2 == vg2est & HEIGHT_POWER_850$cor12 == cor12est & HEIGHT_POWER_850$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 72976,
	    		      		     	      	       		  	   			    		  	           n = c(709706,850),
			  		    										  	     	   vg1 = 0.347,
			 		    										  	     	   cov12 = cor12est*sqrt(vg2est*0.347),
			  		    										  	     	   pi0 = 0.95,
			  		    										  	     	   pupper = c(0,0.001,1),
			  		    										  	     	   nested = TRUE,
			  												  	     	   weighted = TRUE,
			  												  	     	   binary = FALSE,
			  												  	     	   shrinkage = FALSE,
			  												  	     	   logrisk = FALSE,
			  												  	     	   alpha = alphathresh)$power[1]
            HEIGHT_POWER_850[HEIGHT_POWER_850$vg2 == vg2est & HEIGHT_POWER_850$cor12 == cor12est & HEIGHT_POWER_850$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 72976,
	    		      		     	      	       		  	   			    		  	  	n = c(709706,850),
			  		    										   	  	vg1 = 0.347,
			 		    										  	  	cov12 = cor12est*sqrt(vg2est*0.347),
			  		    										  	  	pi0 = 0.95,
			  		    										  	  	pupper = c(0,0.001,1),
			  		    										  	  	nested = TRUE,
			  												  	  	weighted = TRUE,
			  												  	  	binary = FALSE,
			  												  	  	shrinkage = FALSE,
			  												  	  	logrisk = FALSE,
			  												  	  	alpha = alphathresh)$power[2]
		}
	}
}


## Print output

write.table(BMI_POWER_850, "iPSC_PRS/HipSci/Analysis/BMI_Power_Table_850.txt", col.names=T, row.names=F, quote=F)
write.table(HEIGHT_POWER_850, "iPSC_PRS/HipSci/Analysis/Height_Power_Table_850.txt", col.names=T, row.names=F, quote=F)
write.table(SCZ_POWER_850, "iPSC_PRS/HipSci/Analysis/SCZ_CLOZUK_Power_Table_850.txt", col.names=T, row.names=F, quote=F)
write.table(PSYCH_POWER_850, "iPSC_PRS/HipSci/Analysis/PSYCH_Power_Table_850.txt", col.names=T, row.names=F, quote=F)

# 2435

SCZ_POWER_2435<-data.frame(vg2 = rep(seq(0.05,1,0.05), 40),
                      cor12 = rep(seq(0.05,1,0.05), each = 40),
                      alpha = rep(c(rep(0.05, 20),rep(0.00625, 20)),20),
                      power_best = rep(NA, 800),
                      power_1 = rep(NA, 800))

PSYCH_POWER_2435 <- SCZ_POWER_2435
BMI_POWER_2435 <- SCZ_POWER_2435
HEIGHT_POWER_2435 <- SCZ_POWER_2435


## Fill data frames

for(alphathresh in c(0.05,0.00625)){
    for(cor12est in seq(0.05,1,0.05)){
        for(vg2est in seq(0.05,1,0.05)){
            SCZ_POWER_2435[SCZ_POWER_2435$vg2 == vg2est & SCZ_POWER_2435$cor12 == cor12est & SCZ_POWER_2435$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 88633,
	    		      		     	      	       		  	   			    		             n = c(105318,2435),
			  		    										       	     vg1 = 0.225,
			 		    										       	     cov12 = cor12est*sqrt(vg2est*0.225),
			  		    										       	     pi0 = 0.95,
															       	     pupper = c(0,0.05,1),
			  		    										       	     nested = TRUE,
															       	     weighted = TRUE,
			  												       	     binary = c(TRUE, FALSE),
			  												       	     prevalence = c(0.007, NA),
			  												       	     sampling = c(0.386, NA),
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            SCZ_POWER_2435[SCZ_POWER_2435$vg2 == vg2est & SCZ_POWER_2435$cor12 == cor12est & SCZ_POWER_2435$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 88633,
                                                                                                                           	  n = c(105318,2435),
                                                                                                                            	  vg1 = 0.225,
                                                                                                                            	  cov12 = cor12est*sqrt(vg2est*0.225),
                                                                                                                            	  pi0 = 0.95,
                                                                                                                            	  pupper = c(0,0.05,1),
                                                                                                                            	  nested = TRUE,
                                                                                                                            	  weighted = TRUE,
                                                                                                                            	  binary = c(TRUE, FALSE),
                                                                                                                            	  prevalence = c(0.007, NA),
                                                                                                                            	  sampling = c(0.386, NA),
                                                                                                                            	  shrinkage = FALSE,
                                                                                                                            	  logrisk = FALSE,
                                                                                                                            	  alpha = alphathresh)$power[2]
            PSYCH_POWER_2435[PSYCH_POWER_2435$vg2 == vg2est & PSYCH_POWER_2435$cor12 == cor12est & PSYCH_POWER_2435$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 87849,
	    		      		     	      	       		  	   			    		             n = c(438997, 2435),
			  		    										       	     vg1 = 0.25,
																     cov12 = cor12est*sqrt(vg2est*0.25),
			  		    										       	     pi0 = 0.95,
															       	     pupper = c(0,0.001,1),
			  		    										       	     nested = TRUE,
															       	     weighted = TRUE,
			  												       	     binary = c(TRUE, FALSE),
			  												       	     prevalence = c(0.3, NA),
			  												       	     sampling = c(0.368, NA),
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            PSYCH_POWER_2435[PSYCH_POWER_2435$vg2 == vg2est & PSYCH_POWER_2435$cor12 == cor12est & PSYCH_POWER_2435$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 87849,
                                                                                                                           	  n = c(438997,2435),
                                                                                                                            	  vg1 = 0.25,
                                                                                                                            	  cov12 = cor12est*sqrt(vg2est*0.25),
                                                                                                                            	  pi0 = 0.95,
                                                                                                                            	  pupper = c(0,0.001,1),
                                                                                                                            	  nested = TRUE,
                                                                                                                            	  weighted = TRUE,
                                                                                                                            	  binary = c(TRUE, FALSE),
                                                                                                                            	  prevalence = c(0.3, NA),
                                                                                                                            	  sampling = c(0.368, NA),
                                                                                                                            	  shrinkage = FALSE,
                                                                                                                            	  logrisk = FALSE,
                                                                                                                            	  alpha = alphathresh)$power[2]
            BMI_POWER_2435[BMI_POWER_2435$vg2 == vg2est & BMI_POWER_2435$cor12 == cor12est & BMI_POWER_2435$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 72919,
	    		      		     	      	       		  	   			    		       	     n = c(795640,2435),
			  		    										       	     vg1 = 0.138,
			 		    										       	     cov12 = cor12est*sqrt(vg2est*0.138),
			  		    										       	     pi0 = 0.95,
			  		    										       	     pupper = c(0,0.001,1),
			  		    										       	     nested = TRUE,
			  												       	     weighted = TRUE,
			  												       	     binary = FALSE,
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            BMI_POWER_2435[BMI_POWER_2435$vg2 == vg2est & BMI_POWER_2435$cor12 == cor12est & BMI_POWER_2435$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 72919,
	    		      		     	      	       		  	   			    		    	  n = c(795640,2435),
			  		    										    	  vg1 = 0.138,
			 		    										    	  cov12 = cor12est*sqrt(vg2est*0.138),
			  		    										    	  pi0 = 0.95,
			  		    										    	  pupper = c(0,0.001,1),
			  		    										    	  nested = TRUE,
			  												    	  weighted = TRUE,
			  												    	  binary = FALSE,
			  												    	  shrinkage = FALSE,
			  												    	  logrisk = FALSE,
			  												    	  alpha = alphathresh)$power[2]
            HEIGHT_POWER_2435[HEIGHT_POWER_2435$vg2 == vg2est & HEIGHT_POWER_2435$cor12 == cor12est & HEIGHT_POWER_2435$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 72976,
	    		      		     	      	       		  	   			    		  	           n = c(709706,2435),
			  		    										  	     	   vg1 = 0.347,
			 		    										  	     	   cov12 = cor12est*sqrt(vg2est*0.347),
			  		    										  	     	   pi0 = 0.95,
			  		    										  	     	   pupper = c(0,0.001,1),
			  		    										  	     	   nested = TRUE,
			  												  	     	   weighted = TRUE,
			  												  	     	   binary = FALSE,
			  												  	     	   shrinkage = FALSE,
			  												  	     	   logrisk = FALSE,
			  												  	     	   alpha = alphathresh)$power[1]
            HEIGHT_POWER_2435[HEIGHT_POWER_2435$vg2 == vg2est & HEIGHT_POWER_2435$cor12 == cor12est & HEIGHT_POWER_2435$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 72976,
	    		      		     	      	       		  	   			    		  	  	n = c(709706,2435),
			  		    										   	  	vg1 = 0.347,
			 		    										  	  	cov12 = cor12est*sqrt(vg2est*0.347),
			  		    										  	  	pi0 = 0.95,
			  		    										  	  	pupper = c(0,0.001,1),
			  		    										  	  	nested = TRUE,
			  												  	  	weighted = TRUE,
			  												  	  	binary = FALSE,
			  												  	  	shrinkage = FALSE,
			  												  	  	logrisk = FALSE,
			  												  	  	alpha = alphathresh)$power[2]
		}
	}
}


## Print output

write.table(BMI_POWER_2435, "iPSC_PRS/HipSci/Analysis/BMI_Power_Table_2435.txt", col.names=T, row.names=F, quote=F)
write.table(HEIGHT_POWER_2435, "iPSC_PRS/HipSci/Analysis/Height_Power_Table_2435.txt", col.names=T, row.names=F, quote=F)
write.table(SCZ_POWER_2435, "iPSC_PRS/HipSci/Analysis/SCZ_CLOZUK_Power_Table_2435.txt", col.names=T, row.names=F, quote=F)
write.table(PSYCH_POWER_2435, "iPSC_PRS/HipSci/Analysis/PSYCH_Power_Table_2435.txt", col.names=T, row.names=F, quote=F)

## What does it take to go even further? What does N = 5000 look like?


SCZ_POWER_5000<-data.frame(vg2 = rep(seq(0.05,1,0.05), 40),
                      cor12 = rep(seq(0.05,1,0.05), each = 40),
                      alpha = rep(c(rep(0.05, 20),rep(0.00625, 20)),20),
                      power_best = rep(NA, 800),
                      power_1 = rep(NA, 800))

PSYCH_POWER_5000 <- SCZ_POWER_5000
BMI_POWER_5000 <- SCZ_POWER_5000
HEIGHT_POWER_5000 <- SCZ_POWER_5000


## Fill data frames

for(alphathresh in c(0.05,0.00625)){
    for(cor12est in seq(0.05,1,0.05)){
        for(vg2est in seq(0.05,1,0.05)){
            SCZ_POWER_5000[SCZ_POWER_5000$vg2 == vg2est & SCZ_POWER_5000$cor12 == cor12est & SCZ_POWER_5000$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 88633,
	    		      		     	      	       		  	   			    		             n = c(105318,5000),
			  		    										       	     vg1 = 0.225,
			 		    										       	     cov12 = cor12est*sqrt(vg2est*0.225),
			  		    										       	     pi0 = 0.95,
															       	     pupper = c(0,0.05,1),
			  		    										       	     nested = TRUE,
															       	     weighted = TRUE,
			  												       	     binary = c(TRUE, FALSE),
			  												       	     prevalence = c(0.007, NA),
			  												       	     sampling = c(0.386, NA),
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            SCZ_POWER_5000[SCZ_POWER_5000$vg2 == vg2est & SCZ_POWER_5000$cor12 == cor12est & SCZ_POWER_5000$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 88633,
                                                                                                                           	  n = c(105318,5000),
                                                                                                                            	  vg1 = 0.225,
                                                                                                                            	  cov12 = cor12est*sqrt(vg2est*0.225),
                                                                                                                            	  pi0 = 0.95,
                                                                                                                            	  pupper = c(0,0.05,1),
                                                                                                                            	  nested = TRUE,
                                                                                                                            	  weighted = TRUE,
                                                                                                                            	  binary = c(TRUE, FALSE),
                                                                                                                            	  prevalence = c(0.007, NA),
                                                                                                                            	  sampling = c(0.386, NA),
                                                                                                                            	  shrinkage = FALSE,
                                                                                                                            	  logrisk = FALSE,
                                                                                                                            	  alpha = alphathresh)$power[2]
            PSYCH_POWER_5000[PSYCH_POWER_5000$vg2 == vg2est & PSYCH_POWER_5000$cor12 == cor12est & PSYCH_POWER_5000$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 87849,
	    		      		     	      	       		  	   			    		             n = c(438997, 5000),
			  		    										       	     vg1 = 0.25,
																     cov12 = cor12est*sqrt(vg2est*0.25),
			  		    										       	     pi0 = 0.95,
															       	     pupper = c(0,0.001,1),
			  		    										       	     nested = TRUE,
															       	     weighted = TRUE,
			  												       	     binary = c(TRUE, FALSE),
			  												       	     prevalence = c(0.3, NA),
			  												       	     sampling = c(0.368, NA),
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            PSYCH_POWER_5000[PSYCH_POWER_5000$vg2 == vg2est & PSYCH_POWER_5000$cor12 == cor12est & PSYCH_POWER_5000$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 87849,
                                                                                                                           	  n = c(438997,5000),
                                                                                                                            	  vg1 = 0.25,
                                                                                                                            	  cov12 = cor12est*sqrt(vg2est*0.25),
                                                                                                                            	  pi0 = 0.95,
                                                                                                                            	  pupper = c(0,0.001,1),
                                                                                                                            	  nested = TRUE,
                                                                                                                            	  weighted = TRUE,
                                                                                                                            	  binary = c(TRUE, FALSE),
                                                                                                                            	  prevalence = c(0.3, NA),
                                                                                                                            	  sampling = c(0.368, NA),
                                                                                                                            	  shrinkage = FALSE,
                                                                                                                            	  logrisk = FALSE,
                                                                                                                            	  alpha = alphathresh)$power[2]
            BMI_POWER_5000[BMI_POWER_5000$vg2 == vg2est & BMI_POWER_5000$cor12 == cor12est & BMI_POWER_5000$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 72919,
	    		      		     	      	       		  	   			    		       	     n = c(795640,5000),
			  		    										       	     vg1 = 0.138,
			 		    										       	     cov12 = cor12est*sqrt(vg2est*0.138),
			  		    										       	     pi0 = 0.95,
			  		    										       	     pupper = c(0,0.001,1),
			  		    										       	     nested = TRUE,
			  												       	     weighted = TRUE,
			  												       	     binary = FALSE,
			  												       	     shrinkage = FALSE,
			  												       	     logrisk = FALSE,
			  												       	     alpha = alphathresh)$power[1]
            BMI_POWER_5000[BMI_POWER_5000$vg2 == vg2est & BMI_POWER_5000$cor12 == cor12est & BMI_POWER_5000$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 72919,
	    		      		     	      	       		  	   			    		    	  n = c(795640,5000),
			  		    										    	  vg1 = 0.138,
			 		    										    	  cov12 = cor12est*sqrt(vg2est*0.138),
			  		    										    	  pi0 = 0.95,
			  		    										    	  pupper = c(0,0.001,1),
			  		    										    	  nested = TRUE,
			  												    	  weighted = TRUE,
			  												    	  binary = FALSE,
			  												    	  shrinkage = FALSE,
			  												    	  logrisk = FALSE,
			  												    	  alpha = alphathresh)$power[2]
            HEIGHT_POWER_5000[HEIGHT_POWER_5000$vg2 == vg2est & HEIGHT_POWER_5000$cor12 == cor12est & HEIGHT_POWER_5000$alpha == alphathresh, "power_best"] <- polygenescore(nsnp = 72976,
	    		      		     	      	       		  	   			    		  	           n = c(709706,5000),
			  		    										  	     	   vg1 = 0.347,
			 		    										  	     	   cov12 = cor12est*sqrt(vg2est*0.347),
			  		    										  	     	   pi0 = 0.95,
			  		    										  	     	   pupper = c(0,0.001,1),
			  		    										  	     	   nested = TRUE,
			  												  	     	   weighted = TRUE,
			  												  	     	   binary = FALSE,
			  												  	     	   shrinkage = FALSE,
			  												  	     	   logrisk = FALSE,
			  												  	     	   alpha = alphathresh)$power[1]
            HEIGHT_POWER_5000[HEIGHT_POWER_5000$vg2 == vg2est & HEIGHT_POWER_5000$cor12 == cor12est & HEIGHT_POWER_5000$alpha == alphathresh, "power_1"] <- polygenescore(nsnp = 72976,
	    		      		     	      	       		  	   			    		  	  	n = c(709706,5000),
			  		    										   	  	vg1 = 0.347,
			 		    										  	  	cov12 = cor12est*sqrt(vg2est*0.347),
			  		    										  	  	pi0 = 0.95,
			  		    										  	  	pupper = c(0,0.001,1),
			  		    										  	  	nested = TRUE,
			  												  	  	weighted = TRUE,
			  												  	  	binary = FALSE,
			  												  	  	shrinkage = FALSE,
			  												  	  	logrisk = FALSE,
			  												  	  	alpha = alphathresh)$power[2]
		}
	}
}


## Print output

write.table(BMI_POWER_5000, "iPSC_PRS/HipSci/Analysis/BMI_Power_Table_5000.txt", col.names=T, row.names=F, quote=F)
write.table(HEIGHT_POWER_5000, "iPSC_PRS/HipSci/Analysis/Height_Power_Table_5000.txt", col.names=T, row.names=F, quote=F)
write.table(SCZ_POWER_5000, "iPSC_PRS/HipSci/Analysis/SCZ_CLOZUK_Power_Table_5000.txt", col.names=T, row.names=F, quote=F)
write.table(PSYCH_POWER_5000, "iPSC_PRS/HipSci/Analysis/PSYCH_Power_Table_5000.txt", col.names=T, row.names=F, quote=F)



```
### Plot power graphs

For each power table, want to plot how power changes as a function of genetic correlation and variance explained

Run from command line

```{sh RunPlotPowerInR.sh}

for table in BMI_Power_Table.txt Height_Power_Table.txt SCZ_CLOZUK_Power_Table.txt PSYCH_Power_Table.txt \
BMI_Power_Table_200.txt Height_Power_Table_200.txt SCZ_CLOZUK_Power_Table_200.txt PSYCH_Power_Table_200.txt \
BMI_Power_Table_1000.txt Height_Power_Table_1000.txt SCZ_CLOZUK_Power_Table_1000.txt PSYCH_Power_Table_1000.txt \
BMI_Power_Table_850.txt Height_Power_Table_850.txt SCZ_CLOZUK_Power_Table_850.txt PSYCH_Power_Table_850.txt \
BMI_Power_Table_2435.txt Height_Power_Table_2435.txt SCZ_CLOZUK_Power_Table_2435.txt PSYCH_Power_Table_2435.txt \
BMI_Power_Table_5000.txt Height_Power_Table_5000.txt SCZ_CLOZUK_Power_Table_5000.txt PSYCH_Power_Table_5000.txt 
do
    pdfname=$(echo $table | sed 's/.txt/.pdf/g')
    trait=$(echo $table | sed 's/_.*//g')
    Npower=$(echo $table | sed 's/.*_//g' | sed 's/.txt//g')
    if [ $Npower = "Table" ]
    then
        Npower=60
    fi
    Rscript PlotPower.R $table $pdfname $trait $Npower
done

```

```{R PlotPower.R}

### Initialise command line 

args = commandArgs(trailingOnly=TRUE)

### Load packages
library(ggplot2)
library(data.table)

## Load data
PowerTable<-fread(args[1], data.table=F)

## Limit data for plotting - just alpha = 0.05 and vg2 0.1,0.2 etc. - note need to have modulo as 1, not 0.1

PowerTableReduced <- with(PowerTable, PowerTable[alpha == 0.05 & 
                                                   (vg2*10) %% 1 == 0,])

PowerTableReduced$vg2 <- as.factor(PowerTableReduced$vg2) 

## Plot

cbbPalette10 <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#999999", "#00FF00")

gplot <- ggplot(data = PowerTableReduced, mapping = aes(x = cor12, y = power_1, colour = vg2))

pdf(args[2])
gplot + 
  geom_point() +
  geom_line() +
  xlim(c(0,1)) +
  ylim(c(0,1)) +
  ggtitle(paste("Power Plot:",args[3]," N:",args[4], sep="")) +
  xlab("Genetic Correlation") +
  ylab("Power") +
  labs(colour = "Common\nGenetic\nComponent") +
  scale_color_manual(values = cbbPalette10) +
  theme_bw(base_size = 14)
dev.off()

```

### Run analyses in R

```{r Analyses}

## Set options
options(stringsAsFactors=F)

## Set seed
set.seed(604)

## Get packages
library(interactions)
library(purrr)
library(tibble)
library(dplyr)
library(tidyr)
library(ggpubr)
library(data.table)
library(lmerTest)
library(HLMdiag)

## Get data
# Read in all cell data files
temp = list.files(pattern="*.tsv")
# Create list of all cell data files
myfiles = lapply(temp, read.delim)

## Create single tibble comprising data needed for analysis
dataframe_data<- tibble(
name = map(myfiles, "name"),
experiment_name = map(myfiles, "experiment_name"),
cell_area = map(myfiles, "cell_area"),
cell_roundness = map(myfiles, "roundness"),
cell_ratio_w2l = map(myfiles, "ratio_w2l"),
p_col = map(myfiles, "p_col"),
p_row = map(myfiles, "p_row"),
field = map(myfiles, "field"),
clump_size = map(myfiles, "clump_size"))

## Convert lists to a data frame of all entries
Cell_Tibble<-dataframe_data %>%
select(name, experiment_name, cell_area, cell_roundness, cell_ratio_w2l, p_col, p_row, field, clump_size) %>%
unnest()

## Some rows in Cell Tibble are repeated due to repeat assessments - remove these

Cell_Tibble_Distinct <- distinct(Cell_Tibble)

## Get external info on cell lines

temp_plate = list.files(path="iPSC_PRS/HipSci/Images/Plate_Results", pattern="*.tsv", full.names=T)
myfiles_plate = lapply(temp_plate, read.delim)

### Plate fields

    # name                   The official name for each line as used within HipSci.  This is especially needed to link to other data from HipSci.

    # experiment_name        We assigned a number to each experiment as an internal record     
    # passage_number         Cell line passage number at the time of the experiment
    # p_row                  number corresponding to position of wells in row (1-8)
    # p_col    		     number corresponding to position of wells in column (A-H)

    # num_cells              number of objects segmented per well

    # nuc_area_mean          morphological value, mean per well
    # nuc_area_sd            morphological value, standard dev per well
    # nuc_area_sum           morphological value, sum per well
    # nuc_area_max           morphological value, max value
    # nuc_area_min           morphological value, min value
    # nuc_area_median        morphological value, median per well

    # nuc_roundness_mean     morphological value, mean per well
    # nuc_roundness_sd       morphological value, st dev per well
    # nuc_roundness_sum      morphological value, sum per well
    # nuc_roundness_max      morphological value, max per well
    # nuc_roundness_min      morphological value, min per well
    # nuc_roundness_median   morphological value, median per well

    # nuc_ratio_w2l_mean     morphological value, mean per well
    # nuc_ratio_w2l_sd       morphological value, standard dev per well
    # nuc_ratio_w2l_sum      morphological value, sum per well
    # nuc_ratio_w2l_max      morphological value, max value
    # nuc_ratio_w2l_min      morphological value, min value
    # nuc_ratio_w2l_median   morphological value, median per well

    # edu_median_mean        intensity value, mean per well
    # edu_median_sd          intensity value, median per well
    # edu_median_sum         intensity value, sum per well
    # edu_median_max         intensity value, max per well
    # edu_median_min         intensity value, min per well
    # edu_median_median      intensity value, median per well

    # oct4_median_mean       N/A
    # oct4_median_sd         N/A
    # oct4_median_sum        N/A
    # oct4_median_max        N/A
    # oct4_median_min        N/A
    # oct4_median_median     N/A

    # inten_nuc_dapi_median_mean    intensity value, mean per well
    # inten_nuc_dapi_median_sd      intensity value, st dev per well
    # inten_nuc_dapi_median_sum     intensity value, median per well
    # inten_nuc_dapi_median_max     intensity value, max per well
    # inten_nuc_dapi_median_min     intensity value, min per well
    # inten_nuc_dapi_median_median  intensity value, median per well

    # cells_per_clump_mean          context value, mean per well
    # cells_per_clump_sd            context value, st dev per well
    # cells_per_clump_sum           context value, sum per well
    # cells_per_clump_max           context value, max per well
    # cells_per_clump_min           context value, min per well
    # cells_per_clump_median        context value, median per well

    # area_mean              morphological value, mean per well
    # area_sd                morphological value, st dev per well
    # area_sum               morphological value, sum per well
    # area_max               morphological value, max per well
    # area_min               morphological value, min per well
    # area_median            morphological value, median per well

    # roundness_mean         morphological value, mean per well
    # roundness_sd           morphological value, st dev per well
    # roundness_sum          morphological value, sum per well
    # roundness_max          morphological value, max per well
    # roundness_min          morphological value, min per well
    # roundness_median       morphological value, median per well

    # ratio_w2l_mean         morphological value, mean per well
    # ratio_w2l_sd           morphological value, st dev per well
    # ratio_w2l_sum          morphological value, sum per well
    # ratio_w2l_max          morphological value, max per well
    # ratio_w2l_min          morphological value, min per well
    # ratio_w2l_median       morphological value, median per well

    # compound               FN=Fibronectin
    # concentration          1 - 5 - 25 ug/ml (0 for wells N/A)
    # cell_count             number of cells plated, typically 3000
    # num_fields             9

## Get fibronectin conc from plate file - can match on experiment, p_col and p_row

## Create single tibble comprising lists of names and six proposed phenos
dataframe_data_plate<- tibble(
name = map(myfiles_plate, "name"),
experiment_name = map(myfiles_plate, "experiment_name"),
p_col = map(myfiles_plate, "p_col"),
p_row = map(myfiles_plate, "p_row"),
fibro_conc = map(myfiles_plate, "concentration"))

## Convert lists to a data frame of all entries
Plate_Tibble<-dataframe_data_plate %>%
select(name, experiment_name, p_col, p_row, fibro_conc) %>%
unnest()

## Some rows in Plate Tibble are repeated due to repeat assessments - remove these

Plate_Tibble_Distinct <- distinct(Plate_Tibble)

## Add fibronectin to Cell Tibble

Cell_Tibble_Distinct$name_exp_col_row<-paste(Cell_Tibble_Distinct$name, Cell_Tibble_Distinct$experiment_name, Cell_Tibble_Distinct$p_col, Cell_Tibble_Distinct$p_row, sep="_")
Plate_Tibble_Distinct$name_exp_col_row<-paste(Plate_Tibble_Distinct$name, Plate_Tibble_Distinct$experiment_name, Plate_Tibble_Distinct$p_col, Plate_Tibble_Distinct$p_row, sep="_")

Plate_Tibble_Distinct_Short<-Plate_Tibble_Distinct[,c("name_exp_col_row", "fibro_conc")]

Cell_Tibble_Distinct_Fibro<-left_join(Cell_Tibble_Distinct, Plate_Tibble_Distinct_Short)

dim(Cell_Tibble_Distinct)
dim(Cell_Tibble_Distinct_Fibro)

## Add PRS and genomic PCs

BMI_PRS<-fread("../../Analysis/PRS_Output/BMI.all_score")
CDX_FULL_PRS<-fread("../../Analysis/PRS_Output/CDX_FULL.all_score")
CDX_10K_PRS<-fread("../../Analysis/PRS_Output/CXD_10K.all_score")
Height_PRS<-fread("../../Analysis/PRS_Output/Height.all_score")
SCZ_CLOZUK_PRS<-fread("../../Analysis/PRS_Output/SCZ_CLOZUK.all_score")

names(BMI_PRS)[3:4]<-paste("BMI",names(BMI_PRS)[3:4],sep="_")
names(CDX_FULL_PRS)[3:4]<-paste("CDX_FULL",names(CDX_FULL_PRS)[3:4],sep="_")
names(CDX_10K_PRS)[3]<-paste("CDX_10K",names(CDX_10K_PRS)[3],sep="_")
names(Height_PRS)[3:4]<-paste("Height",names(Height_PRS)[3:4],sep="_")
names(SCZ_CLOZUK_PRS)[3:4]<-paste("SCZ_CLOZUK",names(SCZ_CLOZUK_PRS)[3:4],sep="_")

PRS_TEMP1<-full_join(BMI_PRS, CDX_FULL_PRS)
PRS_TEMP2<-full_join(PRS_TEMP1, CDX_10K_PRS)
PRS_TEMP3<-full_join(PRS_TEMP2, Height_PRS)
PRS<-full_join(PRS_TEMP3, SCZ_CLOZUK_PRS)

PCs<-fread("../../Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF_LD3.pca_For_R.txt")

PRS_PCs<-full_join(PRS, PCs)

## External data from Vigilante / Kilpinen

## Combine all data

# Separate name into donor and line

Cell_Tibble_Distinct_Fibro<-separate(Cell_Tibble_Distinct_Fibro, "name", into=c("Donor","line"), sep="_", remove=F)
Cell_Tibble_Distinct_Fibro$line<-as.numeric(Cell_Tibble_Distinct_Fibro$line)

# Rename FID and IID to Donor and line

names(PRS_PCs)[1:2]<-c("Donor","line")

# Remove line from PRS data

PRS_PCs_No_Line<-PRS_PCs[,-grep("line", names(PRS_PCs))]


## Some lines are biological replicates of others (that is, the same donor across multiple lines)
## Change the names of these lines so that "donor" reflects all lines derived from that genome
## Cell_Tibble_Distinct_Fibro has no lines over 40, so add new lines for 40+

# # Donor fpdl is 4 donors:
#     HPSI0813i-fpdl_1
#     HPSI0913i-ffdl_1 [EXCLUDE --> Duplicate_Lines.txt]
#     HPSI1213i-foqj_2 [EXCLUDE --> Duplicate_Lines.txt]
#     HPSI0913i-ffdz_1 {EXCLUDE --> Duplicate_Lines.txt]
# # Donor fpdr is 3 donors:
#     HPSI0813i-ffdr_3 [EXCLUDE --> Duplicate_Lines.txt]
#     HPSI0813i-fpdr_1
#     HPSI1113i-fumb_1 [EXCLUDE --> Duplicate_Lines.txt]
# # Donor fpdk is 3 donors:
#     HPSI0813i-fpdk_2
#     HPSI1113i-zuta_1 [EXCLUDE --> Duplicate_Lines.txt]
# # Donor ffdj is 3 donors:
#     HPSI0813i-ffdj_1 [EXCLUDE --> Duplicate_Lines.txt]
#     HPSI0813i-fpdj_1
#     HPSI1113i-nibo_1 [EXCLUDE --> Duplicate_Lines.txt]
# # Donor ffdm is 4 donors:
#     HPSI0813i-ffdm_1 [EXCLUDE --> Duplicate_Lines.txt]
#     HPSI0813i-fpdm_1

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI0913i-ffdl", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0913i-ffdl" & line == 1), 40, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0913i-ffdl" & line == 2), 41, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0913i-ffdl"), "HPSI0813i-fpdl", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI1213i-foqj", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1213i-foqj" & line == 2), 42, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1213i-foqj"), "HPSI0813i-fpdl", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI0913i-ffdz", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0913i-ffdz" & line == 2), 43, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0913i-ffdz"), "HPSI0813i-fpdl", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI0813i-ffdr", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0813i-ffdr" & line == 3), 44, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0813i-ffdr"), "HPSI0813i-fpdr", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI1113i-fumb", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1113i-fumb" & line == 2), 45, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1113i-fumb"), "HPSI0813i-fpdr", Cell_Tibble_Distinct_Fibro$Donor)

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1113i-zuta" & line == 1), 46, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1113i-zuta"), "HPSI0813i-fpdk", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI0813i-ffdj", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0813i-ffdj" & line == 1), 47, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0813i-ffdj"), "HPSI0813i-fpdj", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI1113i-nibo", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1113i-nibo" & line == 2), 48, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1113i-nibo"), "HPSI0813i-fpdj", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI0813i-ffdm", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0813i-ffdm" & line == 3), 49, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0813i-ffdm"), "HPSI0813i-fpdm", Cell_Tibble_Distinct_Fibro$Donor)

# Merge data

Final_Temp1<-left_join(PRS_PCs_No_Line, Cell_Tibble_Distinct_Fibro)

# Confirm expected reduction in length

dim(Final_Temp1)

# Confirm 60 donors

length(unique(Final_Temp1$Donor))

# Confirm >60 names, implying multiple lines for some donors

length(unique(Final_Temp1$name))

## Add external data - tissue type, reprogramming and feeder from HipSci database / Vigilante Supplement --> CellOfOriginInfo.txt

CellOfOrigin <- fread("CellOfOriginInfo.txt")

Final<-left_join(Final_Temp1, CellOfOrigin)

Final$derived_from_tissue_type <- as.factor(Final$derived_from_tissue_type)
Final$reprogramming <- as.factor(Final$reprogramming)
Final$on_feeder <- as.factor(Final$on_feeder)

## Clean up

#rm(BMI_PRS,
CDX_10K_PRS,
CDX_FULL_PRS,
CellOfOrigin,
Cell_Tibble,
Cell_Tibble_Distinct,
Cell_Tibble_Distinct_Fibro,
dataframe_data,
dataframe_data_plate,
Final_Temp1,
Height_PRS,
myfiles,
myfiles_plate,
PCs,
Plate_Tibble,
Plate_Tibble_Distinct,
Plate_Tibble_Distinct_Short,
PRS,
PRS_PCs,
PRS_PCs_No_Line,
PRS_TEMP1,
PRS_TEMP2,
PRS_TEMP3,
SCZ_CLOZUK_PRS,
temp,
temp_plate)

## Save set-up

save.image("MLM_SetUp.Rdata")

### Examine data structure

## Is continuous data approximately normal and appropriate for scaling

pdf("Distributions_iPSC.pdf")
hist(Final$BMI_Pt_0.001, 100)
hist(Final$BMI_Pt_1,100)
hist(Final$CDX_FULL_Pt_0.001,100)
hist(Final$CDX_FULL_Pt_1,100)
hist(Final$CDX_10K_Pt_1,100)
hist(Final$Height_Pt_0.001,100)
hist(Final$Height_Pt_1,100)
hist(Final$SCZ_CLOZUK_Pt_0.05,100)
hist(Final$SCZ_CLOZUK_Pt_1,100)
hist(Final$PC1,100)
hist(Final$PC2,100)
hist(Final$PC3,100)
hist(Final$PC4,100)
hist(Final$PC5,100)
hist(Final$PC6,100)
hist(Final$PC7,100)
hist(Final$PC8,100)
hist(Final$PC9,100)
hist(Final$PC10,100)
hist(Final$PC11,100)
hist(Final$PC12,100)
hist(Final$PC13,100)
hist(Final$PC14,100)
hist(Final$PC15,100)
hist(Final$PC16,100)
hist(Final$PC17,100)
hist(Final$PC18,100)
hist(Final$PC19,100)
hist(Final$PC20,100)
hist(Final$cell_area,100)
hist(Final$cell_roundness,100)
hist(Final$cell_ratio_w2l,100)
hist(Final$clump_size,100)
dev.off()

## Scale continuous data for analysis

Final$BMI_Pt_0.001<-scale(Final$BMI_Pt_0.001)
Final$BMI_Pt_1<-scale(Final$BMI_Pt_1)
Final$CDX_FULL_Pt_0.001<-scale(Final$CDX_FULL_Pt_0.001)
Final$CDX_FULL_Pt_1<-scale(Final$CDX_FULL_Pt_1)
Final$CDX_10K_Pt_1<-scale(Final$CDX_10K_Pt_1)
Final$Height_Pt_0.001<-scale(Final$Height_Pt_0.001)
Final$Height_Pt_1<-scale(Final$Height_Pt_1)
Final$SCZ_CLOZUK_Pt_0.05<-scale(Final$SCZ_CLOZUK_Pt_0.05)
Final$SCZ_CLOZUK_Pt_1<-scale(Final$SCZ_CLOZUK_Pt_1)
Final$PC1<-scale(Final$PC1)
Final$PC2<-scale(Final$PC2)
Final$PC3<-scale(Final$PC3)
Final$PC4<-scale(Final$PC4)
Final$PC5<-scale(Final$PC5)
Final$PC6<-scale(Final$PC6)
Final$PC7<-scale(Final$PC7)
Final$PC8<-scale(Final$PC8)
Final$PC9<-scale(Final$PC9)
Final$PC10<-scale(Final$PC10)
Final$PC11<-scale(Final$PC11)
Final$PC12<-scale(Final$PC12)
Final$PC13<-scale(Final$PC13)
Final$PC14<-scale(Final$PC14)
Final$PC15<-scale(Final$PC15)
Final$PC16<-scale(Final$PC16)
Final$PC17<-scale(Final$PC17)
Final$PC18<-scale(Final$PC18)
Final$PC19<-scale(Final$PC19)
Final$PC20<-scale(Final$PC20)
Final$cell_area<-scale(Final$cell_area)
Final$cell_roundness<-scale(Final$cell_roundness)
Final$cell_ratio_w2l<-scale(Final$cell_ratio_w2l)
Final$clump_size<-scale(Final$clump_size)

### Note to self - the below requires ~16 threads (more than 8, anyway...) for interaction

### Initial analyses miscoded fibro_conc as linear, and did not account for NA fibro concs.

### Drop fibro_conc == 0

dim(Final)
#1579905      46

Final <- Final[Final$fibro_conc > 0, ]

dim(Final)
#1543624      46

### Set fibro_conc as a factor

Final$fibro_conc <- as.factor(Final$fibro_conc)

### Analysis Set 1 - no interaction
### 2020-10-02 Initial analyses used line as the nested factor, and did not consider well-level effects. Replaced line with well

sink(paste("Analysis_No_Interaction_",Sys.Date(),".txt",sep=""))

for(prs in c("BMI_Pt_0.001",
             "BMI_Pt_1",
             "CDX_FULL_Pt_0.001",
             "CDX_FULL_Pt_1",
             "CDX_10K_Pt_1",
             "Height_Pt_0.001",
             "Height_Pt_1",
             "SCZ_CLOZUK_Pt_0.05",
             "SCZ_CLOZUK_Pt_1")){
    for(pheno in c("cell_area", "cell_roundness", "cell_ratio_w2l")){
        ## Define formula
        analysis_formula <- as.formula(paste(pheno,
                                             "~",
                                             prs,
                                             "+ fibro_conc + PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder + (1 | Donor/name_exp_col_row)", sep =" "))
        print(analysis_formula)
        ## Run mixed linear model (~ 40-60 seconds)
        lmer_output_REML<-lmer(analysis_formula, data = Final, REML = T)
        ## Print summary of lmer, p from Satterthwaite approximation
        print(summary(lmer_output_REML))
        ## Rerun lmer using ML for likelihood ratio test (LRT) [still declare REML - see https://ourcodingclub.github.io/tutorials/mixed-models/]
        lmer_output_ML <- lmer(analysis_formula, data = Final, REML = F)
        ## Create model without PRS for likelihood ratio test (LRT)
        null_formula<-as.formula(paste(pheno,"~ fibro_conc + PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder + (1 | Donor/name_exp_col_row)", sep =" "))
        lmer_null_ML<-lmer(null_formula, data = Final, REML = F)
        ## Perform LRT (no bootstrap)
        print(anova(lmer_null_ML, lmer_output_ML))
    }
}

sink()

save.image("DataForAnalysis.Rdata")

### Interesting results are seen for CDX_FULL_Pt_1 and cell area.
### There are others significant at p < 0.05, but these are expected under the null - the most interesting of these is for CDX_FULL_Pt_1 and cell roundness.

# Linear mixed model fit by REML. t-tests use Satterthwaite's method [lmerModLmerTest]
# Formula: analysis_formula
#    Data: Final

# REML criterion at convergence: 4035831

# Scaled residuals:
#     Min      1Q  Median      3Q     Max
# -2.7710 -0.6586 -0.1695  0.4700  6.0209

# Random effects:
#  Groups                 Name        Variance Std.Dev.
#  name_exp_col_row:Donor (Intercept) 0.07389  0.2718
#  Donor                  (Intercept) 0.03851  0.1962
#  Residual                           0.79497  0.8916
# Number of obs: 1543624, groups:  name_exp_col_row:Donor, 2484; Donor, 60

# Fixed effects:
#                                Estimate Std. Error         df  t value Pr(>|t|)
# (Intercept)                  -4.887e-01  4.755e-02  2.481e+02  -10.278  < 2e-16 ***
# CDX_FULL_Pt_1                 8.450e-02  2.768e-02  5.150e+01    3.052  0.00359 ** 
# fibro_conc5                   4.070e-01  1.367e-02  2.438e+03   29.784  < 2e-16 ***
# fibro_conc25                  7.435e-01  1.366e-02  2.432e+03   54.440  < 2e-16 ***
# PC1                          -3.276e-02  2.810e-02  5.017e+01   -1.166  0.24918    
# PC2                          -1.454e-02  2.666e-02  5.130e+01   -0.545  0.58779    
# PC3                           4.645e-02  2.634e-02  5.166e+01    1.763  0.08378 .  
# PC4                           5.099e-02  2.678e-02  5.554e+01    1.904  0.06209 .  
# clump_size                   -1.913e-01  1.694e-03  1.399e+06 -112.950  < 2e-16 ***
# derived_from_tissue_typepbmc -4.086e-02  4.298e-02  1.664e+03   -0.951  0.34198    
# reprogrammingsendai           9.711e-02  4.389e-02  9.128e+02    2.213  0.02718 *  
# on_feederfeeder-free         -4.592e-01  3.505e-02  1.923e+03  -13.101  < 2e-16 ***
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# Correlation of Fixed Effects:
#             (Intr) CDX_FU fbr_c5 fbr_25 PC1    PC2    PC3    PC4    clmp_s drv___ rprgrm
# CDX_FULL_P_  0.062							   		 
# fibro_conc5 -0.146  0.000						   		 
# fibro_cnc25 -0.146  0.000  0.508					   		 
# PC1          0.041  0.013 -0.001  0.000					   		 
# PC2          0.035 -0.091  0.000  0.000  0.003				   		 
# PC3         -0.068  0.164  0.000  0.000  0.004 -0.009			   		 
# PC4         -0.187  0.090  0.000  0.000 -0.013 -0.012  0.023		   		 
# clump_size   0.010  0.000 -0.007 -0.007  0.000  0.000  0.001  0.000	   		 
# drvd_frm_t_  0.513  0.060 -0.001 -0.001  0.056  0.011  0.050 -0.150  0.000 		 
# rprgrmmngsn -0.807 -0.055  0.001  0.001 -0.045 -0.039 -0.035  0.213  0.000 -0.657	 
# on_fdrfdr-f -0.092  0.001 -0.008 -0.008 -0.022  0.115  0.063  0.007 -0.004 -0.015  0.028

### Follow-up - does altering the coding of clump_size have an important effect?
### It is difficult for more than four cells to aggregate without at least one cell being "inner".
### Inner cells 

load("MLM_SetUp.Rdata")

## Scale continuous data for analysis

Final$BMI_Pt_0.001<-scale(Final$BMI_Pt_0.001)
Final$BMI_Pt_1<-scale(Final$BMI_Pt_1)
Final$CDX_FULL_Pt_0.001<-scale(Final$CDX_FULL_Pt_0.001)
Final$CDX_FULL_Pt_1<-scale(Final$CDX_FULL_Pt_1)
Final$CDX_10K_Pt_1<-scale(Final$CDX_10K_Pt_1)
Final$Height_Pt_0.001<-scale(Final$Height_Pt_0.001)
Final$Height_Pt_1<-scale(Final$Height_Pt_1)
Final$SCZ_CLOZUK_Pt_0.05<-scale(Final$SCZ_CLOZUK_Pt_0.05)
Final$SCZ_CLOZUK_Pt_1<-scale(Final$SCZ_CLOZUK_Pt_1)
Final$PC1<-scale(Final$PC1)
Final$PC2<-scale(Final$PC2)
Final$PC3<-scale(Final$PC3)
Final$PC4<-scale(Final$PC4)
Final$PC5<-scale(Final$PC5)
Final$PC6<-scale(Final$PC6)
Final$PC7<-scale(Final$PC7)
Final$PC8<-scale(Final$PC8)
Final$PC9<-scale(Final$PC9)
Final$PC10<-scale(Final$PC10)
Final$PC11<-scale(Final$PC11)
Final$PC12<-scale(Final$PC12)
Final$PC13<-scale(Final$PC13)
Final$PC14<-scale(Final$PC14)
Final$PC15<-scale(Final$PC15)
Final$PC16<-scale(Final$PC16)
Final$PC17<-scale(Final$PC17)
Final$PC18<-scale(Final$PC18)
Final$PC19<-scale(Final$PC19)
Final$PC20<-scale(Final$PC20)
Final$cell_area<-scale(Final$cell_area)
Final$cell_roundness<-scale(Final$cell_roundness)
Final$cell_ratio_w2l<-scale(Final$cell_ratio_w2l)

Final <- Final[Final$fibro_conc > 0, ]

Final$fibro_conc <- as.factor(Final$fibro_conc)

Final$clump_size_factor <- factor(ifelse(Final$clump_size == 1, "One",
			   	         ifelse(Final$clump_size == 2 | Final$clump_size == 3, "Two or Three",
				                ifelse(Final$clump_size > 3, "More than Three", NA))),
				  ordered = TRUE, 
                                  levels = c("One", "Two or Three", "More than Three"))

Final$clump_size<-scale(Final$clump_size)

table(Final$clump_size_factor, useNA="a")

#            One    Two or Three More than Three            <NA>
#         372962          354894          815768               0

prs <- "CDX_FULL_Pt_1"
pheno <- "cell_area"

analysis_formula <- as.formula(paste(pheno,
                                     "~",
                                     prs,
                                     "+ fibro_conc + PC1 + PC2 + PC3 + PC4 + clump_size_factor + derived_from_tissue_type + reprogramming + on_feeder + (1 | Donor/name_exp_col_row)", sep =" "))
print(analysis_formula)
## Run mixed linear model (~ 40-60 seconds)
lmer_output_REML<-lmer(analysis_formula, data = Final, REML = T)
## Print summary of lmer, p from Satterthwaite approximation
print(summary(lmer_output_REML))

# Linear mixed model fit by REML. t-tests use Satterthwaite's method [lmerModLmerTest]
# Formula: analysis_formula
#    Data: Final

# REML criterion at convergence: 4034936

# Scaled residuals:
#     Min      1Q  Median      3Q     Max
# -2.7725 -0.6588 -0.1776  0.4698  5.9504

# Random effects:
#  Groups                 Name        Variance Std.Dev.
#  name_exp_col_row:Donor (Intercept) 0.07577  0.2753
#  Donor                  (Intercept) 0.03938  0.1984
#  Residual                           0.79447  0.8913
# Number of obs: 1543624, groups:  name_exp_col_row:Donor, 2484; Donor, 60

# Fixed effects:
#                                Estimate Std. Error         df t value Pr(>|t|)
# (Intercept)                  -4.349e-01  4.811e-02  2.475e+02  -9.038  < 2e-16 ***
# CDX_FULL_Pt_1                 8.360e-02  2.800e-02  5.137e+01   2.986  0.00432 ** 
# fibro_conc5                   3.959e-01  1.383e-02  2.443e+03  28.628  < 2e-16 ***
# fibro_conc25                  7.278e-01  1.382e-02  2.437e+03  52.650  < 2e-16 ***
# PC1                          -3.101e-02  2.842e-02  5.005e+01  -1.091  0.28035    
# PC2                          -2.044e-02  2.696e-02  5.118e+01  -0.758  0.45190    
# PC3                           4.381e-02  2.664e-02  5.153e+01   1.644  0.10618    
# PC4                           4.808e-02  2.708e-02  5.541e+01   1.775  0.08132 .  
# clump_size_factor.L          -1.398e-01  1.464e-03  1.540e+06 -95.505  < 2e-16 ***
# clump_size_factor.Q           1.061e-01  1.460e-03  1.543e+06  72.701  < 2e-16 ***
# derived_from_tissue_typepbmc -2.148e-02  4.350e-02  1.663e+03  -0.494  0.62159    
# reprogrammingsendai           8.642e-02  4.442e-02  9.098e+02   1.945  0.05204 .  
# on_feederfeeder-free         -4.476e-01  3.548e-02  1.923e+03 -12.617  < 2e-16 ***
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

### No significant difference in estimate (even though the coding of clump size does effect whether the estimate is significant or not)

### Follow-up - does altering the number of PCs alter the CDX result?

prs <- "CDX_FULL_Pt_1"
pheno <- "cell_area"

# Rerun initial model

analysis_formula <- as.formula(paste(pheno,
                                     "~",
                                     prs,
                                     "+ fibro_conc + PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder + (1 | Donor/name_exp_col_row)", sep =" "))
print(analysis_formula)
## Run mixed name_exp_col_rowar model (~ 40-60 seconds)
lmer_output_REML<-lmer(analysis_formula, data = Final, REML = T)
## Print summary of lmer, p from Satterthwaite approximation
print(summary(lmer_output_REML))

# Add PCs 5-10

larger_analysis_formula <- as.formula(paste(pheno,
                                     "~",
                                     prs,
                                     "+ fibro_conc + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + clump_size + derived_from_tissue_type + reprogramming + on_feeder + (1 | Donor/name_exp_col_row)", sep =" "))
print(larger_analysis_formula)
## Run mixed linear model (~ 40-60 seconds)
lmer_larger_output_REML<-lmer(larger_analysis_formula, data = Final, REML = T)
## Print summary of lmer, p from Satterthwaite approximation
print(summary(lmer_larger_output_REML))

# Linear mixed model fit by REML. t-tests use Satterthwaite's method [
# lmerModLmerTest]
# Formula: larger_analysis_formula
#    Data: Final

# REML criterion at convergence: 4035858

# Scaled residuals:
#     Min      1Q  Median      3Q     Max
# -2.7710 -0.6586 -0.1695  0.4700  6.0209

# Random effects:
#  Groups                 Name        Variance Std.Dev.
#  name_exp_col_row:Donor (Intercept) 0.07386  0.2718
#  Donor                  (Intercept) 0.03925  0.1981
#  Residual                           0.79497  0.8916
# Number of obs: 1543624, groups:  name_exp_col_row:Donor, 2484; Donor, 60

# Fixed effects:
#                                Estimate Std. Error         df  t value Pr(>|t|)
# (Intercept)                  -4.968e-01  4.867e-02  2.458e+02  -10.207  < 2e-16 ***
# CDX_FULL_Pt_1                 8.076e-02  2.874e-02  4.734e+01    2.810  0.00718	** 
# fibro_conc5                   4.070e-01  1.366e-02  2.439e+03   29.791  < 2e-16	***
# fibro_conc25                  7.435e-01  1.366e-02  2.433e+03   54.451  < 2e-16	***
# PC1                          -3.274e-02  2.836e-02  4.575e+01   -1.155  0.25422	   
# PC2                          -1.464e-02  2.690e-02  4.674e+01   -0.544  0.58888	   
# PC3                           4.527e-02  2.660e-02  4.710e+01    1.702  0.09535	.  
# PC4                           5.201e-02  2.702e-02  5.081e+01    1.925  0.05988	.  
# PC5                           1.035e-02  2.536e-02  4.612e+01    0.408  0.68518	   
# PC6                          -2.843e-02  2.652e-02  4.698e+01   -1.072  0.28907	   
# PC7                           1.071e-02  2.549e-02  4.760e+01    0.420  0.67633	   
# PC8                           3.855e-02  2.841e-02  4.866e+01    1.357  0.18104	   
# PC9                           3.383e-02  2.177e-02  4.804e+01    1.554  0.12680	   
# PC10                          5.583e-03  2.602e-02  4.665e+01    0.215  0.83100	   
# clump_size                   -1.934e-01  1.712e-03  1.399e+06 -112.950  < 2e-16	***
# derived_from_tissue_typepbmc -5.003e-02  4.361e-02  1.838e+03   -1.147  0.25143	   
# reprogrammingsendai           1.093e-01  4.483e-02  1.070e+03    2.437  0.01495	*  
# on_feederfeeder-free         -4.620e-01  3.520e-02  1.946e+03  -13.125  < 2e-16	***
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


# Remove PCs 3-4

smaller_analysis_formula <- as.formula(paste(pheno,
                                     "~",
                                     prs,
                                     "+ fibro_conc + PC1 + PC2 + clump_size + derived_from_tissue_type + reprogramming + on_feeder + (1 | Donor/name_exp_col_row)", sep =" "))
print(smaller_analysis_formula)
## Run mixed linear model (~ 40-60 seconds)
lmer_smaller_output_REML<-lmer(smaller_analysis_formula, data = Final, REML = T)
## Print summary of lmer, p from Satterthwaite approximation
print(summary(lmer_smaller_output_REML))

# Linear mixed model fit by REML. t-tests use Satterthwaite's method [
# lmerModLmerTest]
# Formula: smaller_analysis_formula
#    Data: Final

# REML criterion at convergence: 4035827

# Scaled residuals:
#     Min      1Q  Median      3Q     Max
# -2.7709 -0.6586 -0.1695  0.4700  6.0209

# Random effects:
#  Groups                 Name        Variance Std.Dev.
#  name_exp_col_row:Donor (Intercept) 0.07390  0.2719
#  Donor                  (Intercept) 0.04141  0.2035
#  Residual                           0.79497  0.8916
# Number of obs: 1543624, groups:  name_exp_col_row:Donor, 2484; Donor, 60

# Fixed effects:
#                                Estimate Std. Error         df  t value Pr(>|t|)
# (Intercept)                  -4.698e-01  4.737e-02  2.282e+02   -9.917   <2e-16 ***
# CDX_FULL_Pt_1                 7.198e-02  2.815e-02  5.217e+01    2.557   0.0135 *  
# fibro_conc5                   4.070e-01  1.367e-02  2.437e+03   29.781   <2e-16 ***
# fibro_conc25                  7.435e-01  1.366e-02  2.431e+03   54.435   <2e-16 ***
# PC1                          -3.238e-02  2.908e-02  5.112e+01   -1.113   0.2708   
# PC2                          -1.376e-02  2.758e-02  5.227e+01   -0.499   0.6199	   
# clump_size                   -1.934e-01  1.712e-03  1.399e+06 -112.952   <2e-16 ***
# derived_from_tissue_typepbmc -3.447e-02  4.263e-02  1.564e+03   -0.808   0.4190   
# reprogrammingsendai           8.495e-02  4.320e-02  7.889e+02    1.966   0.0496 *  
# on_feederfeeder-free         -4.640e-01  3.510e-02  1.967e+03  -13.220   <2e-16 ***
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# Correlation of Fixed Effects:
#             (Intr) CDX_FU fbr_c5 fbr_25 PC1    PC2    clmp_s drv___ rprgrm
# CDX_FULL_P_  0.091
# fibro_conc5 -0.147  0.000
# fibro_cnc25 -0.147  0.000  0.508
# PC1          0.038  0.013 -0.001  0.000
# PC2          0.033 -0.090  0.000  0.000  0.003
# clump_size   0.010  0.000 -0.007 -0.007  0.000  0.000
# drvd_frm_t_  0.503  0.065 -0.001 -0.001  0.053  0.010  0.000
# rprgrmmngsn -0.798 -0.069  0.001  0.001 -0.042 -0.037 -0.001 -0.650
# on_fdrfdr-f -0.086 -0.010 -0.008 -0.008 -0.022  0.113 -0.004 -0.017  0.028

# Different, but not significantly so

mean1 <- 8.450e-02  
mean2 <- 7.198e-02
se1 <- 2.768e-02
se2 <- 2.815e-02

zdiff <- (mean1 - mean2)/sqrt((se1*se1) + (se2*se2))
pdiff <- 2*pnorm(-abs(zdiff))
print(pdiff)

# 0.7511455

## Test robustness of the CDX effect to deletion of one sample (leave-one-sample-out analysis)

# Define formula
analysis_formula <- as.formula(paste(pheno,
                                     "~",
                                     prs,
                                     "+ fibro_conc + PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder + (1 | Donor/name_exp_col_row)", sep =" "))
print(analysis_formula)

# Run mixed linear model (~ 40-60 seconds)
lmer_output_REML<-lmer(analysis_formula, data = Final, REML = T)

# Leave-one-sample-out analyses

Donors<-unique(Final$Donor)

sink(paste("Analysis_Without_Interaction_Leave_One_Donor_Out_",Sys.Date(),".txt",sep=""))

for(prs in c("BMI_Pt_0.001",
         "BMI_Pt_1",
         "CDX_FULL_Pt_0.001",
         "CDX_FULL_Pt_1",
         "CDX_10K_Pt_1",
         "Height_Pt_0.001",
         "Height_Pt_1",
         "SCZ_CLOZUK_Pt_0.05",
         "SCZ_CLOZUK_Pt_1")){
    for(pheno in c("cell_area", "cell_roundness", "cell_ratio_w2l")){
        for(dropdonor in Donors){
            ## Define formula
	        analysis_formula <- as.formula(paste(pheno,
                                                     "~",
                                                     prs,
                                                     "+ fibro_conc + PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder + (1 | Donor/name_exp_col_row)", sep =" "))
	        print(analysis_formula)
            ## Run mixed linear model (~ 40-60 seconds)
            lmer_output_REML<-lmer(analysis_formula, data = Final[Final$Donor != dropdonor, ], REML = T)
            ## Print summary of lmer, p from Satterthwaite approximation
            print(summary(lmer_output_REML))	    
        }
    }
}

sink()

### Analysis Set 2 - interactions, including plotting

sink(paste("Analysis_With_Interaction_",Sys.Date(),".txt",sep=""))

for(prs in c("BMI_Pt_0.001",
             "BMI_Pt_1",
             "CDX_FULL_Pt_0.001",
             "CDX_FULL_Pt_1",
             "CDX_10K_Pt_1",
             "Height_Pt_0.001",
             "Height_Pt_1",
             "SCZ_CLOZUK_Pt_0.05",
             "SCZ_CLOZUK_Pt_1")){
    for(pheno in c("cell_area", "cell_roundness", "cell_ratio_w2l")){
        ## Define formula
        interaction_analysis_formula <- as.formula(paste(pheno,
                                             "~",
                                             prs,
                                             "*(fibro_conc + PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder) +
					      fibro_conc*(PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder) + (1 | Donor/name_exp_col_row)", sep =" "))
        print(interaction_analysis_formula)
        ## Run mixed linear model
        lmer_output_REML<-lmer(interaction_analysis_formula, data = Final, REML = T)
        ## Print summary of lmer, p from Satterthwaite approximation)
        print(summary(lmer_output_REML))
        ## Rerun lmer using ML for likelihood ratio test (LRT) [still declare REML - see https://ourcodingclub.github.io/tutorials/mixed-models/]
        lmer_output_ML <- lmer(interaction_analysis_formula, data = Final, REML = F)
        ## Create model without PRS for likelihood ratio test (LRT)
        null_formula<-as.formula(paste(pheno,
				       "~",
                                       prs,
                                       "*(PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder) +
                                       fibro_conc*(PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder) + (1 | Donor/name_exp_col_row)", sep =" "))
        lmer_null_ML<-lmer(null_formula, data = Final, REML = F)
        ## Perform LRT (no bootstrap)
        print(anova(lmer_null_ML, lmer_output_ML))
    }
}

sink()

### Visualise the data to enable understanding

for(prs in c("BMI_Pt_0.001",
             "BMI_Pt_1",
             "CDX_FULL_Pt_0.001",
             "CDX_FULL_Pt_1",
             "CDX_10K_Pt_1",
             "Height_Pt_0.001",
             "Height_Pt_1",
             "SCZ_CLOZUK_Pt_0.05",
             "SCZ_CLOZUK_Pt_1")){
    for(pheno in c("cell_area", "cell_roundness", "cell_ratio_w2l")){
        ## Define formula
        interaction_analysis_formula <- as.formula(paste(pheno,
                                             "~",
                                             prs,
                                             "*(fibro_conc + PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder) +
                                              fibro_conc*(PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder) + (1 | Donor/name_exp_col_row)", sep =" "))
        print(interaction_analysis_formula)
        ## Run mixed linear model
        lmer_output_REML<-lmer(interaction_analysis_formula, data = Final, REML = T)

	## Add fit statistics to Final

	Final$fit <- predict(lmer_output_REML)

	Final_agg <- Final %>% 
		     group_by(Donor, BMI_Pt_0.001, BMI_Pt_1, CDX_FULL_Pt_0.001, CDX_FULL_Pt_1, CDX_10K_Pt_1, Height_Pt_0.001, Height_Pt_1, SCZ_CLOZUK_Pt_0.05, SCZ_CLOZUK_Pt_1, fibro_conc) %>% 
		     summarise(mean_fit = mean(fit))
	Final_TEMP<-suppressWarnings(left_join(Final, Final_agg))
	Final_TEMP<-unique(Final_TEMP[,c("Donor",prs,"mean_fit","fibro_conc")])
	names(Final_TEMP)[3:4]<-c(pheno,"modx_group")

	pheno_pretty<-ifelse(pheno == "cell_area", "Cell area",
		      ifelse(pheno == "cell_roundness", "Cell roundness",
		      ifelse(pheno == "cell_ratio_w2l", "Cell width:length", NA)))
	
	prs_pretty<-ifelse(prs == "BMI_Pt_0.001", "BMI PGS (pThresh < 0.001)",
		    ifelse(prs == "BMI_Pt_1", "BMI PGS (pThresh < 1)",
		    ifelse(prs == "CDX_FULL_Pt_0.001", "Cross-psychiatric PGS (pThresh < 0.001)",
		    ifelse(prs == "CDX_FULL_Pt_1", "Cross-psychiatric PGS (pThresh < 1)",
             	    ifelse(prs == "CDX_10K_Pt_1", "Cross-psychiatric + 23andME PGS",
             	    ifelse(prs == "Height_Pt_0.001", "Height PGS (pThresh < 0.001)",
             	    ifelse(prs == "Height_Pt_1", "Height PGS (pThresh < 1)",
             	    ifelse(prs == "SCZ_CLOZUK_Pt_0.05", "Schiziophrenia PGS (pThresh < 0.05)",
             	    ifelse(prs == "SCZ_CLOZUK_Pt_1", "Schiziophrenia PGS (pThresh < 1)", NA)))))))))

	pdfname<-paste("SplitEffectPlot_",prs,"_",pheno,".pdf",sep="")
	pdf(pdfname)
	    print(interact_plot(lmer_output_REML,pred = !!prs, modx = fibro_conc, interval=T, colors="CUD", legend.main="Fibronectin\nConcentration", x.label = prs_pretty, y.label = pheno_pretty) +
	    geom_point(data=Final_TEMP, aes(col = modx_group, shape = modx_group)) +
	    xlim(c(-4,4)) +
	    ylim(c(-1.5,1.5)) +
	    theme(text = element_text(size = 14)) + 
	    guides(shape = FALSE, fill=FALSE))
	dev.off()
    }
}


```


Determine appropriate multiple testing correction in R

```{R MultipleTestingCorrection}

### Load packages
library(data.table)
library(dplyr)

### Assess multiple testing as effective number of tests across 3 phenotypes, using number of PCs to explain 99.5% of variance in the correlation matrix

## Import phenotypes from analyses above

load("DataForAnalysis.Rdata")

PhenoCor <- cor(Final[,c("cell_area", "cell_roundness", "cell_ratio_w2l")])

PhenoCorPCA <- prcomp(PhenoCor, scale=F)

summary(PhenoCorPCA)

# Importance of components:
#                           PC1     PC2       PC3
# Standard deviation     1.2629 0.17989 6.946e-17
# Proportion of Variance 0.9801 0.01989 0.000e+00
# Cumulative Proportion  0.9801 1.00000 1.000e+00

# Phenotypes reflect 2 effective tests

## Does this differ if you collapse multiple measurements?

Unique_Donors<-data.frame(Donor=unique(Final$Donor), 
			  mean_area = NA, 
 			  mean_roundness = NA, 
			  mean_width_length = NA, 
			  median_area = NA, 
			  median_roundness = NA, 
			  median_width_length = NA)

for(donor in Unique_Donors$Donor){
    Unique_Donors[Unique_Donors$Donor == donor, "mean_area"]<-mean(Final[Final$Donor == donor, "cell_area"])
    Unique_Donors[Unique_Donors$Donor == donor, "median_area"]<-median(Final[Final$Donor == donor, "cell_area"])
    Unique_Donors[Unique_Donors$Donor == donor, "mean_roundness"]<-mean(Final[Final$Donor == donor, "cell_roundness"])
    Unique_Donors[Unique_Donors$Donor == donor, "median_roundness"]<-median(Final[Final$Donor == donor, "cell_roundness"])
    Unique_Donors[Unique_Donors$Donor == donor, "mean_width_length"]<-mean(Final[Final$Donor == donor, "cell_ratio_w2l"])
    Unique_Donors[Unique_Donors$Donor == donor, "median_width_length"]<-median(Final[Final$Donor == donor, "cell_ratio_w2l"])
}

PhenoCorMeans <- cor(Unique_Donors[,c("mean_area", "mean_roundness", "mean_width_length")])
PhenoCorMeansPCA <- prcomp(PhenoCorMeans, scale=F)

summary(PhenoCorMeansPCA)

# Importance of components:
#                           PC1    PC2       PC3
# Standard deviation     1.4018 0.1882 1.061e-16
# Proportion of Variance 0.9823 0.0177 0.000e+00
# Cumulative Proportion  0.9823 1.0000 1.000e+00

PhenoCorMedians <- cor(Unique_Donors[,c("median_area", "median_roundness", "median_width_length")])
PhenoCorMediansPCA <- prcomp(PhenoCorMedians, scale=F)

summary(PhenoCorMediansPCA)

# Importance of components:
#                           PC1     PC2       PC3
# Standard deviation     1.3186 0.19418 4.574e-17
# Proportion of Variance 0.9788 0.02123 0.000e+00
# Cumulative Proportion  0.9788 1.00000 1.000e+00

### Assess multiple testing as effective number of tests across 9 PRS, using number of PCs to explain 99.5% of variance in the correlation matrix

## Import PRS

BMI_PRS<-fread("../../Analysis/PRS_Output/BMI.all_score")
CDX_FULL_PRS<-fread("../../Analysis/PRS_Output/CDX_FULL.all_score")
CDX_10K_PRS<-fread("../../Analysis/PRS_Output/CXD_10K.all_score")
Height_PRS<-fread("../../Analysis/PRS_Output/Height.all_score")
SCZ_CLOZUK_PRS<-fread("../../Analysis/PRS_Output/SCZ_CLOZUK.all_score")

names(BMI_PRS)[3:4]<-paste("BMI",names(BMI_PRS)[3:4],sep="_")
names(CDX_FULL_PRS)[3:4]<-paste("CDX_FULL",names(CDX_FULL_PRS)[3:4],sep="_")
names(CDX_10K_PRS)[3]<-paste("CDX_10K",names(CDX_10K_PRS)[3],sep="_")
names(Height_PRS)[3:4]<-paste("Height",names(Height_PRS)[3:4],sep="_")
names(SCZ_CLOZUK_PRS)[3:4]<-paste("SCZ_CLOZUK",names(SCZ_CLOZUK_PRS)[3:4],sep="_")

PRS_TEMP1<-full_join(BMI_PRS, CDX_FULL_PRS)
PRS_TEMP2<-full_join(PRS_TEMP1, CDX_10K_PRS)
PRS_TEMP3<-full_join(PRS_TEMP2, Height_PRS)
PRS<-full_join(PRS_TEMP3, SCZ_CLOZUK_PRS)

PRSCor <- subset(PRS, select=-c(FID,IID))

PRSCorPCA <- prcomp(PRSCor, scale=F)

summary(PRSCorPCA)

# Importance of components:
                             PC1       PC2       PC3       PC4       PC5       PC6       PC7       PC8       PC9
Standard deviation     0.0004821 0.0001224 0.0000844 5.134e-05 4.056e-05 3.515e-05 1.351e-05 4.332e-06 2.819e-06
Proportion of Variance 0.8930900 0.0575400 0.0273700 1.013e-02 6.320e-03 4.750e-03 7.000e-04 7.000e-05 3.000e-05
Cumulative Proportion  0.8930900 0.9506300 0.9780000 9.881e-01 9.944e-01 9.992e-01 9.999e-01 1.000e+00 1.000e+00

# PRS reflect 2 effective tests => total is 2 * 6 = 12 independent tests (and so alpha = 0.05/12 = 0.00416)

```


### Supplementary analyses

Are the 280 cell lines genetically identical without filtering by LD? That is, to what extent do different cell lines differ?

```{sh Check_Genetic_Similarity_Of_Lines}

### Extract just genotyped SNPs from data before filtering to unique individuals

cd iPSC_PRS/HipSci

Software/plink \
--bfile Genotypes/Intermediate_Files/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001 \
--extract Genotypes/OpenAccess/MergedOA_Data.bim \
--genome \
--out Genotypes/Intermediate_Files/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_Just_Genotyped

awk 'NR == 1 || $10 > 0.1 || $1 == $3 {print}' Genotypes/Intermediate_Files/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_Just_Genotyped.genome | \
sort -k10,10g | less -S

## Results are very highly concordant - lowest discordance is:
#
#  HPSI1113i-eofe                3   HPSI1113i-eofe                2 OT     0  0.0001  0.0313  0.9686  0.9843  -1  0.994325  1.0000      NA
#

### Compare this to the LD-pruned genome results

Software/plink \
--bfile Genotypes/Intermediate_Files/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001 \
--extract Genotypes/OpenAccess/MergedOA_Data.bim \
--indep-pairwise 1500 150 0.2 \
--out Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_Just_Genotyped_LD1

## 40678 variants

Software/plink \
--bfile Genotypes/Intermediate_Files/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001 \
--extract Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_Just_Genotyped_LD1.prune.in \
--genome \
--out Genotypes/Intermediate_Files/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001_Just_Genotyped_LD2

## Results are very highly concordant - lowest discordance is:
#
#    HPSI1113i-eofe                3   HPSI1113i-eofe                2 OT     0  0.0000  0.0315  0.9685  0.9842  -1  0.993604  1.0000      NA
#

### Difference between lines is minimal and within the range of normal error when genotyping duplicate samples 
### Acceptable to calculate PRS as a donor-level value, not a cellular level value

```

### Obtain meta-data for initial cell lines included

```{bash GetInitialMetaData}

## Unique_Cell...fam contains IDs for the initial 280 lines considered
## hipsci_20151028.ped contains meta-data for all HipSci cell lines
## Extract just the 280 lines from the meta-data


## The coding for the non-iPSCs is not consistent between the files
## The first set of commands below pull out all entries for each of the 103 donors (including lines without gtarray data
## It keeps only the source cells (which have a 0 in the maternal ID column)
## Then the second line pulls out the iPSC cell lines and appends them to the previously-made file

LANG=C grep -wf \
<(awk '{print $1}' Genotypes/Intermediate_Files/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001.fam) \
iPSC_PRS/HipSci/Genotypes/OpenAccess/hipsci_20151028.ped | awk '$7 == 0 {print}' \
> iPSC_PRS/HipSci/Genotypes/Initial_Cell_Lines.ped

LANG=C grep -wf \
<(awk '{print $1"_"$2}' Genotypes/Intermediate_Files/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001.fam) \
iPSC_PRS/HipSci/Genotypes/OpenAccess/hipsci_20151028.ped \
>> iPSC_PRS/HipSci/Genotypes/Initial_Cell_Lines.ped

## Confirm lines are the same 

vimdiff <(awk '{print $2}' iPSC_PRS/HipSci/Genotypes/Initial_Cell_Lines.ped | tr '_' ' ' | sort -k1,1) <(awk '{print $1, $2}' Genotypes/Intermediate_Files/Unique_Cell_Result_Individuals_Imputed_Phased_20200115_MAF001.fam | sort -k1,1)

## Check line types

awk '{print $9}' iPSC_PRS/HipSci/Genotypes/Initial_Cell_Lines.ped | sort | uniq -c

#
#     60 Fibroblast
#    215 iPSC
#      5 Pbmc
#

```

### SNP-based Heritability of CrossDisorder w/o 23andMe

The whole-genome cross disorder data used for PRS does not contain participants from 23andMe.
This means that the parameters of this data were not reported in the original publication.
Need to estimate heritability using LD Score itself

```{sh Calculate SNP-heritability of CXD}

cd iPSC_PRS/HipSci/Analysis
mkdir LDSC
cd LDSC

## INFO column is a single comma separated column of multiple values
## Split this into separate columns for processing by LDSC

cat <(echo "CHROM POS ID REF ALT BETA SE PVAL FCAS FCON IMPINFO1 IMPINFO2 IMPINFO3 IMPINFO4 IMPINFO5 IMPINFO6 IMPINFO7 IMPINFO8 NCAS NCON") \
<(awk 'NR > 1 {print $1, $2, $3, $4, $5, $6, $7, $8, $10, $11, $12, $14, $15}' ../../Sumstats/pgc_cdg2_meta_no23andMe_oct2019_v2.txt.daner.txt | tr ',' ' ') \
> pgc_cdg2_meta_no23andMe_oct2019_v2_FOR_LDSC

## Take average of info columns

R InfoAverage.R

### Munge CDX sumstats

python \
Software/ldscore_2020_06/ldsc/munge_sumstats.py \
--sumstats pgc_cdg2_meta_no23andMe_oct2019_v2_FOR_LDSC_AV_INFO \
--N-cas-col NCAS \
--N-con-col NCON \
--merge-alleles Software/ldscore_2020_06/w_hm3.snplist \
--snp ID \
--a1 ALT \
--a2 REF \
--p PVAL \
--frq FCON \
--info INFO \
--out pgc_cdg2_meta_no23andMe_oct2019_v2_MUNGED

## Lose quite a lot of SNPs here (279789 left) due to N difference - check how much difference this makes

python \
Software/ldscore_2020_06/ldsc/munge_sumstats.py \
--sumstats pgc_cdg2_meta_no23andMe_oct2019_v2_FOR_LDSC_AV_INFO \
--N-cas-col NCAS \
--N-con-col NCON \
--merge-alleles Software/ldscore/w_hm3.snplist \
--snp ID \
--a1 ALT \
--a2 REF \
--p PVAL \
--frq FCON \
--info INFO \
--out pgc_cdg2_meta_no23andMe_oct2019_v2_NO_N_CUTOFF_MUNGED \
--n-min 1

## This has 930116 SNPs - compare results

### Calculate h2 - sample prev is 162151 / 438997

python \
Software/ldscore_2020_06/ldsc/ldsc.py \
--h2 pgc_cdg2_meta_no23andMe_oct2019_v2_MUNGED.sumstats.gz \
--ref-ld-chr Software/ldscore_2020_06/eur_ref_ld_chr/ \  
--w-ld-chr Software/ldscore_2020_06/eur_w_ld_chr/ \
--out pgc_cdg2_meta_no23andMe_oct2019_v2_LDSC_K0.3 \
--samp-prev 0.368 \
--pop-prev 0.3

# Total Liability scale h2: 0.2495 (0.0137)

## Check effect of N cutoff

python \
Software/ldscore_2020_06/ldsc/ldsc.py \
--h2 pgc_cdg2_meta_no23andMe_oct2019_v2_NO_N_CUTOFF_MUNGED.sumstats.gz \
--ref-ld-chr Software/ldscore_2020_06/eur_ref_ld_chr/ \
--w-ld-chr Software/ldscore_2020_06/eur_w_ld_chr/ \
--out pgc_cdg2_meta_no23andMe_oct2019_v2_NO_N_CUTOFF_LDSC_K0.3 \
--samp-prev 0.368 \
--pop-prev 0.3

# Total Liability scale h2: 0.3944 (0.0127)

## Quite a big difference - but the first estimate seems stable, so go with the default LDSC approach

## Check effect of estimate of pop prev at 0.2

python \
Software/ldscore_2020_06/ldsc/ldsc.py \
--h2 pgc_cdg2_meta_no23andMe_oct2019_v2_MUNGED.sumstats.gz \
--ref-ld-chr Software/ldscore_2020_06/eur_ref_ld_chr/ \
--w-ld-chr Software/ldscore_2020_06/eur_w_ld_chr/ \
--out pgc_cdg2_meta_no23andMe_oct2019_v2_LDSC_K0.2 \
--samp-prev 0.368 \
--pop-prev 0.2

# 0.2234 (0.0123)

## Check effect of estimate of pop prev at 0.4

python \
Software/ldscore_2020_06/ldsc/ldsc.py \
--h2 pgc_cdg2_meta_no23andMe_oct2019_v2_MUNGED.sumstats.gz \
--ref-ld-chr Software/ldscore_2020_06/eur_ref_ld_chr/ \
--w-ld-chr Software/ldscore_2020_06/eur_w_ld_chr/ \
--out pgc_cdg2_meta_no23andMe_oct2019_v2_LDSC_K0.4 \
--samp-prev 0.368 \
--pop-prev 0.4

# Total Liability scale h2: 0.264 (0.0145)

## Obeserved scale, for sanity check with full 23andME

python \
Software/ldscore_2020_06/ldsc/ldsc.py \
--h2 pgc_cdg2_meta_no23andMe_oct2019_v2_MUNGED.sumstats.gz \
--ref-ld-chr Software/ldscore_2020_06/eur_ref_ld_chr/ \
--w-ld-chr Software/ldscore_2020_06/eur_w_ld_chr/ \
--out pgc_cdg2_meta_no23andMe_oct2019_v2_LDSC

# Total Observed scale h2: 0.1591 (0.0088)  
# Compare  0.146 (SE 0.0058) as published - comparable

## Observed scale without N cutoff

python \
Software/ldscore_2020_06/ldsc/ldsc.py \
--h2 pgc_cdg2_meta_no23andMe_oct2019_v2_NO_N_CUTOFF_MUNGED.sumstats.gz \
--ref-ld-chr Software/ldscore_2020_06/eur_ref_ld_chr/ \
--w-ld-chr Software/ldscore_2020_06/eur_w_ld_chr/ \
--out pgc_cdg2_meta_no23andMe_oct2019_v2_NO_N_CUTOFF_LDSC

# Total Observed scale h2: 0.2514 (0.0081)

## 0.25 seems a reasonable estimate of vg1 for PRS power analysis

```

```{r InfoAverage.R}

### Load packages

library(data.table)

### Load data

CrossDis<-fread("pgc_cdg2_meta_no23andMe_oct2019_v2_FOR_LDSC")

### Calculate mean

CrossDis$INFO <- rowMeans(CrossDis[, c("IMPINFO1","IMPINFO2","IMPINFO3","IMPINFO4","IMPINFO5","IMPINFO6","IMPINFO7","IMPINFO8")], na.rm = TRUE, dims = 1)

### Write output

write.table(file="pgc_cdg2_meta_no23andMe_oct2019_v2_FOR_LDSC_AV_INFO", CrossDis[,-c("IMPINFO1","IMPINFO2","IMPINFO3","IMPINFO4","IMPINFO5","IMPINFO6","IMPINFO7","IMPINFO8")], quote=F, col.names=T, row.names=F)

```

Quick check that PRS is not associated with pheno

```{R Check_No_Association_Between_PRS_and_Cell_Number.Rscript}

## Load packages

library(dplyr)

## Load data

load("DataForAnalysis.Rdata")

## Get cell count by donor

Donor_Count <- as.data.frame(table(Final$Donor))
names(Donor_Count)[1]<-"Donor"

## Combine data to unique by-donor variables

Donor_Count_Final<-unique(left_join(Donor_Count, Final[,1:30]))

## Run analyses

sink("Association_Between_PRS_and_Cell_Number.txt")
for(prs in c("BMI_Pt_0.001",
             "BMI_Pt_1",
             "CDX_FULL_Pt_0.001",
             "CDX_FULL_Pt_1",
             "CDX_10K_Pt_1",
             "Height_Pt_0.001",
             "Height_Pt_1",
             "SCZ_CLOZUK_Pt_0.05",
             "SCZ_CLOZUK_Pt_1")){
    pheno <- "Freq"
    analysis_formula <- as.formula(paste(pheno,
                                         "~",
                                         prs,
                                         "+ PC1 + PC2 + PC3 + PC4", sep =" "))
    print(analysis_formula)
    glm_output<-glm(analysis_formula, data = Donor_Count_Final, family="gaussian")
    print(summary(glm_output))
}
sink()

```


### Revisions -- demonstrate that 5e-8 is not a useful threshold to add

## Add threshold to CDX_FULL

```{sh PRSice}
##Height
if [ $1 = 1 ]
then
Software/PRSice_v2.3.1.e/PRSice_linux \
-b iPSC_PRS/HipSci/Sumstats/Meta-analysis_Wood_et_al+UKBiobank_2018.txt \
-t iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF \
--type bed \
--no-regress \
--bar-levels 0.001,1 \
--beta \
-o iPSC_PRS/HipSci/Analysis/PRS_Output/Height \
-n 10 \
--fastscore \
--clump-kb 250 \
--clump-p 1 \
--clump-r2 0.1 \
--A1 Tested_Allele \
--A2 Other_Allele \
--ld iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/1KG_Phase3.WG.CLEANED.EUR_MAF001
fi

##BMI
if [ $1 = 2 ]
then
Software/PRSice_v2.3.1.e/PRSice_linux \
-b iPSC_PRS/HipSci/Sumstats/Meta-analysis_Locke_et_al+UKBiobank_2018_UPDATED.txt \
-t iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF \
--type bed \
--no-regress \
--bar-levels 0.001,1 \
--beta \
-o iPSC_PRS/HipSci/Analysis/PRS_Output/BMI \
-n 10 \
--fastscore \
--clump-kb 250 \
--clump-p 1 \
--clump-r2 0.1 \
--A1 Tested_Allele \
--A2 Other_Allele \
--ld iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/1KG_Phase3.WG.CLEANED.EUR_MAF001
fi

#### SUPERSEDED
## SCZ
##  Initial run of SCZ produced a set of variants that are duplicated in the base file - these were removed with extract below
## if [ $1 = 3 ]
## then
## Software/PRSice_v2.3.1.e/PRSice_linux \
## -b iPSC_PRS/HipSci/Sumstats/SCZ2_Cleaned_Sumstats \
## -t iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF \
## --type bed \
## --no-regress \
## --bar-levels 0.1,1 \
## --or \
## -o iPSC_PRS/HipSci/Analysis/PRS_Output/SCZ \
## -n 10 \
## --fastscore \
## --clump-kb 250 \
## --clump-p 1 \
## --clump-r2 0.1 \
## --ld iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/1KG_Phase3.WG.CLEANED.EUR_MAF001 \
## --extract iPSC_PRS/HipSci/Analysis/PRS_Output/SCZ.valid
## fi
####


## SCZ_CLOZUK
## Initial run of SCZ_CLOZUK produced a set of variants that are duplicated in the base file - these were removed with extract below
if [ $1 = 4 ]
then
Software/PRSice_v2.3.1.e/PRSice_linux \
-b iPSC_PRS/HipSci/Sumstats/clozuk_pgc2.clumped.sumstats \
-t iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF \
--type bed \
--no-regress \
--bar-levels 0.05,1 \
--or \
-o iPSC_PRS/HipSci/Analysis/PRS_Output/SCZ_CLOZUK \
-n 10 \
--fastscore \
--clump-kb 250 \
--clump-p 1 \
--clump-r2 0.1 \
--ld iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/1KG_Phase3.WG.CLEANED.EUR_MAF001 \
--extract iPSC_PRS/HipSci/Analysis/PRS_Output/SCZ_CLOZUK.valid
fi

## Cross Disorder Limited
## Initial run of CXD_10K produced a set of variants that are duplicated in the base file - these were removed with extract below
if [ $1 = 5 ]
then
Software/PRSice_v2.3.1.e/PRSice_linux \
-b iPSC_PRS/HipSci/Sumstats/pgc_cdg2_meta_10k_oct2019_v2.txt.daner.txt \
-t iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF \
--type bed \
--no-regress \
--bar-levels 1 \
--beta \
-o iPSC_PRS/HipSci/Analysis/PRS_Output/CXD_10K \
-n 10 \
--fastscore \
--clump-kb 250 \
--clump-p 1 \
--clump-r2 0.1 \
--ld iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/1KG_Phase3.WG.CLEANED.EUR_MAF001 \
--chr CHROM \
--bp POS \
--a1 ALT \
--a2 REF \
--pvalue PVAL \
--snp ID \
--extract iPSC_PRS/HipSci/Analysis/PRS_Output/CXD_10K.valid
fi

## Cross Disorder Full
## Initial run of CDX_FULL produced a set of variants that are duplicated in the base file - these were removed with extract below
if [ $1 = 6 ]
then
Software/PRSice_v2.3.1.e/PRSice_linux \
-b iPSC_PRS/HipSci/Sumstats/pgc_cdg2_meta_no23andMe_oct2019_v2.txt.daner.txt \
-t iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF \
--type bed \
--no-regress \
--bar-levels 0.00000005,0.001,1 \
--beta \
-o iPSC_PRS/HipSci/Analysis/PRS_Output/CDX_FULL \
-n 10 \
--fastscore \
--clump-kb 250 \
--clump-p 1 \
--clump-r2 0.1 \
--ld iPSC_PRS/HipSci/Genotypes/HipSciReimputationTake2.vcfs/1KG_Phase3.WG.CLEANED.EUR_MAF001 \
--chr CHROM \
--bp POS \
--a1 ALT \
--a2 REF \
--pvalue PVAL \
--snp ID \
--extract iPSC_PRS/HipSci/Analysis/PRS_Output/CDX_FULL.valid
fi

```

```{r Analyses}

## Set options
options(stringsAsFactors=F)

## Set seed
set.seed(604)

## Get packages
library(interactions)
library(purrr)
library(tibble)
library(dplyr)
library(tidyr)
library(ggpubr)
library(data.table)
library(lmerTest)
library(HLMdiag)

## Get data
# Read in all cell data files
temp = list.files(pattern="*.tsv")
# Create list of all cell data files
myfiles = lapply(temp, read.delim)

## Create single tibble comprising data needed for analysis
dataframe_data<- tibble(
name = map(myfiles, "name"),
experiment_name = map(myfiles, "experiment_name"),
cell_area = map(myfiles, "cell_area"),
cell_roundness = map(myfiles, "roundness"),
cell_ratio_w2l = map(myfiles, "ratio_w2l"),
p_col = map(myfiles, "p_col"),
p_row = map(myfiles, "p_row"),
field = map(myfiles, "field"),
clump_size = map(myfiles, "clump_size"))

## Convert lists to a data frame of all entries
Cell_Tibble<-dataframe_data %>%
select(name, experiment_name, cell_area, cell_roundness, cell_ratio_w2l, p_col, p_row, field, clump_size) %>%
unnest(cols=c(name, experiment_name, cell_area, cell_roundness, cell_ratio_w2l, p_col, p_row, field, clump_size))

## Some rows in Cell Tibble are repeated due to repeat assessments - remove these

Cell_Tibble_Distinct <- distinct(Cell_Tibble)

## Get external info on cell lines

temp_plate = list.files(path="iPSC_PRS/HipSci/Images/Plate_Results", pattern="*.tsv", full.names=T)
myfiles_plate = lapply(temp_plate, read.delim)

### Plate fields

    # name                   The official name for each line as used within HipSci.  This is especially needed to link to other data from HipSci.

    # experiment_name        We assigned a number to each experiment as an internal record     
    # passage_number         Cell line passage number at the time of the experiment
    # p_row                  number corresponding to position of wells in row (1-8)
    # p_col    		     number corresponding to position of wells in column (A-H)

    # num_cells              number of objects segmented per well

    # nuc_area_mean          morphological value, mean per well
    # nuc_area_sd            morphological value, standard dev per well
    # nuc_area_sum           morphological value, sum per well
    # nuc_area_max           morphological value, max value
    # nuc_area_min           morphological value, min value
    # nuc_area_median        morphological value, median per well

    # nuc_roundness_mean     morphological value, mean per well
    # nuc_roundness_sd       morphological value, st dev per well
    # nuc_roundness_sum      morphological value, sum per well
    # nuc_roundness_max      morphological value, max per well
    # nuc_roundness_min      morphological value, min per well
    # nuc_roundness_median   morphological value, median per well

    # nuc_ratio_w2l_mean     morphological value, mean per well
    # nuc_ratio_w2l_sd       morphological value, standard dev per well
    # nuc_ratio_w2l_sum      morphological value, sum per well
    # nuc_ratio_w2l_max      morphological value, max value
    # nuc_ratio_w2l_min      morphological value, min value
    # nuc_ratio_w2l_median   morphological value, median per well

    # edu_median_mean        intensity value, mean per well
    # edu_median_sd          intensity value, median per well
    # edu_median_sum         intensity value, sum per well
    # edu_median_max         intensity value, max per well
    # edu_median_min         intensity value, min per well
    # edu_median_median      intensity value, median per well

    # oct4_median_mean       N/A
    # oct4_median_sd         N/A
    # oct4_median_sum        N/A
    # oct4_median_max        N/A
    # oct4_median_min        N/A
    # oct4_median_median     N/A

    # inten_nuc_dapi_median_mean    intensity value, mean per well
    # inten_nuc_dapi_median_sd      intensity value, st dev per well
    # inten_nuc_dapi_median_sum     intensity value, median per well
    # inten_nuc_dapi_median_max     intensity value, max per well
    # inten_nuc_dapi_median_min     intensity value, min per well
    # inten_nuc_dapi_median_median  intensity value, median per well

    # cells_per_clump_mean          context value, mean per well
    # cells_per_clump_sd            context value, st dev per well
    # cells_per_clump_sum           context value, sum per well
    # cells_per_clump_max           context value, max per well
    # cells_per_clump_min           context value, min per well
    # cells_per_clump_median        context value, median per well

    # area_mean              morphological value, mean per well
    # area_sd                morphological value, st dev per well
    # area_sum               morphological value, sum per well
    # area_max               morphological value, max per well
    # area_min               morphological value, min per well
    # area_median            morphological value, median per well

    # roundness_mean         morphological value, mean per well
    # roundness_sd           morphological value, st dev per well
    # roundness_sum          morphological value, sum per well
    # roundness_max          morphological value, max per well
    # roundness_min          morphological value, min per well
    # roundness_median       morphological value, median per well

    # ratio_w2l_mean         morphological value, mean per well
    # ratio_w2l_sd           morphological value, st dev per well
    # ratio_w2l_sum          morphological value, sum per well
    # ratio_w2l_max          morphological value, max per well
    # ratio_w2l_min          morphological value, min per well
    # ratio_w2l_median       morphological value, median per well

    # compound               FN=Fibronectin
    # concentration          1 - 5 - 25 ug/ml (0 for wells N/A)
    # cell_count             number of cells plated, typically 3000
    # num_fields             9

## Get fibronectin conc from plate file - can match on experiment, p_col and p_row

## Create single tibble comprising lists of names and six proposed phenos
dataframe_data_plate<- tibble(
name = map(myfiles_plate, "name"),
experiment_name = map(myfiles_plate, "experiment_name"),
p_col = map(myfiles_plate, "p_col"),
p_row = map(myfiles_plate, "p_row"),
fibro_conc = map(myfiles_plate, "concentration"))

## Convert lists to a data frame of all entries
Plate_Tibble<-dataframe_data_plate %>%
select(name, experiment_name, p_col, p_row, fibro_conc) %>%
unnest(cols=c(name, experiment_name, p_col, p_row, fibro_conc))

## Some rows in Plate Tibble are repeated due to repeat assessments - remove these

Plate_Tibble_Distinct <- distinct(Plate_Tibble)

## Add fibronectin to Cell Tibble

Cell_Tibble_Distinct$name_exp_col_row<-paste(Cell_Tibble_Distinct$name, Cell_Tibble_Distinct$experiment_name, Cell_Tibble_Distinct$p_col, Cell_Tibble_Distinct$p_row, sep="_")
Plate_Tibble_Distinct$name_exp_col_row<-paste(Plate_Tibble_Distinct$name, Plate_Tibble_Distinct$experiment_name, Plate_Tibble_Distinct$p_col, Plate_Tibble_Distinct$p_row, sep="_")

Plate_Tibble_Distinct_Short<-Plate_Tibble_Distinct[,c("name_exp_col_row", "fibro_conc")]

Cell_Tibble_Distinct_Fibro<-left_join(Cell_Tibble_Distinct, Plate_Tibble_Distinct_Short)

dim(Cell_Tibble_Distinct)
dim(Cell_Tibble_Distinct_Fibro)

## Add PRS and genomic PCs

BMI_PRS<-fread("../../Analysis/PRS_Output/BMI.all_score")
CDX_FULL_PRS<-fread("../../Analysis/PRS_Output/CDX_FULL.all_score")
CDX_10K_PRS<-fread("../../Analysis/PRS_Output/CXD_10K.all_score")
Height_PRS<-fread("../../Analysis/PRS_Output/Height.all_score")
SCZ_CLOZUK_PRS<-fread("../../Analysis/PRS_Output/SCZ_CLOZUK.all_score")

names(BMI_PRS)[3:4]<-paste("BMI",names(BMI_PRS)[3:4],sep="_")
names(CDX_FULL_PRS)[3:4]<-paste("CDX_FULL",names(CDX_FULL_PRS)[3:4],sep="_")
names(CDX_10K_PRS)[3]<-paste("CDX_10K",names(CDX_10K_PRS)[3],sep="_")
names(Height_PRS)[3:4]<-paste("Height",names(Height_PRS)[3:4],sep="_")
names(SCZ_CLOZUK_PRS)[3:4]<-paste("SCZ_CLOZUK",names(SCZ_CLOZUK_PRS)[3:4],sep="_")

PRS_TEMP1<-full_join(BMI_PRS, CDX_FULL_PRS)
PRS_TEMP2<-full_join(PRS_TEMP1, CDX_10K_PRS)
PRS_TEMP3<-full_join(PRS_TEMP2, Height_PRS)
PRS<-full_join(PRS_TEMP3, SCZ_CLOZUK_PRS)

PCs<-fread("../../Genotypes/HipSciReimputationTake2.vcfs/WG_INFO9_MAF5_PLINKMAF_LD3.pca_For_R.txt")

PRS_PCs<-full_join(PRS, PCs)

## External data from Vigilante / Kilpinen

## Combine all data

# Separate name into donor and line

Cell_Tibble_Distinct_Fibro<-separate(Cell_Tibble_Distinct_Fibro, "name", into=c("Donor","line"), sep="_", remove=F)
Cell_Tibble_Distinct_Fibro$line<-as.numeric(Cell_Tibble_Distinct_Fibro$line)

# Rename FID and IID to Donor and line

names(PRS_PCs)[1:2]<-c("Donor","line")

# Remove line from PRS data

PRS_PCs_No_Line<-select(PRS_PCs,-grep("line", names(PRS_PCs)))


## Some lines are biological replicates of others (that is, the same donor across multiple lines)
## Change the names of these lines so that "donor" reflects all lines derived from that genome
## Cell_Tibble_Distinct_Fibro has no lines over 40, so add new lines for 40+

# # Donor fpdl is 4 donors:
#     HPSI0813i-fpdl_1
#     HPSI0913i-ffdl_1 [EXCLUDE --> Duplicate_Lines.txt]
#     HPSI1213i-foqj_2 [EXCLUDE --> Duplicate_Lines.txt]
#     HPSI0913i-ffdz_1 {EXCLUDE --> Duplicate_Lines.txt]
# # Donor fpdr is 3 donors:
#     HPSI0813i-ffdr_3 [EXCLUDE --> Duplicate_Lines.txt]
#     HPSI0813i-fpdr_1
#     HPSI1113i-fumb_1 [EXCLUDE --> Duplicate_Lines.txt]
# # Donor fpdk is 3 donors:
#     HPSI0813i-fpdk_2
#     HPSI1113i-zuta_1 [EXCLUDE --> Duplicate_Lines.txt]
# # Donor ffdj is 3 donors:
#     HPSI0813i-ffdj_1 [EXCLUDE --> Duplicate_Lines.txt]
#     HPSI0813i-fpdj_1
#     HPSI1113i-nibo_1 [EXCLUDE --> Duplicate_Lines.txt]
# # Donor ffdm is 4 donors:
#     HPSI0813i-ffdm_1 [EXCLUDE --> Duplicate_Lines.txt]
#     HPSI0813i-fpdm_1

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI0913i-ffdl", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0913i-ffdl" & line == 1), 40, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0913i-ffdl" & line == 2), 41, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0913i-ffdl"), "HPSI0813i-fpdl", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI1213i-foqj", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1213i-foqj" & line == 2), 42, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1213i-foqj"), "HPSI0813i-fpdl", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI0913i-ffdz", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0913i-ffdz" & line == 2), 43, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0913i-ffdz"), "HPSI0813i-fpdl", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI0813i-ffdr", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0813i-ffdr" & line == 3), 44, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0813i-ffdr"), "HPSI0813i-fpdr", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI1113i-fumb", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1113i-fumb" & line == 2), 45, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1113i-fumb"), "HPSI0813i-fpdr", Cell_Tibble_Distinct_Fibro$Donor)

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1113i-zuta" & line == 1), 46, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1113i-zuta"), "HPSI0813i-fpdk", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI0813i-ffdj", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0813i-ffdj" & line == 1), 47, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0813i-ffdj"), "HPSI0813i-fpdj", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI1113i-nibo", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1113i-nibo" & line == 2), 48, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI1113i-nibo"), "HPSI0813i-fpdj", Cell_Tibble_Distinct_Fibro$Donor)

table(with(Cell_Tibble_Distinct_Fibro,Cell_Tibble_Distinct_Fibro[Donor == "HPSI0813i-ffdm", "name"]), useNA="a")

Cell_Tibble_Distinct_Fibro$line <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0813i-ffdm" & line == 3), 49, Cell_Tibble_Distinct_Fibro$line)
Cell_Tibble_Distinct_Fibro$Donor <- ifelse(with(Cell_Tibble_Distinct_Fibro, Donor == "HPSI0813i-ffdm"), "HPSI0813i-fpdm", Cell_Tibble_Distinct_Fibro$Donor)

# Merge data

Final_Temp1<-left_join(PRS_PCs_No_Line, Cell_Tibble_Distinct_Fibro)

# Confirm expected reduction in length

dim(Final_Temp1)

# Confirm 60 donors

length(unique(Final_Temp1$Donor))

# Confirm >60 names, implying multiple lines for some donors

length(unique(Final_Temp1$name))

## Add external data - tissue type, reprogramming and feeder from HipSci database / Vigilante Supplement --> CellOfOriginInfo.txt

CellOfOrigin <- fread("CellOfOriginInfo.txt")

Final<-left_join(Final_Temp1, CellOfOrigin)

Final$derived_from_tissue_type <- as.factor(Final$derived_from_tissue_type)
Final$reprogramming <- as.factor(Final$reprogramming)
Final$on_feeder <- as.factor(Final$on_feeder)

names(Final)[names(Final) == "CDX_FULL_Pt_5e-08"] <- "CDX_FULL_Pt_GWS"
names(Final)[names(Final) == "Pt_1"] <- "CDX_FULL_Pt_1"

## Clean up

rm(BMI_PRS,
CDX_10K_PRS,
CDX_FULL_PRS,
CellOfOrigin,
Cell_Tibble,
Cell_Tibble_Distinct,
Cell_Tibble_Distinct_Fibro,
dataframe_data,
dataframe_data_plate,
Final_Temp1,
Height_PRS,
myfiles,
myfiles_plate,
PCs,
Plate_Tibble,
Plate_Tibble_Distinct,
Plate_Tibble_Distinct_Short,
PRS,
PRS_PCs,
PRS_PCs_No_Line,
PRS_TEMP1,
PRS_TEMP2,
PRS_TEMP3,
SCZ_CLOZUK_PRS,
temp,
temp_plate)

## Save set-up

save.image("MLM_SetUp_Revisions.Rdata")

### Examine data structure

## Is continuous data approximately normal and appropriate for scaling

pdf("Distributions_iPSC_Revisions.pdf")
hist(Final$BMI_Pt_0.001, 100)
hist(Final$BMI_Pt_1,100)
hist(Final$CDX_FULL_Pt_GWS,100)
hist(Final$CDX_FULL_Pt_0.001,100)
hist(Final$CDX_FULL_Pt_1,100)
hist(Final$CDX_10K_Pt_1,100)
hist(Final$Height_Pt_0.001,100)
hist(Final$Height_Pt_1,100)
hist(Final$SCZ_CLOZUK_Pt_0.05,100)
hist(Final$SCZ_CLOZUK_Pt_1,100)
hist(Final$PC1,100)
hist(Final$PC2,100)
hist(Final$PC3,100)
hist(Final$PC4,100)
hist(Final$PC5,100)
hist(Final$PC6,100)
hist(Final$PC7,100)
hist(Final$PC8,100)
hist(Final$PC9,100)
hist(Final$PC10,100)
hist(Final$PC11,100)
hist(Final$PC12,100)
hist(Final$PC13,100)
hist(Final$PC14,100)
hist(Final$PC15,100)
hist(Final$PC16,100)
hist(Final$PC17,100)
hist(Final$PC18,100)
hist(Final$PC19,100)
hist(Final$PC20,100)
hist(Final$cell_area,100)
hist(Final$cell_roundness,100)
hist(Final$cell_ratio_w2l,100)
hist(Final$clump_size,100)
dev.off()

## Scale continuous data for analysis

Final$BMI_Pt_0.001<-scale(Final$BMI_Pt_0.001)
Final$BMI_Pt_1<-scale(Final$BMI_Pt_1)
Final$CDX_FULL_Pt_GWS<-scale(Final$CDX_FULL_Pt_GWS)
Final$CDX_FULL_Pt_0.001<-scale(Final$CDX_FULL_Pt_0.001)
Final$CDX_FULL_Pt_1<-scale(Final$CDX_FULL_Pt_1)
Final$CDX_10K_Pt_1<-scale(Final$CDX_10K_Pt_1)
Final$Height_Pt_0.001<-scale(Final$Height_Pt_0.001)
Final$Height_Pt_1<-scale(Final$Height_Pt_1)
Final$SCZ_CLOZUK_Pt_0.05<-scale(Final$SCZ_CLOZUK_Pt_0.05)
Final$SCZ_CLOZUK_Pt_1<-scale(Final$SCZ_CLOZUK_Pt_1)
Final$PC1<-scale(Final$PC1)
Final$PC2<-scale(Final$PC2)
Final$PC3<-scale(Final$PC3)
Final$PC4<-scale(Final$PC4)
Final$PC5<-scale(Final$PC5)
Final$PC6<-scale(Final$PC6)
Final$PC7<-scale(Final$PC7)
Final$PC8<-scale(Final$PC8)
Final$PC9<-scale(Final$PC9)
Final$PC10<-scale(Final$PC10)
Final$PC11<-scale(Final$PC11)
Final$PC12<-scale(Final$PC12)
Final$PC13<-scale(Final$PC13)
Final$PC14<-scale(Final$PC14)
Final$PC15<-scale(Final$PC15)
Final$PC16<-scale(Final$PC16)
Final$PC17<-scale(Final$PC17)
Final$PC18<-scale(Final$PC18)
Final$PC19<-scale(Final$PC19)
Final$PC20<-scale(Final$PC20)
Final$cell_area<-scale(Final$cell_area)
Final$cell_roundness<-scale(Final$cell_roundness)
Final$cell_ratio_w2l<-scale(Final$cell_ratio_w2l)
Final$clump_size<-scale(Final$clump_size)

### Note to self - the below requires ~16 threads (more than 8, anyway...) for interaction

### Initial analyses miscoded fibro_conc as linear, and did not account for NA fibro concs.

### Drop fibro_conc == 0

dim(Final)
#1579905      47

Final <- Final[Final$fibro_conc > 0, ]

dim(Final)
#1543624      47

### Set fibro_conc as a factor

Final$fibro_conc <- as.factor(Final$fibro_conc)

### Analysis Set 1 - no interaction
### 2020-10-02 Initial analyses used line as the nested factor, and did not consider well-level effects. Replaced line with well

sink(paste("Analysis_No_Interaction_",Sys.Date(),"_Revisions.txt",sep=""))

for(prs in c("CDX_FULL_Pt_GWS",
             "CDX_FULL_Pt_0.001",
             "CDX_FULL_Pt_1")){
    pheno <- "cell_area"
        ## Define formula
        analysis_formula <- as.formula(paste(pheno,
                                             "~",
                                             prs,
                                             "+ fibro_conc + PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder + (1 | Donor/name_exp_col_row)", sep =" "))
        print(analysis_formula)
        ## Run mixed linear model (~ 40-60 seconds)
        lmer_output_REML<-lmer(analysis_formula, data = Final, REML = T)
        ## Print summary of lmer, p from Satterthwaite approximation
        print(summary(lmer_output_REML))
        ## Rerun lmer using ML for likelihood ratio test (LRT) [still declare REML - see https://ourcodingclub.github.io/tutorials/mixed-models/]
        lmer_output_ML <- lmer(analysis_formula, data = Final, REML = F)
        ## Create model without PRS for likelihood ratio test (LRT)
        null_formula<-as.formula(paste(pheno,"~ fibro_conc + PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder + (1 | Donor/name_exp_col_row)", sep =" "))
        lmer_null_ML<-lmer(null_formula, data = Final, REML = F)
        ## Perform LRT (no bootstrap)
        print(anova(lmer_null_ML, lmer_output_ML))
}

sink()

save.image("DataForAnalysis_Revisions.Rdata")

### Interesting results are seen for CDX_FULL_Pt_1 and cell area.
### There are others significant at p < 0.05, but these are expected under the null - the most interesting of these is for CDX_FULL_Pt_1 and cell roundness.

# Linear mixed model fit by REML. t-tests use Satterthwaite's method [lmerModLmerTest]
# Formula: analysis_formula
#    Data: Final

# REML criterion at convergence: 4035831

# Scaled residuals:
#     Min      1Q  Median      3Q     Max
# -2.7710 -0.6586 -0.1695  0.4700  6.0209

# Random effects:
#  Groups                 Name        Variance Std.Dev.
#  name_exp_col_row:Donor (Intercept) 0.07389  0.2718
#  Donor                  (Intercept) 0.03851  0.1962
#  Residual                           0.79497  0.8916
# Number of obs: 1543624, groups:  name_exp_col_row:Donor, 2484; Donor, 60

# Fixed effects:
#                                Estimate Std. Error         df  t value Pr(>|t|)
# (Intercept)                  -4.887e-01  4.755e-02  2.481e+02  -10.278  < 2e-16 ***
# CDX_FULL_Pt_1                 8.450e-02  2.768e-02  5.150e+01    3.052  0.00359 ** 
# fibro_conc5                   4.070e-01  1.367e-02  2.438e+03   29.784  < 2e-16 ***
# fibro_conc25                  7.435e-01  1.366e-02  2.432e+03   54.440  < 2e-16 ***
# PC1                          -3.276e-02  2.810e-02  5.017e+01   -1.166  0.24918    
# PC2                          -1.454e-02  2.666e-02  5.130e+01   -0.545  0.58779    
# PC3                           4.645e-02  2.634e-02  5.166e+01    1.763  0.08378 .  
# PC4                           5.099e-02  2.678e-02  5.554e+01    1.904  0.06209 .  
# clump_size                   -1.913e-01  1.694e-03  1.399e+06 -112.950  < 2e-16 ***
# derived_from_tissue_typepbmc -4.086e-02  4.298e-02  1.664e+03   -0.951  0.34198    
# reprogrammingsendai           9.711e-02  4.389e-02  9.128e+02    2.213  0.02718 *  
# on_feederfeeder-free         -4.592e-01  3.505e-02  1.923e+03  -13.101  < 2e-16 ***
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# Correlation of Fixed Effects:
#             (Intr) CDX_FU fbr_c5 fbr_25 PC1    PC2    PC3    PC4    clmp_s drv___ rprgrm
# CDX_FULL_P_  0.062							   		 
# fibro_conc5 -0.146  0.000						   		 
# fibro_cnc25 -0.146  0.000  0.508					   		 
# PC1          0.041  0.013 -0.001  0.000					   		 
# PC2          0.035 -0.091  0.000  0.000  0.003				   		 
# PC3         -0.068  0.164  0.000  0.000  0.004 -0.009			   		 
# PC4         -0.187  0.090  0.000  0.000 -0.013 -0.012  0.023		   		 
# clump_size   0.010  0.000 -0.007 -0.007  0.000  0.000  0.001  0.000	   		 
# drvd_frm_t_  0.513  0.060 -0.001 -0.001  0.056  0.011  0.050 -0.150  0.000 		 
# rprgrmmngsn -0.807 -0.055  0.001  0.001 -0.045 -0.039 -0.035  0.213  0.000 -0.657	 
# on_fdrfdr-f -0.092  0.001 -0.008 -0.008 -0.022  0.115  0.063  0.007 -0.004 -0.015  0.028

### Analysis Set 2 - interactions, including plotting

sink(paste("Analysis_With_Interaction_",Sys.Date(),".txt",sep=""))

for(prs in c("BMI_Pt_0.001",
             "BMI_Pt_1",
             "CDX_FULL_Pt_0.001",
             "CDX_FULL_Pt_1",
             "CDX_10K_Pt_1",
             "Height_Pt_0.001",
             "Height_Pt_1",
             "SCZ_CLOZUK_Pt_0.05",
             "SCZ_CLOZUK_Pt_1")){
    for(pheno in c("cell_area", "cell_roundness", "cell_ratio_w2l")){
        ## Define formula
        interaction_analysis_formula <- as.formula(paste(pheno,
                                             "~",
                                             prs,
                                             "*(fibro_conc + PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder) +
					      fibro_conc*(PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder) + (1 | Donor/name_exp_col_row)", sep =" "))
        print(interaction_analysis_formula)
        ## Run mixed linear model
        lmer_output_REML<-lmer(interaction_analysis_formula, data = Final, REML = T)
        ## Print summary of lmer, p from Satterthwaite approximation)
        print(summary(lmer_output_REML))
        ## Rerun lmer using ML for likelihood ratio test (LRT) [still declare REML - see https://ourcodingclub.github.io/tutorials/mixed-models/]
        lmer_output_ML <- lmer(interaction_analysis_formula, data = Final, REML = F)
        ## Create model without PRS for likelihood ratio test (LRT)
        null_formula<-as.formula(paste(pheno,
				       "~",
                                       prs,
                                       "*(PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder) +
                                       fibro_conc*(PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder) + (1 | Donor/name_exp_col_row)", sep =" "))
        lmer_null_ML<-lmer(null_formula, data = Final, REML = F)
        ## Perform LRT (no bootstrap)
        print(anova(lmer_null_ML, lmer_output_ML))
    }
}

sink()

### Visualise the data to enable understanding

for(prs in c("BMI_Pt_0.001",
             "BMI_Pt_1",
             "CDX_FULL_Pt_0.001",
             "CDX_FULL_Pt_1",
             "CDX_10K_Pt_1",
             "Height_Pt_0.001",
             "Height_Pt_1",
             "SCZ_CLOZUK_Pt_0.05",
             "SCZ_CLOZUK_Pt_1")){
    for(pheno in c("cell_area", "cell_roundness", "cell_ratio_w2l")){
        ## Define formula
        interaction_analysis_formula <- as.formula(paste(pheno,
                                             "~",
                                             prs,
                                             "*(fibro_conc + PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder) +
                                              fibro_conc*(PC1 + PC2 + PC3 + PC4 + clump_size + derived_from_tissue_type + reprogramming + on_feeder) + (1 | Donor/name_exp_col_row)", sep =" "))
        print(interaction_analysis_formula)
        ## Run mixed linear model
        lmer_output_REML<-lmer(interaction_analysis_formula, data = Final, REML = T)

	## Add fit statistics to Final

	Final$fit <- predict(lmer_output_REML)

	Final_agg <- Final %>% 
		     group_by(Donor, BMI_Pt_0.001, BMI_Pt_1, CDX_FULL_Pt_0.001, CDX_FULL_Pt_1, CDX_10K_Pt_1, Height_Pt_0.001, Height_Pt_1, SCZ_CLOZUK_Pt_0.05, SCZ_CLOZUK_Pt_1, fibro_conc) %>% 
		     summarise(mean_fit = mean(fit))
	Final_TEMP<-suppressWarnings(left_join(Final, Final_agg))
	Final_TEMP<-unique(Final_TEMP[,c("Donor",prs,"mean_fit","fibro_conc")])
	names(Final_TEMP)[3:4]<-c(pheno,"modx_group")

	pheno_pretty<-ifelse(pheno == "cell_area", "Cell area",
		      ifelse(pheno == "cell_roundness", "Cell roundness",
		      ifelse(pheno == "cell_ratio_w2l", "Cell width:length", NA)))
	
	prs_pretty<-ifelse(prs == "BMI_Pt_0.001", "BMI PGS (pThresh < 0.001)",
		    ifelse(prs == "BMI_Pt_1", "BMI PGS (pThresh < 1)",
		    ifelse(prs == "CDX_FULL_Pt_0.001", "Cross-psychiatric PGS (pThresh < 0.001)",
		    ifelse(prs == "CDX_FULL_Pt_1", "Cross-psychiatric PGS (pThresh < 1)",
             	    ifelse(prs == "CDX_10K_Pt_1", "Cross-psychiatric + 23andME PGS",
             	    ifelse(prs == "Height_Pt_0.001", "Height PGS (pThresh < 0.001)",
             	    ifelse(prs == "Height_Pt_1", "Height PGS (pThresh < 1)",
             	    ifelse(prs == "SCZ_CLOZUK_Pt_0.05", "Schiziophrenia PGS (pThresh < 0.05)",
             	    ifelse(prs == "SCZ_CLOZUK_Pt_1", "Schiziophrenia PGS (pThresh < 1)", NA)))))))))

	pdfname<-paste("SplitEffectPlot_",prs,"_",pheno,".pdf",sep="")
	pdf(pdfname)
	    print(interact_plot(lmer_output_REML,pred = !!prs, modx = fibro_conc, interval=T, colors="CUD", legend.main="Fibronectin\nConcentration", x.label = prs_pretty, y.label = pheno_pretty) +
	    geom_point(data=Final_TEMP, aes(col = modx_group, shape = modx_group)) +
	    xlim(c(-4,4)) +
	    ylim(c(-1.5,1.5)) +
	    theme(text = element_text(size = 14)) + 
	    guides(shape = FALSE, fill=FALSE))
	dev.off()
    }
}


```
